{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brune/miniconda3/envs/sprg/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.classes.gapfill import KEGGapFiller\n",
    "from refinegems.utility.io import load_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435.xml'\n",
    "\n",
    "model = load_model(mpath,'libsbml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'get_missing_genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m gff2 \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435_RefSeq.gff\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 2\u001b[0m missing_genes \u001b[38;5;241m=\u001b[39m \u001b[43mget_missing_genes\u001b[49m(gff2,load_model(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435.xml\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlibsbml\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m      3\u001b[0m missing_genes\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# @TODO : some od the genes do not have an old locus tag -> what to do?\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'get_missing_genes' is not defined"
     ]
    }
   ],
   "source": [
    "gff2 = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435_RefSeq.gff'\n",
    "missing_genes = get_missing_genes(gff2,load_model('/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435.xml','libsbml'))\n",
    "missing_genes\n",
    "\n",
    "# @TODO : some od the genes do not have an old locus tag -> what to do?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brune/miniconda3/envs/sprg/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.classes.gapfill import GeneGapFiller\n",
    "from refinegems.utility.io import load_model\n",
    "\n",
    "modelpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435.xml'\n",
    "model = load_model(modelpath,'libsbml')\n",
    "cmodel = load_model(modelpath,'cobra')\n",
    "\n",
    "gffpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435_RefSeq.gff'\n",
    "\n",
    "gf2 = GeneGapFiller()\n",
    "gf2_missing_genes = gf2.get_missing_genes(gffpath,model)\n",
    "# ncbiprotein | locus_tag | ec-code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ec-code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_011274359.1</td>\n",
       "      <td>SH0001</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_011274360.1</td>\n",
       "      <td>SH0002</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_172458781.1</td>\n",
       "      <td>SH0003</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_011274362.1</td>\n",
       "      <td>SH0004</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_011274363.1</td>\n",
       "      <td>SH0005</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>WP_145424740.1</td>\n",
       "      <td>pSHaeC05</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2574</th>\n",
       "      <td>WP_011276926.1</td>\n",
       "      <td>pSHaeC06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>WP_011276927.1</td>\n",
       "      <td>pSHaeC07</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>WP_011276928.1</td>\n",
       "      <td>pSHaeC08</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2577</th>\n",
       "      <td>WP_011276930.1</td>\n",
       "      <td>pSHaeC10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1692 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ncbiprotein locus_tag ec-code\n",
       "0     WP_011274359.1    SH0001    None\n",
       "1     WP_011274360.1    SH0002    None\n",
       "2     WP_172458781.1    SH0003    None\n",
       "3     WP_011274362.1    SH0004    None\n",
       "4     WP_011274363.1    SH0005    None\n",
       "...              ...       ...     ...\n",
       "2573  WP_145424740.1  pSHaeC05    None\n",
       "2574  WP_011276926.1  pSHaeC06    None\n",
       "2575  WP_011276927.1  pSHaeC07    None\n",
       "2576  WP_011276928.1  pSHaeC08    None\n",
       "2577  WP_011276930.1  pSHaeC10    None\n",
       "\n",
       "[1692 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gf2_missing_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from refinegems.classes.gapfill import map_ec_to_reac\n",
    "from refinegems.curation.db_access.db import get_ec_from_ncbi, get_ec_via_swissprot\n",
    "# for notebook, later should be part of object\n",
    "manual_curation = {}\n",
    "\n",
    "# find missing reaction\n",
    "# @TODO one fildering step towards the end is still missing\n",
    "def get_missing_reacs(missing_genes, \n",
    "                      # NCBI params\n",
    "                      mail:str=None, check_NCBI:bool=False,\n",
    "                      # SwissProt\n",
    "                      fasta:str=None, dmnd_db:str=None, swissprot_map:str=None,\n",
    "                      **kwargs):\n",
    "    \n",
    "    # Case 1:  no EC\n",
    "    # --------------\n",
    "    case_1 = missing_genes[missing_genes['ec-code'].isna()]\n",
    "    not_case_1 = missing_genes[~missing_genes['ec-code'].isna()]\n",
    "    if len(case_1) > 0:\n",
    "        \n",
    "        # Option 1: BLAST against SwissProt\n",
    "        # +++++++++++++++++++++++++++++++++    \n",
    "        # -> BLAST (DIAMOND) against SwissProt to get EC/BRENDA \n",
    "        # @TEST\n",
    "        if fasta and dmnd_db and swissprot_map:\n",
    "            case_1_mapped = get_ec_via_swissprot(fasta,dmnd_db,\n",
    "                                          case_1,\n",
    "                                          swissprot_map,\n",
    "                                          **kwargs) # further optional params for the mapping\n",
    "            case_1.drop('ec-code', inplace=True, axis=1)\n",
    "            case_1 = case_1.merge(case_1_mapped, on='locus_tag', how='left')\n",
    "            not_case_1['UniProt'] = None\n",
    "            \n",
    "        # Option 2: Use ML to predict EC\n",
    "        # ++++++++++++++++++++++++++++++\n",
    "        # @TODO\n",
    "        # -> sth like DeepECTransformer (tools either not good, \n",
    "        #    not installable or no license)\n",
    "        # -> use ECRECer Web service output as input \n",
    "        #    (whole protein fasta -> wait for Gwendolyn's results) -> does not work / not return\n",
    "        # -> use CLEAN webservice\n",
    "        #    same problem as above with the web tool\n",
    "\n",
    "    manual_curation['no ncbiprotein, no EC'] = case_1[case_1['ncbiprotein'].isna() & case_1['ec-code'].isna()] \n",
    "    \n",
    "    mapped_reacs = pd.concat([case_1,not_case_1])\n",
    "\n",
    "    # Case 2: still no EC but ncbiprotein\n",
    "    # -----------------------------------\n",
    "    #       -> access ncbi for ec (optional) \n",
    "    if check_NCBI and mail:\n",
    "        mapped_reacs['ec-code'] = mapped_reacs.progress_apply(lambda x: get_ec_from_ncbi(mail,x['ncbiprotein']) if not x['ec-code'] else x['ec-code'], axis=1)\n",
    "    \n",
    "    manual_curation['no EC'] = mapped_reacs[mapped_reacs['ec-code'].isna()]\n",
    "    mapped_reacs = mapped_reacs[~mapped_reacs['ec-code'].isna()]\n",
    "\n",
    "    # Case 3: EC found\n",
    "    # ----------------\n",
    "    # @TODO : reformatting of table to ensure mapping each EC as less as possible \n",
    "    #         (ideally only once)\n",
    "    # @TODO : how to carry the uniprot information along\n",
    "    #       -> directly access reaction \n",
    "    mapped_reacs = map_ec_to_reac(mapped_reacs[['ec-code','ncbiprotein']])\n",
    "    \n",
    "    return mapped_reacs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfasta = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435_proteins_genome.fasta'\n",
    "spdb = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/swissprot.dmnd'\n",
    "# gf2_missing_genes\n",
    "spmap = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/uniprot_table.tsv'\n",
    "kwargs = {'outdir':'/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435',\n",
    "          'sens':'more-sensitive',\n",
    "          'cov':90.0,\n",
    "          't':4,\n",
    "          'pid':90.0}\n",
    "\n",
    "mapped_res = get_missing_reacs(missing_genes=gf2_missing_genes,\n",
    "                               fasta=tfasta, \n",
    "                               dmnd_db=spdb,\n",
    "                               swissprot_map=spmap,\n",
    "                               **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ec-code</th>\n",
       "      <th>UniProt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_011274359.1</td>\n",
       "      <td>SH0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_011274360.1</td>\n",
       "      <td>SH0002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_172458781.1</td>\n",
       "      <td>SH0003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_011274362.1</td>\n",
       "      <td>SH0004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_011274363.1</td>\n",
       "      <td>SH0005</td>\n",
       "      <td>5.6.2.2</td>\n",
       "      <td>[Q5HK03, Q8CQK4, Q6GKU0, Q2FKQ1, Q6GD85, P0A0K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1698</th>\n",
       "      <td>WP_145424740.1</td>\n",
       "      <td>pSHaeC05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1699</th>\n",
       "      <td>WP_011276926.1</td>\n",
       "      <td>pSHaeC06</td>\n",
       "      <td>2.3.2.22</td>\n",
       "      <td>[Q4L2X9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700</th>\n",
       "      <td>WP_011276927.1</td>\n",
       "      <td>pSHaeC07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1701</th>\n",
       "      <td>WP_011276928.1</td>\n",
       "      <td>pSHaeC08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>WP_011276930.1</td>\n",
       "      <td>pSHaeC10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1703 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ncbiprotein locus_tag   ec-code  \\\n",
       "0     WP_011274359.1    SH0001       NaN   \n",
       "1     WP_011274360.1    SH0002       NaN   \n",
       "2     WP_172458781.1    SH0003       NaN   \n",
       "3     WP_011274362.1    SH0004       NaN   \n",
       "4     WP_011274363.1    SH0005   5.6.2.2   \n",
       "...              ...       ...       ...   \n",
       "1698  WP_145424740.1  pSHaeC05       NaN   \n",
       "1699  WP_011276926.1  pSHaeC06  2.3.2.22   \n",
       "1700  WP_011276927.1  pSHaeC07       NaN   \n",
       "1701  WP_011276928.1  pSHaeC08       NaN   \n",
       "1702  WP_011276930.1  pSHaeC10       NaN   \n",
       "\n",
       "                                                UniProt  \n",
       "0                                                   NaN  \n",
       "1                                                   NaN  \n",
       "2                                                   NaN  \n",
       "3                                                   NaN  \n",
       "4     [Q5HK03, Q8CQK4, Q6GKU0, Q2FKQ1, Q6GD85, P0A0K...  \n",
       "...                                                 ...  \n",
       "1698                                                NaN  \n",
       "1699                                           [Q4L2X9]  \n",
       "1700                                                NaN  \n",
       "1701                                                NaN  \n",
       "1702                                                NaN  \n",
       "\n",
       "[1703 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapped_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @TODO\n",
    "# SwissProt Download\n",
    "# (mappings + sequences can be downloaded from DB)\n",
    "# two tables: \n",
    "#   1.: TSV with UniprotID, BRENDA and EC -7.7MB (26.07.2024)\n",
    "#   2.: FASTA with sequences ~280MB (26.07.2024)\n",
    "# more infos:\n",
    "#   - API generation possible for automatic download\n",
    "#   - maybe write a small script to download stuff and shorten to actual EC/BRENDA mappings\n",
    "\n",
    "# current_release or complete? differences?\n",
    "swissprot_api = 'https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz'\n",
    "swissprot_mapping_api = 'https://rest.uniprot.org/uniprotkb/stream?compressed=true&fields=accession%2Cxref_brenda%2Cec%2Csequence&format=tsv&query=%28*%29+AND+%28reviewed%3Atrue%29'\n",
    "\n",
    "def download_swissprot_for_gapfill():\n",
    "    pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sprg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
