{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGGapFiller Example\n",
    "\n",
    "### Initialisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/doebel/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.classes.gapfill import KEGGapFiller\n",
    "from refinegems.utility.io import load_model\n",
    "\n",
    "# Get model file loaded with COBRA (for missing reactions) & libSBML\n",
    "modelpath = './test_files/JCSC1435.xml' # A link to this model will be added as soon as it is published\n",
    "model = load_model(modelpath,'libsbml')\n",
    "cmodel = load_model(modelpath,'cobra')\n",
    "\n",
    "# Initialise GapFiller subclass to be used with required parameters\n",
    "gfk = KEGGapFiller('sha')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1987/1987 [42:40<00:00,  1.29s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orgid:locus</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>kegg.orthology</th>\n",
       "      <th>ec-code</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sha:pSHaeA01</td>\n",
       "      <td>pSHaeA01</td>\n",
       "      <td>[K11210]</td>\n",
       "      <td>[2.5.1.-]</td>\n",
       "      <td>BAE05988</td>\n",
       "      <td>[Q4L2Y9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sha:pSHaeA02</td>\n",
       "      <td>pSHaeA02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE05989</td>\n",
       "      <td>[Q4L2Y8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sha:pSHaeA03</td>\n",
       "      <td>pSHaeA03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE05990</td>\n",
       "      <td>[Q4L2Y7]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sha:pSHaeB01</td>\n",
       "      <td>pSHaeB01</td>\n",
       "      <td>[K00561]</td>\n",
       "      <td>[2.1.1.184]</td>\n",
       "      <td>BAE05991</td>\n",
       "      <td>[Q4L2Y6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sha:pSHaeB02</td>\n",
       "      <td>pSHaeB02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE05992</td>\n",
       "      <td>[Q4L2Y5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>sha:SH2674</td>\n",
       "      <td>SH2674</td>\n",
       "      <td>[K03501]</td>\n",
       "      <td>[2.1.1.170]</td>\n",
       "      <td>BAE05983</td>\n",
       "      <td>[Q4L2Z4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>sha:SH2675</td>\n",
       "      <td>SH2675</td>\n",
       "      <td>[K03495]</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE05984</td>\n",
       "      <td>[Q4L2Z3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>sha:SH2676</td>\n",
       "      <td>SH2676</td>\n",
       "      <td>[K03650]</td>\n",
       "      <td>[3.6.-.-]</td>\n",
       "      <td>BAE05985</td>\n",
       "      <td>[Q4L2Z2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>sha:SH2677</td>\n",
       "      <td>SH2677</td>\n",
       "      <td>[K03536]</td>\n",
       "      <td>[3.1.26.5]</td>\n",
       "      <td>BAE05986</td>\n",
       "      <td>[Q4L2Z1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>sha:SH2678</td>\n",
       "      <td>SH2678</td>\n",
       "      <td>[K02914]</td>\n",
       "      <td>None</td>\n",
       "      <td>BAE05987</td>\n",
       "      <td>[Q4L2Z0]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1987 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       orgid:locus locus_tag kegg.orthology      ec-code ncbiprotein   uniprot\n",
       "0     sha:pSHaeA01  pSHaeA01       [K11210]    [2.5.1.-]    BAE05988  [Q4L2Y9]\n",
       "1     sha:pSHaeA02  pSHaeA02            NaN         None    BAE05989  [Q4L2Y8]\n",
       "2     sha:pSHaeA03  pSHaeA03            NaN         None    BAE05990  [Q4L2Y7]\n",
       "3     sha:pSHaeB01  pSHaeB01       [K00561]  [2.1.1.184]    BAE05991  [Q4L2Y6]\n",
       "4     sha:pSHaeB02  pSHaeB02            NaN         None    BAE05992  [Q4L2Y5]\n",
       "...            ...       ...            ...          ...         ...       ...\n",
       "1982    sha:SH2674    SH2674       [K03501]  [2.1.1.170]    BAE05983  [Q4L2Z4]\n",
       "1983    sha:SH2675    SH2675       [K03495]         None    BAE05984  [Q4L2Z3]\n",
       "1984    sha:SH2676    SH2676       [K03650]    [3.6.-.-]    BAE05985  [Q4L2Z2]\n",
       "1985    sha:SH2677    SH2677       [K03536]   [3.1.26.5]    BAE05986  [Q4L2Z1]\n",
       "1986    sha:SH2678    SH2678       [K02914]         None    BAE05987  [Q4L2Z0]\n",
       "\n",
       "[1987 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find missing genes\n",
    "gfk.find_missing_genes(model)\n",
    "\n",
    "# Get/show missing genes\n",
    "gfk.missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11/11 [00:43<00:00,  3.93s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec-code</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>id</th>\n",
       "      <th>equation</th>\n",
       "      <th>reference</th>\n",
       "      <th>is_transport</th>\n",
       "      <th>via</th>\n",
       "      <th>add_to_GPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.3.1.313</td>\n",
       "      <td>[BAE04147]</td>\n",
       "      <td>R13186</td>\n",
       "      <td>None</td>\n",
       "      <td>{'brenda': ['2.3.1.313']}</td>\n",
       "      <td>None</td>\n",
       "      <td>KEGG</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.4.99.28</td>\n",
       "      <td>[BAE04395, BAE04771]</td>\n",
       "      <td>R06178</td>\n",
       "      <td>None</td>\n",
       "      <td>{'kegg.pathway': ['ec00550'], 'brenda': ['2.4....</td>\n",
       "      <td>None</td>\n",
       "      <td>KEGG</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.4.99.28</td>\n",
       "      <td>[BAE04395, BAE04771]</td>\n",
       "      <td>R06179</td>\n",
       "      <td>None</td>\n",
       "      <td>{'kegg.pathway': ['ec00550'], 'brenda': ['2.4....</td>\n",
       "      <td>None</td>\n",
       "      <td>KEGG</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.7.10.3</td>\n",
       "      <td>[BAE03698]</td>\n",
       "      <td>R02584</td>\n",
       "      <td>None</td>\n",
       "      <td>{'brenda': ['2.7.10.3']}</td>\n",
       "      <td>None</td>\n",
       "      <td>KEGG</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.7.2.18</td>\n",
       "      <td>[BAE04998]</td>\n",
       "      <td>R12793</td>\n",
       "      <td>None</td>\n",
       "      <td>{'brenda': ['2.7.2.18']}</td>\n",
       "      <td>None</td>\n",
       "      <td>KEGG</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3616</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[BAE04398, BAE05568]</td>\n",
       "      <td>MNXR115570</td>\n",
       "      <td>1 MNXM1@MNXD1 + 1 MNXM21999@MNXD2 + 1 MNXM4033...</td>\n",
       "      <td>seedR:rxn18688</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3617</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[BAE04398, BAE05568]</td>\n",
       "      <td>MNXR121346</td>\n",
       "      <td>1 MNXM12982@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM4033...</td>\n",
       "      <td>seedR:rxn41181</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3618</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[BAE04398, BAE05568]</td>\n",
       "      <td>MNXR121347</td>\n",
       "      <td>1 MNXM163615@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>seedR:rxn43450</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3619</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[BAE04398, BAE05568]</td>\n",
       "      <td>MNXR123981</td>\n",
       "      <td>1 MNXM163613@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>seedR:rxn42597</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3620</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[BAE04398, BAE05568]</td>\n",
       "      <td>MNXR191804</td>\n",
       "      <td>1 MNXM164493@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>metacycR:TRANS-RXN-321</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3614 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ec-code           ncbiprotein          id  \\\n",
       "0     2.3.1.313            [BAE04147]      R13186   \n",
       "1     2.4.99.28  [BAE04395, BAE04771]      R06178   \n",
       "2     2.4.99.28  [BAE04395, BAE04771]      R06179   \n",
       "3      2.7.10.3            [BAE03698]      R02584   \n",
       "4      2.7.2.18            [BAE04998]      R12793   \n",
       "...         ...                   ...         ...   \n",
       "3616    7.5.2.4  [BAE04398, BAE05568]  MNXR115570   \n",
       "3617    7.5.2.4  [BAE04398, BAE05568]  MNXR121346   \n",
       "3618    7.5.2.4  [BAE04398, BAE05568]  MNXR121347   \n",
       "3619    7.5.2.4  [BAE04398, BAE05568]  MNXR123981   \n",
       "3620    7.5.2.4  [BAE04398, BAE05568]  MNXR191804   \n",
       "\n",
       "                                               equation  \\\n",
       "0                                                  None   \n",
       "1                                                  None   \n",
       "2                                                  None   \n",
       "3                                                  None   \n",
       "4                                                  None   \n",
       "...                                                 ...   \n",
       "3616  1 MNXM1@MNXD1 + 1 MNXM21999@MNXD2 + 1 MNXM4033...   \n",
       "3617  1 MNXM12982@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM4033...   \n",
       "3618  1 MNXM163615@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "3619  1 MNXM163613@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "3620  1 MNXM164493@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "\n",
       "                                              reference is_transport  \\\n",
       "0                             {'brenda': ['2.3.1.313']}         None   \n",
       "1     {'kegg.pathway': ['ec00550'], 'brenda': ['2.4....         None   \n",
       "2     {'kegg.pathway': ['ec00550'], 'brenda': ['2.4....         None   \n",
       "3                              {'brenda': ['2.7.10.3']}         None   \n",
       "4                              {'brenda': ['2.7.2.18']}         None   \n",
       "...                                                 ...          ...   \n",
       "3616                                     seedR:rxn18688            T   \n",
       "3617                                     seedR:rxn41181            T   \n",
       "3618                                     seedR:rxn43450            T   \n",
       "3619                                     seedR:rxn42597            T   \n",
       "3620                             metacycR:TRANS-RXN-321            T   \n",
       "\n",
       "           via add_to_GPR  \n",
       "0         KEGG       None  \n",
       "1         KEGG       None  \n",
       "2         KEGG       None  \n",
       "3         KEGG       None  \n",
       "4         KEGG       None  \n",
       "...        ...        ...  \n",
       "3616  MetaNetX       None  \n",
       "3617  MetaNetX       None  \n",
       "3618  MetaNetX       None  \n",
       "3619  MetaNetX       None  \n",
       "3620  MetaNetX       None  \n",
       "\n",
       "[3614 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find missing reactions\n",
    "gfk.find_missing_reactions(cmodel)\n",
    "\n",
    "# Get/show missing reactions\n",
    "gfk.missing_reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfk.missing_reactions.to_csv('./test_files/kegg_missing_reactions.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fill model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding genes to model: 100%|██████████| 18/18 [00:00<00:00, 3138.54it/s]\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'getPlugin'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Fill model\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mgfk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfill_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/refinegems/classes/gapfill.py:1021\u001b[0m, in \u001b[0;36mGapFiller.fill_model\u001b[0;34m(self, model, **kwargs)\u001b[0m\n\u001b[1;32m   1018\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39madd_genes_from_table(model, genes_with_reacs_in_model)\n\u001b[1;32m   1020\u001b[0m \u001b[38;5;66;03m# extend gene production rules \u001b[39;00m\n\u001b[0;32m-> 1021\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd_gene_reac_associations_from_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43mreacs_in_model\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1023\u001b[0m \u001b[38;5;66;03m# what remains:\u001b[39;00m\n\u001b[1;32m   1024\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmissing_reactions \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmissing_reactions[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmissing_reactions[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124madd_to_GPR\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39misnull()]\n",
      "File \u001b[0;32m~/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/refinegems/classes/gapfill.py:810\u001b[0m, in \u001b[0;36mGapFiller.add_gene_reac_associations_from_table\u001b[0;34m(self, model, reac_table)\u001b[0m\n\u001b[1;32m    808\u001b[0m current_reacid \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mR_\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m+\u001b[39mreacid\n\u001b[1;32m    809\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m geneid \u001b[38;5;129;01min\u001b[39;00m model_gene_ids:\n\u001b[0;32m--> 810\u001b[0m     \u001b[43mcreate_gpr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetReaction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcurrent_reacid\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43mgeneid\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    811\u001b[0m \u001b[38;5;66;03m# else, print warning\u001b[39;00m\n\u001b[1;32m    812\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    813\u001b[0m     mes \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCannot find \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgeneid\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m in model. Should be added to \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcurrent_reacid\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/refinegems/utility/entities.py:1779\u001b[0m, in \u001b[0;36mcreate_gpr\u001b[0;34m(reaction, gene)\u001b[0m\n\u001b[1;32m   1777\u001b[0m old_association_str \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1778\u001b[0m old_association_fbc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1779\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mreaction\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetPlugin\u001b[49m(\u001b[38;5;241m0\u001b[39m)\u001b[38;5;241m.\u001b[39mgetGeneProductAssociation():\n\u001b[1;32m   1780\u001b[0m     old_association \u001b[38;5;241m=\u001b[39m reaction\u001b[38;5;241m.\u001b[39mgetPlugin(\u001b[38;5;241m0\u001b[39m)\u001b[38;5;241m.\u001b[39mgetGeneProductAssociation()\u001b[38;5;241m.\u001b[39mgetListOfAllElements()\n\u001b[1;32m   1781\u001b[0m     \u001b[38;5;66;03m# case 1: only a single association\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'getPlugin'"
     ]
    }
   ],
   "source": [
    "# Fill model\n",
    "model = gfk.fill_model(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes': {'missing (before)': 1987,\n",
       "  'duplicates': 78,\n",
       "  'added': 18,\n",
       "  'missing (after)': 0},\n",
       " 'reactions': {'added (total)': 0, 'failed to build': 0},\n",
       " 'metabolites': {}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get raw statistics\n",
    "gfk._statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GeneGapFiller Example\n",
    "\n",
    "### Initialisation\n",
    "#### Get SwissProt DIAMOND database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/doebel/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.utility.set_up import download_url\n",
    "\n",
    "# Get SwissProt\n",
    "# download_url('SwissProt gapfill', directory='./test_files/test_gapfill')\n",
    "\n",
    "# On the command line:\n",
    "# diamond makedb --in ./dev/test_files/test_gapfill/SwissProt.fasta -d ./dev/test_files/test_gapfill/swissprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initialise GeneGapFiller"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/doebel/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.classes.gapfill import GeneGapFiller\n",
    "from refinegems.utility.io import load_model\n",
    "\n",
    "# Get model file loaded with COBRA (for missing reactions) & libSBML\n",
    "modelpath = './test_files/IMITSC147.xml' # A link to this model will be added as soon as it is published\n",
    "model = load_model(modelpath,'libsbml')\n",
    "cmodel = load_model(modelpath,'cobra')\n",
    "\n",
    "# Get files required for the GeneGapFiller\n",
    "gffpath = './test_files/IMITSC147_genome.gff'\n",
    "\n",
    "tfasta = './test_files/IMITSC147_proteins_genome.fasta'\n",
    "spdb = './test_files/test_gapfill/swissprot.dmnd'\n",
    "spmap = './test_files/test_gapfill/SwissProt_mapping.tsv'\n",
    "kwargs = {'outdir':'./test_files/test_gapfill/IMITSC147',\n",
    "          'sens':'more-sensitive',\n",
    "          'cov':95.0,\n",
    "          't':4,\n",
    "          'pid':90.0}\n",
    "\n",
    "# Initialise GapFiller subclass to be used with required parameters\n",
    "gfg = GeneGapFiller()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>ec-code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMITSC147_000001</td>\n",
       "      <td>extdb:IMITSC147_000001</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IMITSC147_000002</td>\n",
       "      <td>extdb:IMITSC147_000002</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IMITSC147_000003</td>\n",
       "      <td>extdb:IMITSC147_000003</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IMITSC147_000004</td>\n",
       "      <td>extdb:IMITSC147_000004</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IMITSC147_000005</td>\n",
       "      <td>extdb:IMITSC147_000005</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>IMITSC147_002523</td>\n",
       "      <td>extdb:IMITSC147_002523</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>IMITSC147_002524</td>\n",
       "      <td>extdb:IMITSC147_002524</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2441</th>\n",
       "      <td>IMITSC147_002525</td>\n",
       "      <td>extdb:IMITSC147_002525</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2442</th>\n",
       "      <td>IMITSC147_002526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2443</th>\n",
       "      <td>IMITSC147_002527</td>\n",
       "      <td>extdb:IMITSC147_002527</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1679 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             locus_tag             ncbiprotein ec-code\n",
       "0     IMITSC147_000001  extdb:IMITSC147_000001    None\n",
       "1     IMITSC147_000002  extdb:IMITSC147_000002    None\n",
       "2     IMITSC147_000003  extdb:IMITSC147_000003    None\n",
       "3     IMITSC147_000004  extdb:IMITSC147_000004    None\n",
       "4     IMITSC147_000005  extdb:IMITSC147_000005    None\n",
       "...                ...                     ...     ...\n",
       "2439  IMITSC147_002523  extdb:IMITSC147_002523    None\n",
       "2440  IMITSC147_002524  extdb:IMITSC147_002524    None\n",
       "2441  IMITSC147_002525  extdb:IMITSC147_002525    None\n",
       "2442  IMITSC147_002526                     NaN    None\n",
       "2443  IMITSC147_002527  extdb:IMITSC147_002527    None\n",
       "\n",
       "[1679 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find missing genes\n",
    "gfg.find_missing_genes(gffpath,model)\n",
    "\n",
    "# Get/show missing genes\n",
    "gfg.missing_genes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4/4 [00:12<00:00,  3.06s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec-code</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>id</th>\n",
       "      <th>equation</th>\n",
       "      <th>reference</th>\n",
       "      <th>is_transport</th>\n",
       "      <th>via</th>\n",
       "      <th>add_to_GPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.-.-.-</td>\n",
       "      <td>[extdb:IMITSC147_000502]</td>\n",
       "      <td>MNXR108769</td>\n",
       "      <td>1 MNXM1@MNXD1 + 1 MNXM736332@MNXD1 + 1 MNXM738...</td>\n",
       "      <td>rheaR:34567</td>\n",
       "      <td>None</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.-.-.-</td>\n",
       "      <td>[extdb:IMITSC147_000502]</td>\n",
       "      <td>MNXR109748</td>\n",
       "      <td>1 MNXM10@MNXD1 + 1 MNXM4978@MNXD1 = 3 MNXM1@MN...</td>\n",
       "      <td>keggR:R05668</td>\n",
       "      <td>None</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.-.-.-</td>\n",
       "      <td>[extdb:IMITSC147_000502]</td>\n",
       "      <td>MNXR111597</td>\n",
       "      <td>1 MNXM1094296@MNXD1 + 2 MNXM1@MNXD1 + 2 MNXM73...</td>\n",
       "      <td>keggR:R08014</td>\n",
       "      <td>None</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.-.-.-</td>\n",
       "      <td>[extdb:IMITSC147_000502]</td>\n",
       "      <td>MNXR111600</td>\n",
       "      <td>1 MNXM1094296@MNXD1 + 2 MNXM10@MNXD1 + 2 MNXM1...</td>\n",
       "      <td>keggR:R08017</td>\n",
       "      <td>None</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.-.-.-</td>\n",
       "      <td>[extdb:IMITSC147_000502]</td>\n",
       "      <td>MNXR111623</td>\n",
       "      <td>1 MNXM1094296@MNXD1 + 2 MNXM10@MNXD1 + 2 MNXM1...</td>\n",
       "      <td>keggR:R08042</td>\n",
       "      <td>None</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3482</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[extdb:IMITSC147_002172]</td>\n",
       "      <td>MNXR121346</td>\n",
       "      <td>1 MNXM12982@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM4033...</td>\n",
       "      <td>seedR:rxn41181</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3483</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[extdb:IMITSC147_002172]</td>\n",
       "      <td>MNXR121347</td>\n",
       "      <td>1 MNXM163615@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>seedR:rxn43450</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3484</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[extdb:IMITSC147_002172]</td>\n",
       "      <td>MNXR123981</td>\n",
       "      <td>1 MNXM163613@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>seedR:rxn42597</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3485</th>\n",
       "      <td>7.5.2.4</td>\n",
       "      <td>[extdb:IMITSC147_002172]</td>\n",
       "      <td>MNXR191804</td>\n",
       "      <td>1 MNXM164493@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...</td>\n",
       "      <td>metacycR:TRANS-RXN-321</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3486</th>\n",
       "      <td>7.6.2.-</td>\n",
       "      <td>[extdb:IMITSC147_000615]</td>\n",
       "      <td>MNXR183687</td>\n",
       "      <td>1 MNXM1101005@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM40...</td>\n",
       "      <td>seedR:rxn40820</td>\n",
       "      <td>T</td>\n",
       "      <td>MetaNetX</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3483 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ec-code               ncbiprotein          id  \\\n",
       "4     1.-.-.-  [extdb:IMITSC147_000502]  MNXR108769   \n",
       "5     1.-.-.-  [extdb:IMITSC147_000502]  MNXR109748   \n",
       "6     1.-.-.-  [extdb:IMITSC147_000502]  MNXR111597   \n",
       "7     1.-.-.-  [extdb:IMITSC147_000502]  MNXR111600   \n",
       "8     1.-.-.-  [extdb:IMITSC147_000502]  MNXR111623   \n",
       "...       ...                       ...         ...   \n",
       "3482  7.5.2.4  [extdb:IMITSC147_002172]  MNXR121346   \n",
       "3483  7.5.2.4  [extdb:IMITSC147_002172]  MNXR121347   \n",
       "3484  7.5.2.4  [extdb:IMITSC147_002172]  MNXR123981   \n",
       "3485  7.5.2.4  [extdb:IMITSC147_002172]  MNXR191804   \n",
       "3486  7.6.2.-  [extdb:IMITSC147_000615]  MNXR183687   \n",
       "\n",
       "                                               equation  \\\n",
       "4     1 MNXM1@MNXD1 + 1 MNXM736332@MNXD1 + 1 MNXM738...   \n",
       "5     1 MNXM10@MNXD1 + 1 MNXM4978@MNXD1 = 3 MNXM1@MN...   \n",
       "6     1 MNXM1094296@MNXD1 + 2 MNXM1@MNXD1 + 2 MNXM73...   \n",
       "7     1 MNXM1094296@MNXD1 + 2 MNXM10@MNXD1 + 2 MNXM1...   \n",
       "8     1 MNXM1094296@MNXD1 + 2 MNXM10@MNXD1 + 2 MNXM1...   \n",
       "...                                                 ...   \n",
       "3482  1 MNXM12982@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM4033...   \n",
       "3483  1 MNXM163615@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "3484  1 MNXM163613@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "3485  1 MNXM164493@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM403...   \n",
       "3486  1 MNXM1101005@MNXD1 + 1 MNXM1@MNXD2 + 1 MNXM40...   \n",
       "\n",
       "                   reference is_transport       via add_to_GPR  \n",
       "4                rheaR:34567         None  MetaNetX       None  \n",
       "5               keggR:R05668         None  MetaNetX       None  \n",
       "6               keggR:R08014         None  MetaNetX       None  \n",
       "7               keggR:R08017         None  MetaNetX       None  \n",
       "8               keggR:R08042         None  MetaNetX       None  \n",
       "...                      ...          ...       ...        ...  \n",
       "3482          seedR:rxn41181            T  MetaNetX       None  \n",
       "3483          seedR:rxn43450            T  MetaNetX       None  \n",
       "3484          seedR:rxn42597            T  MetaNetX       None  \n",
       "3485  metacycR:TRANS-RXN-321            T  MetaNetX       None  \n",
       "3486          seedR:rxn40820            T  MetaNetX       None  \n",
       "\n",
       "[3483 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%autoreload 2\n",
    "# Find missing reactions\n",
    "gfg.find_missing_reactions(model=cmodel,\n",
    "                                   fasta=tfasta, \n",
    "                                   dmnd_db=spdb,\n",
    "                                   swissprot_map=spmap,\n",
    "                                   **kwargs)\n",
    "\n",
    "# Get/show missing reactions\n",
    "gfg.missing_reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfg.missing_reactions.to_csv('./test_files/gene_missing_reactions.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fill model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gfg' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Fill model\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mgfg\u001b[49m\u001b[38;5;241m.\u001b[39mfill_model(model,formula_check\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mexistence\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gfg' is not defined"
     ]
    }
   ],
   "source": [
    "# Fill model\n",
    "model = gfg.fill_model(model,formula_check='existence')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes': {'missing (before)': 1679,\n",
       "  'duplicates': 9,\n",
       "  'added': 76,\n",
       "  'missing (after)': 0,\n",
       "  'no locus tag': 0},\n",
       " 'reactions': {'added (total)': 0,\n",
       "  'failed to build': 0,\n",
       "  'no NCBI, no EC': 4,\n",
       "  'NCBI, no EC': 1369},\n",
       " 'metabolites': {}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get raw statistics\n",
    "gfg._statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BioCycGapFiller example\n",
    "\n",
    "### Initialisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/doebel/miniconda3/envs/refineGEMs/lib/python3.10/site-packages/pydantic/_internal/_config.py:322: UserWarning: Valid config keys have changed in V2:\n",
      "* 'underscore_attrs_are_private' has been removed\n",
      "  warnings.warn(message, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from refinegems.classes.gapfill import BioCycGapFiller\n",
    "from refinegems.utility.io import load_model\n",
    "\n",
    "# Get model file loaded with COBRA (for missing reactions) & libSBML\n",
    "modelpath = './test_files/ATCC29970.xml' # A link to this model will be added as soon as it is published\n",
    "model = load_model(modelpath,'libsbml')\n",
    "cmodel = load_model(modelpath,'cobra')\n",
    "\n",
    "# Get files required for the BioCycGapFiller\n",
    "gffpath = './test_files/ATCC29970_RefSeq.gff'\n",
    "biocyc_gene_tbl_path = './test_files/ATCC29970_Accession-22Reactions.txt'\n",
    "biocyc_reacs_tbl_path = './test_files/ATCC29970_biocyc_rxns.txt'\n",
    "fasta = './test_files/ATCC29970_proteins_genome.fasta'\n",
    "\n",
    "# Initialise GapFiller subclass to be used with required parameters\n",
    "gfbc = BioCycGapFiller(biocyc_gene_tbl_path, biocyc_reacs_tbl_path, gffpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>ncbiprotein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EQ029_01155</td>\n",
       "      <td>OLEATE-HYDRATASE-RXN</td>\n",
       "      <td>oleate hydratase</td>\n",
       "      <td>WP_053030667.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EQ029_06240</td>\n",
       "      <td>TRANS-RXN1WO9-20</td>\n",
       "      <td>nucleotide exchange factor GrpE</td>\n",
       "      <td>WP_011275632.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EQ029_07635</td>\n",
       "      <td>PHOSPHAGLYPSYN-RXN</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>WP_011275916.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EQ029_03370</td>\n",
       "      <td>RXN-8348</td>\n",
       "      <td>molybdopterin molybdotransferase MoeA</td>\n",
       "      <td>WP_011275102.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EQ029_03775</td>\n",
       "      <td>HEME-OXYGENASE-DECYCLIZING-RXN // RXN-18341 //...</td>\n",
       "      <td>heme oxygenase</td>\n",
       "      <td>WP_011275179.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>EQ029_01095</td>\n",
       "      <td>PYROGLUTAMYL-PEPTIDASE-I-RXN</td>\n",
       "      <td>pyroglutamyl-peptidase I</td>\n",
       "      <td>WP_011274622.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>EQ029_00835</td>\n",
       "      <td>TEICHOICSYN3-RXN</td>\n",
       "      <td>CDP-glycerol glycerophosphotransferase family ...</td>\n",
       "      <td>WP_057504685.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>EQ029_06820</td>\n",
       "      <td>TRNA-CYTIDYLYLTRANSFERASE-RXN</td>\n",
       "      <td>CCA tRNA nucleotidyltransferase</td>\n",
       "      <td>WP_037558740.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>EQ029_08625</td>\n",
       "      <td>RXN-18399 // RXN-18428 // RXN-18430</td>\n",
       "      <td>heme A synthase</td>\n",
       "      <td>WP_011276113.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>EQ029_04765</td>\n",
       "      <td>TRANS-RXN1WO9-12</td>\n",
       "      <td>aromatic acid exporter family protein</td>\n",
       "      <td>WP_011275368.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>235 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       locus_tag                                                 id  \\\n",
       "0    EQ029_01155                               OLEATE-HYDRATASE-RXN   \n",
       "1    EQ029_06240                                   TRANS-RXN1WO9-20   \n",
       "2    EQ029_07635                                 PHOSPHAGLYPSYN-RXN   \n",
       "3    EQ029_03370                                           RXN-8348   \n",
       "4    EQ029_03775  HEME-OXYGENASE-DECYCLIZING-RXN // RXN-18341 //...   \n",
       "..           ...                                                ...   \n",
       "230  EQ029_01095                       PYROGLUTAMYL-PEPTIDASE-I-RXN   \n",
       "231  EQ029_00835                                   TEICHOICSYN3-RXN   \n",
       "232  EQ029_06820                      TRNA-CYTIDYLYLTRANSFERASE-RXN   \n",
       "233  EQ029_08625                RXN-18399 // RXN-18428 // RXN-18430   \n",
       "234  EQ029_04765                                   TRANS-RXN1WO9-12   \n",
       "\n",
       "                                                  name     ncbiprotein  \n",
       "0                                     oleate hydratase  WP_053030667.1  \n",
       "1                      nucleotide exchange factor GrpE  WP_011275632.1  \n",
       "2    CDP-diacylglycerol--glycerol-3-phosphate 3-pho...  WP_011275916.1  \n",
       "3                molybdopterin molybdotransferase MoeA  WP_011275102.1  \n",
       "4                                       heme oxygenase  WP_011275179.1  \n",
       "..                                                 ...             ...  \n",
       "230                           pyroglutamyl-peptidase I  WP_011274622.1  \n",
       "231  CDP-glycerol glycerophosphotransferase family ...  WP_057504685.1  \n",
       "232                    CCA tRNA nucleotidyltransferase  WP_037558740.1  \n",
       "233                                    heme A synthase  WP_011276113.1  \n",
       "234              aromatic acid exporter family protein  WP_011275368.1  \n",
       "\n",
       "[235 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find missing genes\n",
    "gfbc.find_missing_genes(model)\n",
    "\n",
    "# Get/show missing genes\n",
    "gfbc.missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>add_to_GPR</th>\n",
       "      <th>ec-code</th>\n",
       "      <th>equation</th>\n",
       "      <th>id</th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>reference</th>\n",
       "      <th>via</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[3HAD80]</td>\n",
       "      <td>[4.2.1.59]</td>\n",
       "      <td>(3R)-3-hydroxyoctanoyl-[acp]  -&gt;  (2E)-oct-2-e...</td>\n",
       "      <td>4.2.1.59-RXN</td>\n",
       "      <td>[WP_011275247.1]</td>\n",
       "      <td>None</td>\n",
       "      <td>BioCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[AMMQLT8]</td>\n",
       "      <td>[2.1.1.163]</td>\n",
       "      <td>S-adenosyl-L-methionine + demethylmenaquinol-8...</td>\n",
       "      <td>ADOMET-DMK-METHYLTRANSFER-RXN</td>\n",
       "      <td>[WP_011275738.1]</td>\n",
       "      <td>None</td>\n",
       "      <td>BioCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[NTP1, ATPM]</td>\n",
       "      <td>[3.6.1.3, 3.6.1.15]</td>\n",
       "      <td>ATP + H2O  -&gt;  ADP + phosphate + H+</td>\n",
       "      <td>ATPASE-RXN</td>\n",
       "      <td>[WP_016931380.1]</td>\n",
       "      <td>None</td>\n",
       "      <td>BioCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[LDH_D]</td>\n",
       "      <td>[1.1.1.28]</td>\n",
       "      <td>(R)-lactate + NAD+  &lt;-  pyruvate + NADH + H+</td>\n",
       "      <td>DLACTDEHYDROGNAD-RXN</td>\n",
       "      <td>[WP_011274818.1]</td>\n",
       "      <td>None</td>\n",
       "      <td>BioCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[DHNAOT4]</td>\n",
       "      <td>[2.5.1.74]</td>\n",
       "      <td>all-trans-octaprenyl diphosphate + 1,4-dihydro...</td>\n",
       "      <td>DMK-RXN</td>\n",
       "      <td>[WP_016931372.1]</td>\n",
       "      <td>None</td>\n",
       "      <td>BioCyc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>None</td>\n",
       "      <td>[6.1.1.2]</td>\n",
       "      <td>1 MNXM11@MNXD1 + 1 MNXM167418@MNXD1 + 1 MNXM72...</td>\n",
       "      <td>MNXR124075</td>\n",
       "      <td>[WP_011276231.1]</td>\n",
       "      <td>{'metacyc.reaction': 'TRYPTOPHAN--TRNA-LIGASE-...</td>\n",
       "      <td>MetaNetX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>None</td>\n",
       "      <td>[6.1.1.1]</td>\n",
       "      <td>1 MNXM11@MNXD1 + 1 MNXM16655@MNXD1 + 1 MNXM728...</td>\n",
       "      <td>MNXR124080</td>\n",
       "      <td>[WP_011275493.1]</td>\n",
       "      <td>{'metacyc.reaction': 'TYROSINE--TRNA-LIGASE-RX...</td>\n",
       "      <td>MetaNetX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>None</td>\n",
       "      <td>[6.3.2.10]</td>\n",
       "      <td>1 MNXM1104679@MNXD1 + 1 MNXM1446@MNXD1 + 1 MNX...</td>\n",
       "      <td>MNXR146695</td>\n",
       "      <td>[WP_011275264.1]</td>\n",
       "      <td>{'metacyc.reaction': 'UDP-NACMURALGLDAPAALIG-R...</td>\n",
       "      <td>MetaNetX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>None</td>\n",
       "      <td>[5.1.3.14]</td>\n",
       "      <td>1 MNXM1101258@MNXD1 = 1 MNXM1104529@MNXD1</td>\n",
       "      <td>MNXR151539</td>\n",
       "      <td>[WP_033079611.1]</td>\n",
       "      <td>{'metacyc.reaction': 'UDPGLCNACEPIM-RXN', 'ori...</td>\n",
       "      <td>MetaNetX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>None</td>\n",
       "      <td>[6.1.1.9]</td>\n",
       "      <td>1 MNXM11@MNXD1 + 1 MNXM167419@MNXD1 + 1 MNXM72...</td>\n",
       "      <td>MNXR124096</td>\n",
       "      <td>[WP_011275561.1]</td>\n",
       "      <td>{'metacyc.reaction': 'VALINE--TRNA-LIGASE-RXN'...</td>\n",
       "      <td>MetaNetX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>238 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       add_to_GPR              ec-code  \\\n",
       "0        [3HAD80]           [4.2.1.59]   \n",
       "1       [AMMQLT8]          [2.1.1.163]   \n",
       "2    [NTP1, ATPM]  [3.6.1.3, 3.6.1.15]   \n",
       "3         [LDH_D]           [1.1.1.28]   \n",
       "4       [DHNAOT4]           [2.5.1.74]   \n",
       "..            ...                  ...   \n",
       "233          None            [6.1.1.2]   \n",
       "234          None            [6.1.1.1]   \n",
       "235          None           [6.3.2.10]   \n",
       "236          None           [5.1.3.14]   \n",
       "237          None            [6.1.1.9]   \n",
       "\n",
       "                                              equation  \\\n",
       "0    (3R)-3-hydroxyoctanoyl-[acp]  ->  (2E)-oct-2-e...   \n",
       "1    S-adenosyl-L-methionine + demethylmenaquinol-8...   \n",
       "2                  ATP + H2O  ->  ADP + phosphate + H+   \n",
       "3         (R)-lactate + NAD+  <-  pyruvate + NADH + H+   \n",
       "4    all-trans-octaprenyl diphosphate + 1,4-dihydro...   \n",
       "..                                                 ...   \n",
       "233  1 MNXM11@MNXD1 + 1 MNXM167418@MNXD1 + 1 MNXM72...   \n",
       "234  1 MNXM11@MNXD1 + 1 MNXM16655@MNXD1 + 1 MNXM728...   \n",
       "235  1 MNXM1104679@MNXD1 + 1 MNXM1446@MNXD1 + 1 MNX...   \n",
       "236          1 MNXM1101258@MNXD1 = 1 MNXM1104529@MNXD1   \n",
       "237  1 MNXM11@MNXD1 + 1 MNXM167419@MNXD1 + 1 MNXM72...   \n",
       "\n",
       "                                id       ncbiprotein  \\\n",
       "0                     4.2.1.59-RXN  [WP_011275247.1]   \n",
       "1    ADOMET-DMK-METHYLTRANSFER-RXN  [WP_011275738.1]   \n",
       "2                       ATPASE-RXN  [WP_016931380.1]   \n",
       "3             DLACTDEHYDROGNAD-RXN  [WP_011274818.1]   \n",
       "4                          DMK-RXN  [WP_016931372.1]   \n",
       "..                             ...               ...   \n",
       "233                     MNXR124075  [WP_011276231.1]   \n",
       "234                     MNXR124080  [WP_011275493.1]   \n",
       "235                     MNXR146695  [WP_011275264.1]   \n",
       "236                     MNXR151539  [WP_033079611.1]   \n",
       "237                     MNXR124096  [WP_011275561.1]   \n",
       "\n",
       "                                             reference       via  \n",
       "0                                                 None    BioCyc  \n",
       "1                                                 None    BioCyc  \n",
       "2                                                 None    BioCyc  \n",
       "3                                                 None    BioCyc  \n",
       "4                                                 None    BioCyc  \n",
       "..                                                 ...       ...  \n",
       "233  {'metacyc.reaction': 'TRYPTOPHAN--TRNA-LIGASE-...  MetaNetX  \n",
       "234  {'metacyc.reaction': 'TYROSINE--TRNA-LIGASE-RX...  MetaNetX  \n",
       "235  {'metacyc.reaction': 'UDP-NACMURALGLDAPAALIG-R...  MetaNetX  \n",
       "236  {'metacyc.reaction': 'UDPGLCNACEPIM-RXN', 'ori...  MetaNetX  \n",
       "237  {'metacyc.reaction': 'VALINE--TRNA-LIGASE-RXN'...  MetaNetX  \n",
       "\n",
       "[238 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find missing reactions\n",
    "gfbc.find_missing_reactions(cmodel)\n",
    "\n",
    "# Get/show missing reactions\n",
    "gfbc.missing_reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fill model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding genes to model: 100%|██████████| 13/13 [00:00<00:00, 8296.71it/s]\n",
      "Trying to add missing reacs: 100%|██████████| 226/226 [02:29<00:00,  1.51it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['add_to_GPR', 'ec-code', 'equation', 'id', 'ncbiprotein', 'reference',\n",
      "       'via'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding genes to model: 100%|██████████| 88/88 [00:00<00:00, 32306.24it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Model ATCC29970>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fill model\n",
    "gfbc.fill_model(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genes': {'missing (before)': 235,\n",
       "  'duplicates': 0,\n",
       "  'added': 101,\n",
       "  'missing (after)': 134,\n",
       "  'missing (unmappable)': 0},\n",
       " 'reactions': {'added (total)': 183,\n",
       "  'failed to build': 40,\n",
       "  'missing (before)': 240,\n",
       "  'add to GPR (BioCyc)': 11,\n",
       "  'mapped2MNX': 227,\n",
       "  'mapped2BiGG': 0,\n",
       "  'remaining_unmapped': 2},\n",
       " 'metabolites': {}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get raw statistics\n",
    "gfbc._statistics"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sprg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
