{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "from refinegems.classes.gapfill import KEGGapFiller\n",
    "from refinegems.utility.io import load_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435.xml'\n",
    "\n",
    "model = load_model(mpath,'libsbml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ec-code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[WP_011274359.1]</td>\n",
       "      <td>SH0001</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[WP_011274360.1]</td>\n",
       "      <td>SH0002</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[WP_172458781.1]</td>\n",
       "      <td>SH0003</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[WP_011274362.1]</td>\n",
       "      <td>SH0004</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[WP_011274363.1]</td>\n",
       "      <td>SH0005</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>[WP_145424740.1]</td>\n",
       "      <td>pSHaeC05</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2574</th>\n",
       "      <td>[WP_011276926.1]</td>\n",
       "      <td>pSHaeC06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>[WP_011276927.1]</td>\n",
       "      <td>pSHaeC07</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>[WP_011276928.1]</td>\n",
       "      <td>pSHaeC08</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2577</th>\n",
       "      <td>[WP_011276930.1]</td>\n",
       "      <td>pSHaeC10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1692 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ncbiprotein locus_tag ec-code\n",
       "0     [WP_011274359.1]    SH0001    None\n",
       "1     [WP_011274360.1]    SH0002    None\n",
       "2     [WP_172458781.1]    SH0003    None\n",
       "3     [WP_011274362.1]    SH0004    None\n",
       "4     [WP_011274363.1]    SH0005    None\n",
       "...                ...       ...     ...\n",
       "2573  [WP_145424740.1]  pSHaeC05    None\n",
       "2574  [WP_011276926.1]  pSHaeC06    None\n",
       "2575  [WP_011276927.1]  pSHaeC07    None\n",
       "2576  [WP_011276928.1]  pSHaeC08    None\n",
       "2577  [WP_011276930.1]  pSHaeC10    None\n",
       "\n",
       "[1692 rows x 3 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff2 = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435_RefSeq.gff'\n",
    "missing_genes = get_missing_genes(gff2,load_model('/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JCSC1435.xml','libsbml'))\n",
    "missing_genes\n",
    "\n",
    "# @TODO : some od the genes do not have an old locus tag -> what to do?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "from refinegems.classes.gapfill import GeneGapFiller\n",
    "from refinegems.utility.io import load_model\n",
    "\n",
    "modelpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435.xml'\n",
    "model = load_model(modelpath,'libsbml')\n",
    "cmodel = load_model(modelpath,'cobra')\n",
    "\n",
    "gffpath = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435_RefSeq.gff'\n",
    "\n",
    "gf2 = GeneGapFiller()\n",
    "gf2_missing_genes = gf2.get_missing_genes(gffpath,model)\n",
    "# ncbiprotein | locus_tag | ec-code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbiprotein</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>ec-code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_011274359.1</td>\n",
       "      <td>SH0001</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_011274360.1</td>\n",
       "      <td>SH0002</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_172458781.1</td>\n",
       "      <td>SH0003</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_011274362.1</td>\n",
       "      <td>SH0004</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_011274363.1</td>\n",
       "      <td>SH0005</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>WP_145424740.1</td>\n",
       "      <td>pSHaeC05</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2574</th>\n",
       "      <td>WP_011276926.1</td>\n",
       "      <td>pSHaeC06</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>WP_011276927.1</td>\n",
       "      <td>pSHaeC07</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>WP_011276928.1</td>\n",
       "      <td>pSHaeC08</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2577</th>\n",
       "      <td>WP_011276930.1</td>\n",
       "      <td>pSHaeC10</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1692 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ncbiprotein locus_tag ec-code\n",
       "0     WP_011274359.1    SH0001    None\n",
       "1     WP_011274360.1    SH0002    None\n",
       "2     WP_172458781.1    SH0003    None\n",
       "3     WP_011274362.1    SH0004    None\n",
       "4     WP_011274363.1    SH0005    None\n",
       "...              ...       ...     ...\n",
       "2573  WP_145424740.1  pSHaeC05    None\n",
       "2574  WP_011276926.1  pSHaeC06    None\n",
       "2575  WP_011276927.1  pSHaeC07    None\n",
       "2576  WP_011276928.1  pSHaeC08    None\n",
       "2577  WP_011276930.1  pSHaeC10    None\n",
       "\n",
       "[1692 rows x 3 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gf2_missing_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'map_ec_to_reac' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[120], line 24\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# find missing reaction\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# case: no EC no ncbiprotein\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;66;03m# case: already has EC\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;66;03m#       -> directly access reaction \u001b[39;00m\n\u001b[0;32m---> 24\u001b[0m mapped_reacs \u001b[38;5;241m=\u001b[39m \u001b[43mmap_ec_to_reac\u001b[49m(missing_genes[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mec-code\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mncbiprotein\u001b[39m\u001b[38;5;124m'\u001b[39m]])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'map_ec_to_reac' is not defined"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from refinegems.classes.gapfill import map_ec_to_reac\n",
    "from refinegems.curation.db_access.db import get_ec_from_ncbi\n",
    "# for notebook, later should be part of object\n",
    "manual_curation = {}\n",
    "\n",
    "# find missing reaction\n",
    "def get_missing_reacs(missing_genes, mail:str=None, check_NCBI:bool=False):\n",
    "    \n",
    "    # Case 1:  no EC, no ncbiprotein\n",
    "    # ------------------------------\n",
    "    case_1 = missing_genes[missing_genes['ncbiprotein'].isna() & missing_genes['ec-code'].isna()]\n",
    "    if len(case_1) > 0:\n",
    "        pass\n",
    "#       -> BLAST against SwissProt to get EC/BRENDA \n",
    "#          (mappings + sequences can be downloaded from DB)\n",
    "            # two tables: \n",
    "            #   1.: TSV with UniprotID, BRENDA and EC -7.7MB (26.07.2024)\n",
    "            #   2.: FASTA with sequences ~280MB (26.07.2024)\n",
    "            # more infos:\n",
    "            #   - API generation possible for automatic download\n",
    "            #   - UniprotID can be extracted from FASTA header sp|A0A1B0GTW7|...\n",
    "            #   - maybe write a small script to download stuff and shorten to actual EC/BRENDA mappings\n",
    "#       -> sth like DeepECTransformer (tools either not good, \n",
    "#          not installable or no license)\n",
    "#       -> use ECRECer Web service output as input \n",
    "#          (whole protein fasta -> wait for Gwendolyn's results) -> does not work / not return\n",
    "#       -> use CLEAN webservice\n",
    "\n",
    "    manual_curation['no ncbiprotein, no EC'] = case_1[case_1['ncbiprotein'].isna() & case_1['ec-code'].isna()] \n",
    "    mapped_reacs = pd.concat(case_1[~(case_1['ncbiprotein'].isna() & case_1['ec-code'].isna())] , \n",
    "                missing_genes[~(missing_genes['ncbiprotein'].isna() & missing_genes['ec-code'].isna())])\n",
    "\n",
    "    # Case 2: no EC but ncbiprotein\n",
    "    # ----------------------------- \n",
    "    #       -> access ncbi for ec (optional) \n",
    "    if check_NCBI and mail:\n",
    "        mapped_reacs['ec-code'] = mapped_reacs.progress_apply(lambda x: get_ec_from_ncbi(mail,x['ncbiprotein']) if not x['ec-code'] else x['ec-code'], axis=1)\n",
    "    \n",
    "    manual_curation['no EC'] = mapped_reacs[mapped_reacs['ec-code'].isna()]\n",
    "    mapped_reacs = mapped_reacs[~mapped_reacs['ec-code'].isna()]\n",
    "\n",
    "    # Case 3: already has EC\n",
    "    # ----------------------\n",
    "    #       -> directly access reaction \n",
    "    mapped_reacs = map_ec_to_reac(mapped_reacs[['ec-code','ncbiprotein']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/missing_genes.fasta')"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from pathlib import Path\n",
    "\n",
    "outdir = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/'\n",
    "fasta = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/JCSC1435_proteins_genome.fasta'\n",
    "\n",
    "\n",
    "def create_missing_genes_protein_fasta(fasta,outdir, missing_genes):\n",
    "    \n",
    "    # format the missing genes' locus tags\n",
    "    locus_tags_list = list(missing_genes['locus_tag'])\n",
    "    locus_tags_dict = {_:'[locus_tag='+_+']' for _ in list(missing_genes['locus_tag'])}\n",
    "    \n",
    "    # parse the protein FASTA\n",
    "    protfasta = SeqIO.parse(fasta,'fasta')\n",
    "    \n",
    "    # extract the sequences of the missing genes only\n",
    "    missing_seqs = []\n",
    "    for seq in protfasta:\n",
    "        # Case 1: locus tag equals FASTA header\n",
    "        if seq.id in locus_tags_dict.keys():\n",
    "            missing_seqs.append(seq)\n",
    "        # Case 2: locus tag as descriptor\n",
    "        elif any([k for k,v in locus_tags_dict.items() if v in seq.description]):\n",
    "            for k,v in locus_tags_dict.items():\n",
    "                if v in seq.description:\n",
    "                    seq.id = k\n",
    "                    missing_seqs.append(seq)\n",
    "                    break\n",
    "        # Case _: locus tag either in model or errounous\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    # save the collected sequences in a new file\n",
    "    outfile = Path(outdir,'missing_genes.fasta')\n",
    "    SeqIO.write(missing_seqs, outfile, 'fasta')\n",
    "    \n",
    "    return outfile\n",
    "\n",
    "\n",
    "create_missing_genes_protein_fasta(fasta,outdir,gf2_missing_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "from typing import Literal\n",
    "\n",
    "def run_DIAMOND_blastp(fasta:str, db:str, \n",
    "                       sensitivity:Literal['sensitive', 'more-sensitive', 'very-sensitive','ultra-sensitive']='more-sensitive',\n",
    "                       coverage:float=95.0,\n",
    "                       threads:int=2,\n",
    "                       outdir:str=None, outname:str='DIAMOND_blastp_res.tsv'):\n",
    "    \n",
    "    if outdir:\n",
    "        outname = Path(outdir,'DIAMOND_blastp_res.tsv')\n",
    "        logfile = Path(outdir,'log_DIAMOND_blastp.txt')\n",
    "    else:\n",
    "        outname = Path(outname)\n",
    "        logfile = Path('log_DIAMOND_blastp.txt')\n",
    "      \n",
    "    # @TODO: test, if it works with different paths and their problems  \n",
    "    # @TODO: write additional output to a logfile, not stderr\n",
    "    subprocess.run([F'diamond blastp -d {db} -q {fasta} --{sensitivity} --query-cover {coverage} -p {int(threads)} -o {outname} --outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore > {logfile}'], shell=True)\n",
    "\n",
    "    return outname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "diamond v2.1.9.163 (C) Max Planck Society for the Advancement of Science, Benjamin Buchfink, University of Tuebingen\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 2\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: /Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435\n",
      "#Target sequences to report alignments for: 25\n",
      "Opening the database...  [0.017s]\n",
      "Database: /Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/swissprot.dmnd (type: Diamond database, sequences: 571864, letters: 207016062)\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0s]\n",
      "Masking queries...  [0s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0s]\n",
      "Seeking in database...  [0s]\n",
      "Loading reference sequences...  [0.238s]\n",
      "Masking reference...  [9.99s]\n",
      "Initializing temporary storage...  [0.001s]\n",
      "Building reference histograms...  [6.957s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.884s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.117s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.003s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.859s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.858s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.859s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.112s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.846s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.846s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.104s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.868s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.112s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.857s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.112s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.855s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.856s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.857s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.113s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.855s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.114s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.854s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.102s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.853s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.111s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.853s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.85s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.85s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.108s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.872s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.87s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.871s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.111s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.869s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.105s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.848s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.846s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.111s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.111s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.846s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.105s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.85s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.104s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.855s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.113s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.855s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.108s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.853s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.854s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.843s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.102s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.843s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.852s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.116s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.851s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.842s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.846s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.104s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.108s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.106s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.112s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.842s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.096s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.864s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.863s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.102s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.865s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.864s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.107s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.85s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.115s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.849s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.105s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.847s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.11s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.844s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.1s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.109s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.111s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.845s]\n",
      "Building query seed array...  [0s]\n",
      "Computing hash join...  [0.105s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0s]\n",
      "Deallocating memory...  [0s]\n",
      "Deallocating buffers...  [0.003s]\n",
      "Clearing query masking...  [0s]\n",
      "Computing alignments... Loading trace points...  [0.003s]\n",
      "Sorting trace points...  [0s]\n",
      "Computing alignments...  [0.008s]\n",
      "Deallocating buffers...  [0s]\n",
      "Loading trace points...  [0s]\n",
      " [0.011s]\n",
      "Deallocating reference...  [0.003s]\n",
      "Loading reference sequences...  [0s]\n",
      "Deallocating buffers...  [0s]\n",
      "Deallocating queries...  [0s]\n",
      "Loading query sequences...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the output file...  [0s]\n",
      "Closing the database...  [0s]\n",
      "Cleaning up...  [0s]\n",
      "Total time = 78.842s\n",
      "Reported 2 pairwise alignments, 2 HSPs.\n",
      "1 queries aligned.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/DIAMOND_blastp_res.tsv')"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_fasta = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/mini_test.fasta'\n",
    "db = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/swissprot.dmnd'\n",
    "outdir = '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/'\n",
    "\n",
    "run_DIAMOND_blastp(test_fasta, db, outdir=outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_DIAMOND_blastp_results(blasttsv:str, pid_theshold:float=90.0):\n",
    "    \n",
    "    if pid_theshold > 100.0 or pid_theshold < 0.0:\n",
    "        raise ValueError('PID threshold has to be between 0.0 and 100.0')\n",
    "    \n",
    "    # load diamond results\n",
    "    diamond_results = pd.read_csv(blasttsv, sep='\\t', header=None)\n",
    "    diamond_results.columns = ['query_ID', 'subject_ID', 'PID', 'align_len', 'no_mismatch', 'no_gapopen', 'query_start', 'query_end', 'subject_start', 'subject_end','E-value','bitscore']\n",
    "    # filter by PID\n",
    "    diamond_results = diamond_results[diamond_results['PID']>=pid_theshold]\n",
    "    # trim cols\n",
    "    diamond_results = diamond_results[['query_ID','subject_ID']]\n",
    "    \n",
    "    return diamond_results\n",
    "    \n",
    "dmnd_res = filter_DIAMOND_blastp_results('/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/JSC1435/DIAMOND_blastp_res.tsv', pid_theshold=40.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>UniProt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH0062</td>\n",
       "      <td>O05052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag UniProt\n",
       "0    SH0062  O05052"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dmnd_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>ec-code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SH0062</td>\n",
       "      <td>O05052</td>\n",
       "      <td>3.1.21.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag UniProt   ec-code\n",
       "0    SH0062  O05052  3.1.21.3"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def map_dmnd_res_to_sp_ec_brenda(dmnd_results, swissprot_mapping_path):\n",
    "    \n",
    "    def _combine_EC_BRENDA(brenda,ec):\n",
    "    \n",
    "        nums_list = []\n",
    "        if brenda and not pd.isna(brenda):\n",
    "            for num in brenda.split(';'):\n",
    "                nums_list.append(num.strip())\n",
    "\n",
    "        if ec and not pd.isna(ec):\n",
    "            for num in ec.split(';'):\n",
    "                nums_list.append(num.strip())\n",
    "            \n",
    "        return nums_list\n",
    "    \n",
    "    # load the SwissProt mapping file\n",
    "    swissprot_mapping = pd.read_csv(swissprot_mapping_path, sep='\\t')\n",
    "    swissprot_mapping.dropna(subset=['BRENDA','EC number'], how='all',inplace=True)\n",
    "    \n",
    "    # extract SwissProts IDs from subject_ID\n",
    "    dmnd_results.columns = ['locus_tag','UniProt']\n",
    "    dmnd_results['UniProt'] = dmnd_results['UniProt'].apply(lambda x: x.split('|')[1])\n",
    "    \n",
    "    # match\n",
    "    dmnd_results = dmnd_results.merge(swissprot_mapping, left_on='UniProt', right_on='Entry', how='left')\n",
    "    dmnd_results.drop('Entry', axis=1, inplace=True)\n",
    "    dmnd_results['ec-code'] = dmnd_results.apply(lambda x: _combine_EC_BRENDA(x['BRENDA'],x['EC number']), axis=1)\n",
    "    dmnd_results.drop(['BRENDA','EC number'], axis=1, inplace=True)\n",
    "    dmnd_results = dmnd_results.explode('ec-code')\n",
    "    \n",
    "    return dmnd_results\n",
    "    \n",
    "    \n",
    "map_dmnd_res_to_sp_ec_brenda(dmnd_res.copy(), '/Users/brune/Documents/11_Test_Data/test_refinegems/test_gapfill/uniprot_table.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sprg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
