Command Line access points
==========================

After a successfull installation, refineGEMs can be accessed via the command line
from inside the Python environment it was installed in using:

.. code-block:: bash

    refinegems [OPTIONS] COMMAND [ARGS]


The following commands are available:

- ``analyse``: Provides commands for analysing the model(s).
- ``gaps``: Provides commands for performing gap-filling.
- ``media``: Provides commands for interacting with the media database.
- ``polish``: Provides commands for polising a model generated by an automated pipeline.
- ``refine``: Provides commands for refining a model. 
- ``setup``: Provides commands for the set-up or download of files and directories for the tool.

General Options
---------------

- ``--help``: Call the help page of the command.
- ``--version``: Show the version  and exit.

refinegems analyse 
------------------

.. code:: bash 
    
    refinegems analyse compare [MODELPATH]

Compare multiple models.

Options:

- ``--type/-t <sboterm|entities>`` : Type of comparison to perform. Multiple can be added.
- ``--all`` : Perform all comparisons. Overwrites previous option.

.. code:: bash  
    
    refinegems analyse memote [MODELPATH]
    
Perform a memote analsis on a model.

Options:

- ``--score-only/-s`` : Return only the final score of the analysis.
- ``--file/-f`` :  Name/Path to save output to. Only relecant if ``-s`` was not set.

.. code:: bash 
  
  refinegems analyse pancore [MODELPATH] [PCPATH]

Compare the pan-core(-novel) information content of a model to a pan-core model.

Options:

- ``--based-on/-b`` : How to compare the models. Deafult is a comparison based on the IDs 
  of the model entities
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse pathways [MODELPATH]

Analyse the KEGG pathways contained in a model.

Options:

- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse stats [MODELPATH]

Perform a statistical analysis of the model.

Options:

- ``--dir/-d`` : Path to the output directory. 
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.

refinegems analyse growth 
^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: bash 
  
  refinegems analyse growth auxothrophies [MODELPATH]

For a given set of media, test simulate the amino acid auxotrohpies of the model.

Options:

- ``--media/-m`` : Path to a media config file. REQUIRED.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth minimal-medium [MODELPATH]

Calculate the minimal medium of a model. Can either be by minimising the fluxes of the current medium
(``flux``), finding the minimal number of compounds needed for growth based on the current medium (``medium``) or the
minimal number of compounds based on the available exchange reactions (``exchanges``).

Options:

- ``--objective/-o`` : One of the three options listed to choose what to optimize for.
- ``--growth-rate/-r`` : Minimal growth rate that should be reached on the minimal medium. The smaller the value, the more computational expensive the calculation will be.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth simulate [MODELPATHS]

Simulate the growth of one or multiple models one or more media.

Options:

- ``--media/-m`` : Path to a media config file. REQUIRED.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth sources [MODELPATH]

Simulate the growth for different sources. When choosing the starting medium, 
it is useful to have at least one additional source of every other neccessary element separated 
from the one that is tested against.

Options:

- ``--element/-e``: The element to perform the source test for. Should be a valid chemical symbol.
- ``--substances/-s``: Substances for substitution. Multiple can be given. If none are given, all options found in the database will be utelised as potential sources.
- ``--medium/-m``: Medium abbreviation of a medium in the database, if the medium currently saved with the model should be substitued before testing.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 

refinegems gaps
---------------

.. code:: bash 
  
  refinegems gaps find [MODELPATH] [GFF_FILE] [ORGANISMID] [GAPFILL_PARAMS] [FILENAME]

Find gaps in a model based on the genes/gene products of the underlying organism.

.. code:: bash 
  
  refinegems gaps fill [MODEL] [GAP_ANALYSIS_RESULTS]

Fill the gaps in a model based on a user-provided input file.

.. code:: bash 
  
  refinegems gaps autofill [MODELPATH] [GAFILL_PARAMS] [FILENAME]

Automatically find and fill the gaps based on the genes/gene products.

refinegems media
----------------

.. code:: bash 
  
  refinegems media info

Retrieve information about the media database.

Options:

- ``--list`` : List the available media.

.. code:: bash  
  
  refinegems media initialise

Initialse or update the media database.

refinegems polish
-----------------

.. code:: bash 
  
  refinegems polish run [MODEL] [EMAIL] [PATH]

Options:

- ``--id-db/-i`` : Main database that is the source of the IDs.
- ``--refseq-gff``: Path to RefSeq GFF file of organism
- ``--protein-fasta/-p``: File used as input for CarveMe, the protein fasta file.
- ``--lab-strain/-l``: True if the strain was sequenced in a local lab.
- ``--kegg-organism-id/-k``: KEGG Identifier for the organism, if available.

refinegems refine
-----------------

.. code:: bash 
  
  refinegems refine biomass [MODELPATH]

Normalise the biomass objective functions of a model to improve the model consistency.

Options:

- ``--cycles/-c`` : Maximal number of normalisation cycles to ensure the program is not running endlessly. The default is 10.
- ``--outfile/-o`` : Optional, filename to save the updated model under. 

.. code:: bash 
  
  refinegems refine charges [MODELPATH]

Compare the charges in a model to the ModelSEED database and adjust them accordingly, if neccessary.

Options:

- ``--dir/-d`` : Path to the output directory. 


.. code:: bash 
  
  refinegems refine direction [MODELPATH]

Check and - if neccessary - correct the direction of th reactions in a model.

.. code:: bash 
  
  refinegems refine egcs [MODELPATH]

Indentify problematic energy generating cycles (EGCs) in the model 
an optionally, try to resolve them.

Options:

- ``--solver/-s``: Options to solve the EGCs. Following Solvers are currently available:
    - greedy: Solve the EGCs by trying out modifications of single reactions (reversible, backward,forward,remove)
- ``--namespace/-n``: Namespace of the model.
- ``--compartment/-c``: Compartments to check, separated by comma. Default is ``c,p``
- ``--outfile/-o`` : File to save the updated model to.

refinegems refine annot
^^^^^^^^^^^^^^^^^^^^^^^

.. code:: bash 
  
  refinegems refine annot sboterms [MODELPATH]

Call the SBOannotator on a model to enhance/add SBOterms to the annotations.

.. code:: bash 
  
  refinegems refine annot pathways [MODELPATH]

Add the KEGG pathways as group entities to the model.


refinegems setup 
----------------

.. code:: bash 
  
  refinegems setup build_pancore [MODELS]

Using the given models, construct a pan-core model.

Options:

- ``--based-on/-o`` : Option on how to compare the models, defaults to 'id', which compares and combines the models based in the IDs only.
- ``--name/-n``: Set the name of the constructed pan-core model.
- ``--keep-genes/-g`` : Keep the genes in the pan-core model, otherwise they are deleted and only reactions and metabolites are kept.
- ``--resolve-compartments/--rcomp`` : Try to standardise the model's compartment names.
- ``--dir/-d`` : Path to the output directory. 