Command Line access points
==========================

After a successfull installation, refineGEMs can be accessed via the command line
from inside the Python environment it was installed in using:

.. code-block:: bash

    refinegems [OPTIONS] COMMAND [ARGS]


The following commands are available:

- ``analyse``: Provides commands for analysing the model(s).
- ``database``: Provides commands for handling the internal database.
- ``gaps``: Provides commands for performing gap-filling.
- ``media``: Provides commands for interacting with the media (database).
- ``polish``: Provides commands for polising a model generated by an automated pipeline.
- ``refine``: Provides commands for refining a model. 
- ``setup``: Provides commands for the set-up or download of files and directories for the tool.

General Options
---------------

- ``--help``: Call the help page of the command.
- ``--version``: Show the version  and exit.

----


refinegems analyse 
------------------

Analyse a model by testing for auxotrophies and growth on different media along 
with finding EGCs and looking at overall model statistics. 

.. code:: bash 
    
    refinegems analyse compare [MODELPATH]

Compare multiple models.

Options:

- ``--type/-t <sboterm|entities>`` : Type of comparison to perform. Multiple can be added.
- ``--all`` : Perform all comparisons. Overwrites previous option.

.. code:: bash  
    
    refinegems analyse memote [MODELPATH]
    
Perform a memote analysis on a model.

Options:

- ``--score-only/-s`` : Return only the final score of the analysis.
- ``--file/-f`` :  Name/Path to save output to. Only relevant if ``-s`` was not set.

.. code:: bash 
  
  refinegems analyse pancore [MODELPATH] [PCPATH]

Compare the pan-core(-novel) information content of a model to a pan-core model.

Options:

- ``--based-on/-b`` : How to compare the models. Deafult is a comparison based on the IDs 
  of the model entities
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse pathways [MODELPATH]

Analyse the KEGG pathways contained in a model.

Options:

- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse stats [MODELPATH]

Perform a statistical analysis of the model.

Options:

- ``--dir/-d`` : Path to the output directory. 
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.

refinegems analyse growth 
^^^^^^^^^^^^^^^^^^^^^^^^^

Analyse the growth under different conditions.

.. code:: bash 
  
  refinegems analyse growth auxothrophies [MODELPATH]

For a given set of media, test simulate the amino acid auxotrohpies of the model.

Options:

- ``--media/-m`` : Path to a media config file. REQUIRED.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth minimal-medium [MODELPATH]

Calculate the minimal medium of a model. Can either be calculated by minimising the fluxes of the current medium
(``flux``), finding the minimal number of compounds needed for growth based on the current medium (``medium``) or the
minimal number of compounds based on the available exchange reactions (``exchanges``).

Options:

- ``--objective/-o`` : One of the three options listed to choose what to optimize for.
- ``--growth-rate/-r`` : Minimal growth rate that should be reached on the minimal medium. The smaller the value, the more computational expensive the calculation will be.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth simulate [MODELPATHS]

Simulate the growth of one or multiple models on one or more media.

Options:

- ``--media/-m`` : Path to a media config file. REQUIRED.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 

.. code:: bash 
  
  refinegems analyse growth sources [MODELPATH]

Simulate the growth for different sources. When choosing the starting medium, 
it is useful to have at least one additional source of every other neccessary element separated 
from the one that is tested against.

Options:

- ``--element/-e``: The element to perform the source test for. Should be a valid chemical symbol.
- ``--substances/-s``: Substances for substitution. Multiple can be given. If none are given, all options found in the database will be utilised as potential sources.
- ``--medium/-m``: Medium abbreviation of a medium in the database, if the medium currently saved with the model should be substitued before testing.
- ``--namespace/-n`` : Namespace of the model.
- ``--colors/-c`` : Abbreviation of a matplotlib color palette for the graphics.
- ``--dir/-d`` : Path to the output directory. 


refinegems database
-------------------

Access, curate, etc. the in-build database.

.. code:: bash  
  
  refinegems database initialise

Initialise or update the in-build database.

.. code:: bash 

  refinegems database add_namespace [DATABASENAME]

Add or update tables for additional namespaces/databases into/of the in-build database.

Options:

- ``--chunksize/-c``: Size (in kB) of data to download per chunk, if a download is required.

.. code:: bash

  refinegems database reset

Reset the database by removing all additionally added tables (added with e.g. add_namespace)


refinegems gaps
---------------

.. code:: bash 
  
  refinegems gaps automated_gapfill [ALG] [MODELPATH] [OPTIONS]

Find gaps in a model and optionally try to fill them.
Algorithm for the gap-filling is chosen by the first argument.

General Options

- ``-o``, ``--outdir``: Path to a directory to write the output to.  [default: ``./``]
- ``-f``, ``--fill``: If True, tries to fill the gaps in the model.
- ``-fc``, ``--formula-check``: [none,existence,wildcard,strict] Set the filter for which metabolite formulas are valid to be added to the model. [default: existence]
- ``--no-dna``: Exclude DNA reactions (name-based) from being added to the model.
- ``--no-rna``: Exclude RNA reactions (name-based) from being added to the model.
- ``-p``, ``--idprefix``: Prefix for the pseudo-IDs, if an ID does not exists for the given namespace. [default: refineGEMs]
- ``-n``, ``--namespace``: [BiGG]  Namespace used in the model.  [default: BiGG]

| KEGG required parameters: [all required if ``ALG="KEGG"``]
| Parameters required when running the KEGGapFiller algorithm

- ``--orgid``: KEGG organism ID

| BioCyc required parameters: [all required if ``ALG="BioCyc"``]
| Parameters required when running the BioCycGapFiller algorithm

- ``-gt``, ``--genetable``: Path to the BioCyc gene SmartTable.
- ``-rt``, ``--reactable``: Path to the BioCyc reaction SmartTable.
- ``--gff-bc``: Path to the GFF.

| Gene required parameters: [all required if ``ALG="Gene"``]
| Parameters required when running the GeneGapFiller algorithm

- ``--gff-g``: Path to the GFF.

| Gene optional parameters:
| Optional / conditionally interdependant parameters for the GeneGapFiller algorithm

- ``--prot-prefix``: Prefix for pseudo-protein IDs.  [default: refineGEMs]
- ``--mail``: Mail address for NCBI requests.
- ``--ncbi``, ``--check-ncbi``: Enable searching protein IDs in NCBI. This increases the runtime significantly.
- ``--fasta``: Path to the protein FASTA of the model.
- ``--dmnd-db``: Path to the SwissProt DIAMOND database.
- ``-sp-map``, ``--swissprot-mapping``: Path to the SwissProt mapping file (ID against EC and BRENDA)
- ``-s``, ``--sensitivity``: [sensitive,more-sensitive,very-sensitive,ultra-sensitive] Sensitivity mode for running DIAMOND.  [default: more-sensitive]
- ``--cov``: Coverage value (passed to DIAMOND)  [default: 90.0]
- ``--pid``: Percentage identity threshold value for filtering DIAMOND results. [default: 95.0]
- ``-t``, ``--threads``: Number of threads to be used by DIAMOND.  [default: 2]

Constraints:

- ``--mail`` is required if ``--check-ncbi`` is set
- if one of ``--fasta``, ``--dmnd-db``, ``--swissprot-mapping`` is set, all need to be set
- ``--fasta``, ``--dmnd-db``, ``--swissprot-mapping`` are all required if any of ``--sensitivity``, ``--cov``, ``--pid`` and ``--threads`` is set


refinegems media
----------------

Access the media part of the database.

.. code:: bash 
  
  refinegems media info

Retrieve information about the media database.

Options:

- ``--list`` : List the available media.


refinegems polish
-----------------

.. code:: bash 
  
  refinegems polish run [MODEL] [EMAIL] [PATH]

Options:

- ``--id-db/-i`` : Main database that is the source of the IDs.
- ``--refseq-gff``: Path to RefSeq GFF file of organism
- ``--protein-fasta/-p``: File used as input for CarveMe, the protein fasta file.
- ``--lab-strain/-l``: True if the strain was sequenced in a local lab.
- ``--kegg-organism-id/-k``: KEGG Identifier for the organism, if available.

refinegems refine
-----------------

Refine a model. Includes steps like biomass, charges, SBO annotation, reaction direction correction and addition 
of pathways and further gene product annotations.

.. code:: bash 
  
  refinegems refine biomass [MODELPATH]

Normalise the biomass objective function(s) of a model to improve the model consistency.

Options:

- ``--cycles/-c`` : Maximal number of normalisation cycles to ensure the program is not running endlessly. The default is 10.
- ``--outfile/-o`` : Optional, filename to save the updated model under. 

.. code:: bash 
  
  refinegems refine charges [MODELPATH]

Compare the charges in a model to the ModelSEED database and adjust them accordingly, if neccessary.

Options:

- ``--dir/-d`` : Path to the output directory. 


.. code:: bash 
  
  refinegems refine direction [MODELPATH]

Check and - if neccessary - correct the direction of the reactions in a model.

.. code:: bash 
  
  refinegems refine egcs [MODELPATH]

Indentify problematic energy generating cycles (EGCs) in the model 
an optionally, try to resolve them.

Options:

- ``--solver/-s``: Options to solve the EGCs. Following Solvers are currently available:

      - greedy: Solve the EGCs by trying out modifications of single reactions (reversible, backward,forward,remove)

- ``--namespace/-n``: Namespace of the model.
- ``--compartment/-c``: Compartments to check, separated by comma. Default is ``c,p``
- ``--outfile/-o`` : File to save the updated model to.

refinegems refine annot
^^^^^^^^^^^^^^^^^^^^^^^

Add annotations to your model.

.. code:: bash 
  
  refinegems refine annot sboterms [MODELPATH]

Call the SBOannotator on a model to enhance/add SBOterms to the annotations.

.. code:: bash 
  
  refinegems refine annot pathways [MODELPATH]

Add the KEGG pathways as group entities to the model.


refinegems setup 
----------------

Set-up tools, folder structure and more for running the program.

.. code:: bash

  refinegems setup config 

Download a configuration file for a specific functionality of the toolbox.

Options:

- ``--filename/-f`` : Name or path of a file to save the config under.
- ``--type/-t`` : Type of config to download. Defaults to ``media``.

.. code:: bash

  refinegems setup data [DOWNLOADTYPE]

Download file(s) needed for a given functionality of the toolbox.

Current options include:

- 'SwissProt\_gapfill': download the SwissProt sequences as a FASTA and a mapping of their ID to BRENDA/EC for gap-filling

Options:

- ``--dir/-d`` : Path to directory to save the downloaded files to.
- ``--chunksize/-c`` : Size of the chunk to download (in kB) at once.

.. code:: bash 
  
  refinegems setup build_pancore [MODELS]

Using the given models, construct a pan-core model.

Options:

- ``--based-on/-o`` : Option on how to compare the models, defaults to ``id``, which compares and combines the models based in the IDs only.
- ``--name/-n``: Set the name of the constructed pan-core model.
- ``--keep-genes/-g`` : Keep the genes in the pan-core model, otherwise they are deleted and only reactions and metabolites are kept.
- ``--resolve-compartments/--rcomp`` : Try to standardise the model's compartment names.
- ``--dir/-d`` : Path to the output directory. 