<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage as standalone application &mdash; refinegems 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples on how to use refinegems" href="examples.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            refinegems
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage as standalone application</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples on how to use refinegems</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage as python module</a></li>
<li class="toctree-l1"><a class="reference internal" href="explanation.html">Ideas behind some implementations</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">refinegems package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">refinegems</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage as standalone application</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/application.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="usage-as-standalone-application">
<h1>Usage as standalone application<a class="headerlink" href="#usage-as-standalone-application" title="Permalink to this heading"></a></h1>
<p>The script <code class="docutils literal notranslate"><span class="pre">main.py</span></code> can be used directly in the command line after
entering the virtual environment with <code class="docutils literal notranslate"><span class="pre">pipenv</span> <span class="pre">shell</span></code>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file contains defaults for all variables that need
to be set by the user.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">Description</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">&gt;</span>
<span class="w">  </span><span class="no">This file can be adapted to choose what refinegems should do.</span>
<span class="w">  </span><span class="no">Note: For windows use \ instead of / for the paths</span>


<span class="nt">General Setting</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">&gt;</span>
<span class="w">  </span><span class="no">Path to GEM to be investigated</span>

<span class="nt">model</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;data/e_coli_core.xml&#39;</span><span class="w"> </span><span class="c1">#&#39;../C_striatum_GEMs/models/Cstr_14.xml&#39;</span>
<span class="nt">entrez_email</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;famke.baeuerle@student.uni-tuebingen.de&#39;</span>


<span class="nt">Settings for scripts that manipulate the model</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">&gt;</span>
<span class="w">  </span><span class="no">They are all split into the ON / OFF switch (TRUE / FALSE) and additional settings like a path to where the new model should be saved.</span>

<span class="c1">### Addition of KEGG Pathways as Groups ###</span>
<span class="nt">keggpathways</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">kegg_path</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;../Nextcloud/master_thesis/models/Cstr_17_kegg.xml&#39;</span><span class="w"> </span><span class="c1"># path where to save model with KEGG Groups</span>

<span class="c1">### SBO-Term Annotation (requires PostgreSQL) ###</span>
<span class="nt">sboterms</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">sbo_path</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;../Nextcloud/master_thesis/models/Cstr_17_sbo.xml&#39;</span><span class="w"> </span><span class="c1"># path where to save model with sbo terms</span>

<span class="c1">### CarveMe polishing ###</span>
<span class="nt">polish</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">id_db</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;BIGG&#39;</span><span class="w"> </span><span class="c1"># Required!</span>
<span class="c1"># Possible identifiers, currently: BiGG &amp; VMH</span>
<span class="c1"># For other IDs the `polish` function in `polish.py` might need adjustment</span>
<span class="nt">lab_strain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span><span class="w"> </span><span class="c1"># Needs to be set to ensure that protein IDs get the &#39;bqbiol:isHomologTo&#39; qualifier</span>
<span class="w">               </span><span class="c1"># &amp; to set the locus_tag to the ones obtained by the annotation</span>
<span class="nt">protein_fasta</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="w"> </span><span class="c1"># Path to used CarveMe input file, if exists; Needs to be set for lab_strain: True</span>
<span class="nt">polish_path</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;../Nextcloud/master_thesis/models/Cstr_17_genes.xml&#39;</span><span class="w"> </span><span class="c1"># path where to save the polished model</span>

<span class="c1">### Charge correction ###</span>
<span class="nt">charge_corr</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">charge_path</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;../Nextcloud/master_thesis/models/Cstr_16_charges.xml&#39;</span>


<span class="nt">Settings for scripts that investigate the model</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">&gt;</span>
<span class="w">  </span><span class="no">These are only necessary if none of the scripts to manipulate the model are used.</span>

<span class="c1"># media to simulate growth from, available: SNM, LB, M9, SMM, CGXII, RPMI</span>
<span class="nt">media</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="s">&#39;SNM3&#39;</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="s">&#39;LB&#39;</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="s">&#39;M9&#39;</span><span class="p p-Indicator">]</span>

<span class="c1"># determine whether the memote score should be calculated, default: FALSE</span>
<span class="nt">memote</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>

<span class="c1"># Determine if output file should be created, default: cl</span>
<span class="c1"># Filename is set as the models name</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">xlsx</span><span class="w"> </span><span class="c1">#cl, xlsx, csv</span>

<span class="c1">### Gapfill ###</span>
<span class="c1"># All parameters are required for all db_to_compare choices except:</span>
<span class="c1"># - organismid which is only required for db_to_compare: &#39;KEGG&#39;/&#39;KEGG+BioCyc&#39;</span>
<span class="c1"># - and biocyc_tables which is not required for &#39;KEGG&#39;</span>
<span class="nt">gapfill_analysis</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">gapfill_analysis_params</span><span class="p">:</span>
<span class="w">  </span><span class="nt">db_to_compare</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;KEGG+BioCyc&#39;</span><span class="w">  </span><span class="c1"># One of the choices KEGG|BioCyc|GFF|KEGG+BioCyc</span>
<span class="w">  </span><span class="nt">organismid</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;Three-letter</span><span class="nv"> </span><span class="s">KEGG</span><span class="nv"> </span><span class="s">OrganismID&#39;</span><span class="w">  </span><span class="c1"># Needs to be specified for KEGG</span>
<span class="w">  </span><span class="nt">gff_file</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">RefSeq</span><span class="nv"> </span><span class="s">GFF</span><span class="nv"> </span><span class="s">file&#39;</span><span class="w">  </span><span class="c1"># Obtainable via the NCBI Assembly Accession page for the organism</span>
<span class="w">  </span><span class="nt">biocyc_files</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;Path0&#39;</span><span class="w">  </span><span class="c1"># Path to TXT file containing a SmartTable from BioCyc with the columns &#39;Accession-2&#39; &#39;Reaction of gene&#39; (-)</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;Path1&#39;</span><span class="w">  </span><span class="c1"># Path to TXT file containing a SmartTable with all reaction relevant information (*)</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;Path2&#39;</span><span class="w">  </span><span class="c1"># Path to TXT file containing a SmartTable with all metabolite relevant information (+)</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;Path3&#39;</span><span class="w">  </span><span class="c1"># Path to protein FASTA file used as input for CarveMe (Needed to get the protein IDs from the locus tags)</span>
<span class="c1"># (-) If the organism is not in BioCyc retrieve a table mapping all reactions in BioCyc to the corresponding sequence</span>
<span class="c1"># (*) &#39;Reaction&#39; &#39;Reactants of reaction&#39; &#39;Products of reaction&#39; &#39;EC-Number&#39; &#39;Reaction-Direction&#39; &#39;Spontaneous?&#39;</span>
<span class="c1"># (+) &#39;Compound&#39; &#39;Object ID&#39; &#39;Chemical Formula&#39; &#39;InChI-Key&#39; &#39;ChEBI&#39;</span>
<span class="c1"># For all BioCyc files the order should be the same. If the organism does not occur in the BioCyc database the</span>
<span class="c1"># complete tables for reactions can be used with the same columns.</span>
<span class="nt">gapfill_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FALSE</span>
<span class="nt">gapfill_model_in</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">Excel</span><span class="nv"> </span><span class="s">file</span><span class="nv"> </span><span class="s">with</span><span class="nv"> </span><span class="s">which</span><span class="nv"> </span><span class="s">gaps</span><span class="nv"> </span><span class="s">in</span><span class="nv"> </span><span class="s">model</span><span class="nv"> </span><span class="s">should</span><span class="nv"> </span><span class="s">be</span><span class="nv"> </span><span class="s">filled&#39;</span>
<span class="c1"># Either obtained by running gapfill_analysis/Created by hand with the same structure as the result file from gapfill_analysis</span>
<span class="c1"># Example Excel file to fill in by hand: data/modelName_gapfill_analysis_date_example.xlsx</span>
<span class="nt">gapfill_model_out</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;Path</span><span class="nv"> </span><span class="s">where</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">save</span><span class="nv"> </span><span class="s">gap</span><span class="nv"> </span><span class="s">filled</span><span class="nv"> </span><span class="s">model&#39;</span>

<span class="c1">### ModelSEED comparison ###</span>
<span class="nt">modelseed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRUE</span><span class="w"> </span><span class="c1"># set to False if not needed</span>
</pre></div>
</div>
<p>The repository structure has the following intention:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">refinegems/</span></code> contains all the functions needed in <code class="docutils literal notranslate"><span class="pre">main.py</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">data/</span></code> contains all tables that are used by different parts of the script as well as a toy model <code class="docutils literal notranslate"><span class="pre">e_coli_core.xml</span></code></p></li>
<li><p>Instead of using the files given in <code class="docutils literal notranslate"><span class="pre">data/</span></code>, you can use your own files and just change the paths in <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. Please be aware that some functions rely on input in a certain format so make sure to check the files given in the <code class="docutils literal notranslate"><span class="pre">data/</span></code> folder and use the same formatting.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">databases/</span></code> contains the <code class="docutils literal notranslate"><span class="pre">sql</span></code> file as well as the <code class="docutils literal notranslate"><span class="pre">db</span></code> file necessary for the SBOAnn script by Elisabeth Fritze as well as the modules <cite>gapfill</cite>, <cite>growth</cite> and <cite>modelseed</cite>.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">setup.py</span></code> and <code class="docutils literal notranslate"><span class="pre">pyproject.toml</span></code> enable creating a PyPi package called <code class="docutils literal notranslate"><span class="pre">refinegems</span></code>.</p></li>
</ul>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="examples.html" class="btn btn-neutral float-right" title="Examples on how to use refinegems" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Famke Bäuerle.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>