<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Filling gaps with refineGEMs &mdash; refinegems 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom_theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/refineGEMs_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="application.html">Usage as standalone application</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage as python module</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline.html">Pipeline: From genome sequence to draft model</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">refinegems package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">refinegems</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Filling gaps with refineGEMs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/gapfill.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="filling-gaps-with-refinegems">
<h1>Filling gaps with refineGEMs<a class="headerlink" href="#filling-gaps-with-refinegems" title="Permalink to this heading"></a></h1>
<p>There are two possibilities to use refineGEMs to fill gaps.</p>
<section id="manual-gap-filling">
<h2>Manual gap filling<a class="headerlink" href="#manual-gap-filling" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">refinegems.curate.</span></span><span class="sig-name descname"><span class="pre">add_reactions_from_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span></em><span class="sig-paren">)</span></dt>
<dd><dl class="simple">
<dt>adds all reactions with their info given in the table to the given model</dt><dd><p>(wrapper function to use with table given in the data folder)</p>
</dd>
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libSBML
table (pd-DataFrame): manual_curation tabel loaded with pandas, sheet gapfill
email (string): User Email to access the NCBI Entrez database</p>
</dd>
<dt>Returns:</dt><dd><p>model: modified model with new reactions</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="automated-gap-filling">
<h2>Automated gap filling<a class="headerlink" href="#automated-gap-filling" title="Permalink to this heading"></a></h2>
<p>The <cite>gapfill</cite> module was created to enable an automatic way of filling gaps in a model via genes.
This module can be used either with KEGG were you only need the KEGG organism ID or with BioCyc or with both (Options: ‘KEGG’, ‘BioCyc’, ‘KEGG+BioCyc’).</p>
<blockquote>
<div><div class="admonition warning">
<p class="admonition-title">Warning</p>
<dl class="simple">
<dt>Current restrictions:</dt><dd><ul class="simple">
<li><p>Only bacteria supported as missing reactions are filtered by compartments (‘c’, ‘e’, ‘p’).</p></li>
<li><p>Only models where the organisms have entries in KEGG and/or BioCyc can be gap filled.</p></li>
</ul>
</dd>
</dl>
</div>
<ol class="arabic">
<li><p>For the gap filling using BioCyc you will need to create an  <cite>account &lt;https://biocyc.org/&gt;</cite>.</p></li>
<li><p>Then you will search for the strain of your organism.</p></li>
<li><p>Then you click on <cite>Tools</cite> and select <cite>Special SmartTables</cite>.
There you can download the tables “All genes of &lt;organism&gt;” and “All reactions of &lt;organism&gt;”.</p></li>
<li><p>For the gene to reaction mapping table:</p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Remove all columns except ‘Gene Name’,</p></li>
<li><p>then <cite>choose a transform</cite> (‘Reactions of gene’),</p></li>
<li><p>then add <cite>property</cite> (‘Accession-2’)</p></li>
<li><p>and after that delete the ‘Gene Name’ column.</p></li>
<li><p>Afterwards choose ‘Accession-2’,</p></li>
<li><p>and use the filter function to delete all empty rows.</p></li>
<li><p>Then <cite>Export to Spreadsheet File</cite> and choose <cite>frame ids</cite>.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>For the reactions table:</p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Remove all columns except ‘Reaction’,</p></li>
<li><p>then <cite>choose a transform</cite>:</p>
<ol class="loweralpha simple">
<li><p>‘Reactants of reaction’</p></li>
<li><p>‘Products of reaction’</p></li>
</ol>
</li>
<li><p>and then <cite>choose property</cite>:</p></li>
</ol>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>‘EC-Number’</p></li>
<li><p>‘Reaction-Direction’</p></li>
<li><p>‘Spontaneous?’ (property)</p></li>
</ol>
</div></blockquote>
<ol class="lowerroman simple" start="4">
<li><p>Afterwards <cite>Export to Spreadsheet File</cite> and choose <cite>frame ids</cite>.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>For the metabolites table:</p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Use MetaCyc database -&gt; all compounds</p></li>
<li><p>Remove all columns except ‘Compound’,</p></li>
<li><p>then choose <cite>property</cite>:</p></li>
</ol>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>‘Object ID’</p></li>
<li><p>‘Chemical Formula’</p></li>
<li><p>‘InChI-Key’</p></li>
<li><p>database links -&gt; ‘ChEBI’</p></li>
</ol>
</div></blockquote>
<ol class="lowerroman simple" start="4">
<li><p>Afterwards <cite>Export to Spreadsheet File</cite> and choose <cite>common names</cite>.</p></li>
</ol>
</div></blockquote>
</li>
</ol>
<p>Run times:</p>
<blockquote>
<div><p>‘KEGG’: ~ 2h
‘BioCyc’: ~ 45mins - 1h
‘KEGG+BioCyc’: ~ 3 - 4h</p>
</div></blockquote>
</div></blockquote>
<p>You can either use it with refineGEMs, just edit the config …</p>
<p>or you can use it in your local script with</p>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Famke Bäuerle.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>