<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>refinegems package &mdash; refinegems 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom_theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Ideas behind some implementations" href="explanation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/refineGEMs_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="application.html">Usage as standalone application</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples on how to use refinegems</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage as python module</a></li>
<li class="toctree-l1"><a class="reference internal" href="explanation.html">Ideas behind some implementations</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">refinegems package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.charges">refinegems.charges module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.charges.correct_charges_from_db"><code class="docutils literal notranslate"><span class="pre">correct_charges_from_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.charges.correct_charges_modelseed"><code class="docutils literal notranslate"><span class="pre">correct_charges_modelseed()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.comparison">refinegems.comparison module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.get_sbo_mapping_multiple"><code class="docutils literal notranslate"><span class="pre">get_sbo_mapping_multiple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.plot_heatmap_binary"><code class="docutils literal notranslate"><span class="pre">plot_heatmap_binary()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.plot_heatmap_dt"><code class="docutils literal notranslate"><span class="pre">plot_heatmap_dt()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.plot_initial_analysis"><code class="docutils literal notranslate"><span class="pre">plot_initial_analysis()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.plot_rea_sbo_multiple"><code class="docutils literal notranslate"><span class="pre">plot_rea_sbo_multiple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.plot_venn"><code class="docutils literal notranslate"><span class="pre">plot_venn()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.comparison.simulate_all"><code class="docutils literal notranslate"><span class="pre">simulate_all()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.curate">refinegems.curate module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.curate.add_reactions_from_table"><code class="docutils literal notranslate"><span class="pre">add_reactions_from_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.curate.update_annotations_from_others"><code class="docutils literal notranslate"><span class="pre">update_annotations_from_others()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.curate.update_annotations_from_table"><code class="docutils literal notranslate"><span class="pre">update_annotations_from_table()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.cvterms">refinegems.cvterms module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_genes"><code class="docutils literal notranslate"><span class="pre">add_cv_term_genes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_metabolites"><code class="docutils literal notranslate"><span class="pre">add_cv_term_metabolites()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_pathways"><code class="docutils literal notranslate"><span class="pre">add_cv_term_pathways()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_pathways_to_entity"><code class="docutils literal notranslate"><span class="pre">add_cv_term_pathways_to_entity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_reactions"><code class="docutils literal notranslate"><span class="pre">add_cv_term_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.add_cv_term_units"><code class="docutils literal notranslate"><span class="pre">add_cv_term_units()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.generate_cvterm"><code class="docutils literal notranslate"><span class="pre">generate_cvterm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.get_id_from_cv_term"><code class="docutils literal notranslate"><span class="pre">get_id_from_cv_term()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.cvterms.print_cvterm"><code class="docutils literal notranslate"><span class="pre">print_cvterm()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.growth">refinegems.growth module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.find_additives"><code class="docutils literal notranslate"><span class="pre">find_additives()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.find_minimum_essential"><code class="docutils literal notranslate"><span class="pre">find_minimum_essential()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.find_missing_essential"><code class="docutils literal notranslate"><span class="pre">find_missing_essential()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_all_minimum_essential"><code class="docutils literal notranslate"><span class="pre">get_all_minimum_essential()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_default_secretion"><code class="docutils literal notranslate"><span class="pre">get_default_secretion()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_default_uptake"><code class="docutils literal notranslate"><span class="pre">get_default_uptake()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_essential_reactions"><code class="docutils literal notranslate"><span class="pre">get_essential_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_essential_reactions_via_bounds"><code class="docutils literal notranslate"><span class="pre">get_essential_reactions_via_bounds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_growth_selected_media"><code class="docutils literal notranslate"><span class="pre">get_growth_selected_media()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_minimal_uptake"><code class="docutils literal notranslate"><span class="pre">get_minimal_uptake()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.get_missing_exchanges"><code class="docutils literal notranslate"><span class="pre">get_missing_exchanges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.growth_one_medium_from_default"><code class="docutils literal notranslate"><span class="pre">growth_one_medium_from_default()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.growth_one_medium_from_minimal"><code class="docutils literal notranslate"><span class="pre">growth_one_medium_from_minimal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.modify_medium"><code class="docutils literal notranslate"><span class="pre">modify_medium()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.growth.simulate_minimum_essential"><code class="docutils literal notranslate"><span class="pre">simulate_minimum_essential()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.investigate">refinegems.investigate module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_egc"><code class="docutils literal notranslate"><span class="pre">get_egc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_mass_charge_unbalanced"><code class="docutils literal notranslate"><span class="pre">get_mass_charge_unbalanced()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_metabs_with_one_cvterm"><code class="docutils literal notranslate"><span class="pre">get_metabs_with_one_cvterm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_model_info"><code class="docutils literal notranslate"><span class="pre">get_model_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_orphans_deadends_disconnected"><code class="docutils literal notranslate"><span class="pre">get_orphans_deadends_disconnected()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.get_reactions_per_sbo"><code class="docutils literal notranslate"><span class="pre">get_reactions_per_sbo()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.initial_analysis"><code class="docutils literal notranslate"><span class="pre">initial_analysis()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.parse_reaction"><code class="docutils literal notranslate"><span class="pre">parse_reaction()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.plot_rea_sbo_single"><code class="docutils literal notranslate"><span class="pre">plot_rea_sbo_single()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.run_memote"><code class="docutils literal notranslate"><span class="pre">run_memote()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.investigate.run_memote_sys"><code class="docutils literal notranslate"><span class="pre">run_memote_sys()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.io">refinegems.io module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_a_table_from_database"><code class="docutils literal notranslate"><span class="pre">load_a_table_from_database()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_all_media_from_db"><code class="docutils literal notranslate"><span class="pre">load_all_media_from_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_document_libsbml"><code class="docutils literal notranslate"><span class="pre">load_document_libsbml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_manual_annotations"><code class="docutils literal notranslate"><span class="pre">load_manual_annotations()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_manual_gapfill"><code class="docutils literal notranslate"><span class="pre">load_manual_gapfill()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_medium_custom"><code class="docutils literal notranslate"><span class="pre">load_medium_custom()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_medium_from_db"><code class="docutils literal notranslate"><span class="pre">load_medium_from_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_model_cobra"><code class="docutils literal notranslate"><span class="pre">load_model_cobra()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_model_libsbml"><code class="docutils literal notranslate"><span class="pre">load_model_libsbml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.load_multiple_models"><code class="docutils literal notranslate"><span class="pre">load_multiple_models()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.parse_fasta_headers"><code class="docutils literal notranslate"><span class="pre">parse_fasta_headers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.parse_gff_for_gp_info"><code class="docutils literal notranslate"><span class="pre">parse_gff_for_gp_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.save_user_input"><code class="docutils literal notranslate"><span class="pre">save_user_input()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.search_ncbi_for_gpr"><code class="docutils literal notranslate"><span class="pre">search_ncbi_for_gpr()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.search_sbo_label"><code class="docutils literal notranslate"><span class="pre">search_sbo_label()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.validate_libsbml_model"><code class="docutils literal notranslate"><span class="pre">validate_libsbml_model()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.write_report"><code class="docutils literal notranslate"><span class="pre">write_report()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.io.write_to_file"><code class="docutils literal notranslate"><span class="pre">write_to_file()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.modelseed">refinegems.modelseed module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.compare_model_modelseed"><code class="docutils literal notranslate"><span class="pre">compare_model_modelseed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.compare_to_modelseed"><code class="docutils literal notranslate"><span class="pre">compare_to_modelseed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_charge_mismatch"><code class="docutils literal notranslate"><span class="pre">get_charge_mismatch()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_compared_formulae"><code class="docutils literal notranslate"><span class="pre">get_compared_formulae()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_formula_mismatch"><code class="docutils literal notranslate"><span class="pre">get_formula_mismatch()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_model_charges"><code class="docutils literal notranslate"><span class="pre">get_model_charges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_modelseed_charges"><code class="docutils literal notranslate"><span class="pre">get_modelseed_charges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.modelseed.get_modelseed_compounds"><code class="docutils literal notranslate"><span class="pre">get_modelseed_compounds()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.pathways">refinegems.pathways module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.add_kegg_pathways"><code class="docutils literal notranslate"><span class="pre">add_kegg_pathways()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.create_pathway_groups"><code class="docutils literal notranslate"><span class="pre">create_pathway_groups()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.extract_kegg_pathways"><code class="docutils literal notranslate"><span class="pre">extract_kegg_pathways()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.extract_kegg_reactions"><code class="docutils literal notranslate"><span class="pre">extract_kegg_reactions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.get_pathway_groups"><code class="docutils literal notranslate"><span class="pre">get_pathway_groups()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.kegg_pathways"><code class="docutils literal notranslate"><span class="pre">kegg_pathways()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.pathways.load_model_enable_groups"><code class="docutils literal notranslate"><span class="pre">load_model_enable_groups()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.polish">refinegems.polish module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.add_compartment_structure_specs"><code class="docutils literal notranslate"><span class="pre">add_compartment_structure_specs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.add_curie_set"><code class="docutils literal notranslate"><span class="pre">add_curie_set()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.add_fba_units"><code class="docutils literal notranslate"><span class="pre">add_fba_units()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.add_metab"><code class="docutils literal notranslate"><span class="pre">add_metab()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.add_reac"><code class="docutils literal notranslate"><span class="pre">add_reac()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.change_all_qualifiers"><code class="docutils literal notranslate"><span class="pre">change_all_qualifiers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.change_qualifier_per_entity"><code class="docutils literal notranslate"><span class="pre">change_qualifier_per_entity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.change_qualifiers"><code class="docutils literal notranslate"><span class="pre">change_qualifiers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.create_fba_units"><code class="docutils literal notranslate"><span class="pre">create_fba_units()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.create_unit"><code class="docutils literal notranslate"><span class="pre">create_unit()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.create_unit_definition"><code class="docutils literal notranslate"><span class="pre">create_unit_definition()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.cv_ncbiprotein"><code class="docutils literal notranslate"><span class="pre">cv_ncbiprotein()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.cv_notes_metab"><code class="docutils literal notranslate"><span class="pre">cv_notes_metab()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.cv_notes_reac"><code class="docutils literal notranslate"><span class="pre">cv_notes_reac()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.generate_new_curie_set"><code class="docutils literal notranslate"><span class="pre">generate_new_curie_set()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.get_set_of_curies"><code class="docutils literal notranslate"><span class="pre">get_set_of_curies()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.improve_curie_per_entity"><code class="docutils literal notranslate"><span class="pre">improve_curie_per_entity()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.improve_curies"><code class="docutils literal notranslate"><span class="pre">improve_curies()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.polish"><code class="docutils literal notranslate"><span class="pre">polish()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.polish_annotations"><code class="docutils literal notranslate"><span class="pre">polish_annotations()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.polish_entities"><code class="docutils literal notranslate"><span class="pre">polish_entities()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.print_UnitDefinitions"><code class="docutils literal notranslate"><span class="pre">print_UnitDefinitions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.print_remaining_UnitDefinitions"><code class="docutils literal notranslate"><span class="pre">print_remaining_UnitDefinitions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.set_default_units"><code class="docutils literal notranslate"><span class="pre">set_default_units()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.set_initial_amount"><code class="docutils literal notranslate"><span class="pre">set_initial_amount()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.polish.set_units"><code class="docutils literal notranslate"><span class="pre">set_units()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-refinegems.sboann">refinegems.sboann module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforCompartments"><code class="docutils literal notranslate"><span class="pre">addSBOforCompartments()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforGenes"><code class="docutils literal notranslate"><span class="pre">addSBOforGenes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforGroups"><code class="docutils literal notranslate"><span class="pre">addSBOforGroups()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforMetabolites"><code class="docutils literal notranslate"><span class="pre">addSBOforMetabolites()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforModel"><code class="docutils literal notranslate"><span class="pre">addSBOforModel()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOforParameters"><code class="docutils literal notranslate"><span class="pre">addSBOforParameters()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOfromDB"><code class="docutils literal notranslate"><span class="pre">addSBOfromDB()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.addSBOviaEC"><code class="docutils literal notranslate"><span class="pre">addSBOviaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkAcetylationViaEC"><code class="docutils literal notranslate"><span class="pre">checkAcetylationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkActiveTransport"><code class="docutils literal notranslate"><span class="pre">checkActiveTransport()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkBiomass"><code class="docutils literal notranslate"><span class="pre">checkBiomass()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkCoTransport"><code class="docutils literal notranslate"><span class="pre">checkCoTransport()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDeamination"><code class="docutils literal notranslate"><span class="pre">checkDeamination()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDeaminationViaEC"><code class="docutils literal notranslate"><span class="pre">checkDeaminationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDecarbonylation"><code class="docutils literal notranslate"><span class="pre">checkDecarbonylation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDecarboxylation"><code class="docutils literal notranslate"><span class="pre">checkDecarboxylation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDecarboxylationViaEC"><code class="docutils literal notranslate"><span class="pre">checkDecarboxylationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkDemand"><code class="docutils literal notranslate"><span class="pre">checkDemand()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkExchange"><code class="docutils literal notranslate"><span class="pre">checkExchange()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkGlycosylation"><code class="docutils literal notranslate"><span class="pre">checkGlycosylation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkGlycosylationViaEC"><code class="docutils literal notranslate"><span class="pre">checkGlycosylationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkHydrolysisViaEC"><code class="docutils literal notranslate"><span class="pre">checkHydrolysisViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkIsomerisationViaEC"><code class="docutils literal notranslate"><span class="pre">checkIsomerisationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkMethylationViaEC"><code class="docutils literal notranslate"><span class="pre">checkMethylationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkPassiveTransport"><code class="docutils literal notranslate"><span class="pre">checkPassiveTransport()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkPhosphorylation"><code class="docutils literal notranslate"><span class="pre">checkPhosphorylation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkRedox"><code class="docutils literal notranslate"><span class="pre">checkRedox()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkRedoxViaEC"><code class="docutils literal notranslate"><span class="pre">checkRedoxViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkSink"><code class="docutils literal notranslate"><span class="pre">checkSink()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.checkTransaminationViaEC"><code class="docutils literal notranslate"><span class="pre">checkTransaminationViaEC()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentDict"><code class="docutils literal notranslate"><span class="pre">getCompartmentDict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentFromSpeciesRef"><code class="docutils literal notranslate"><span class="pre">getCompartmentFromSpeciesRef()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentList"><code class="docutils literal notranslate"><span class="pre">getCompartmentList()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentlessMetaboliteIds"><code class="docutils literal notranslate"><span class="pre">getCompartmentlessMetaboliteIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentlessProductIds"><code class="docutils literal notranslate"><span class="pre">getCompartmentlessProductIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentlessReactantIds"><code class="docutils literal notranslate"><span class="pre">getCompartmentlessReactantIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getCompartmentlessSpeciesId"><code class="docutils literal notranslate"><span class="pre">getCompartmentlessSpeciesId()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getECNums"><code class="docutils literal notranslate"><span class="pre">getECNums()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getListOfMetabolites"><code class="docutils literal notranslate"><span class="pre">getListOfMetabolites()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getMetaboliteIds"><code class="docutils literal notranslate"><span class="pre">getMetaboliteIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getProductCompartmentList"><code class="docutils literal notranslate"><span class="pre">getProductCompartmentList()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getProductIds"><code class="docutils literal notranslate"><span class="pre">getProductIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getReactantCompartmentList"><code class="docutils literal notranslate"><span class="pre">getReactantCompartmentList()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.getReactantIds"><code class="docutils literal notranslate"><span class="pre">getReactantIds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.hasReactantPair"><code class="docutils literal notranslate"><span class="pre">hasReactantPair()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.isProtonTransport"><code class="docutils literal notranslate"><span class="pre">isProtonTransport()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.moreThanTwoCompartmentTransport"><code class="docutils literal notranslate"><span class="pre">moreThanTwoCompartmentTransport()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.returnCompartment"><code class="docutils literal notranslate"><span class="pre">returnCompartment()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.sbo_annotation"><code class="docutils literal notranslate"><span class="pre">sbo_annotation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.sbo_annotation_write"><code class="docutils literal notranslate"><span class="pre">sbo_annotation_write()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.soleProtonTransported"><code class="docutils literal notranslate"><span class="pre">soleProtonTransported()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.splitSymAntiPorter"><code class="docutils literal notranslate"><span class="pre">splitSymAntiPorter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#refinegems.sboann.splitTransportBiochem"><code class="docutils literal notranslate"><span class="pre">splitTransportBiochem()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">refinegems</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">refinegems package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/modules.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="refinegems-package">
<h1>refinegems package<a class="headerlink" href="#refinegems-package" title="Permalink to this heading"></a></h1>
<p>Here is an overview on all functions. All imports are mocked in <code class="docutils literal notranslate"><span class="pre">autodoc_mock_imports</span></code> in the <code class="docutils literal notranslate"><span class="pre">conf.py</span></code> file to enable automatic building.</p>
<section id="module-refinegems.charges">
<span id="refinegems-charges-module"></span><h2>refinegems.charges module<a class="headerlink" href="#module-refinegems.charges" title="Permalink to this heading"></a></h2>
<p>Provides functions for adding charges to metabolites</p>
<p>When iterating through all metabolites present in a model, you will find several which have no defined charge (metab.getPlugin(‘fbc’).isSetCharge() = false). This can lead to charge imbalanced reactions. This script takes information on metabolite charges from the ModelSEED database. A charge is automatically added to a metabolite if it has no defined charge and if there is only one charge denoted in ModelSEED. When multiple charges are present, the metabolite and the possible charges are noted and later returned in a dictionary.</p>
<p>It is possible to use the correct_charges_from_db function with other databases. The user just needs to make sure that the compounds dataframe has a ‘BiGG’ and a ‘charge’ column.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.charges.correct_charges_from_db">
<span class="sig-prename descclassname"><span class="pre">refinegems.charges.</span></span><span class="sig-name descname"><span class="pre">correct_charges_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compounds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.charges.correct_charges_from_db" title="Permalink to this definition"></a></dt>
<dd><p>Adds charges taken from given database to metabolites which have no defined charge</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
compounds (df): containing database data with ‘BiGG’ (BiGG-Ids) and ‘charge’ (float or int) as columns</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (model with added charges, metabolites with multiple charges as pd df)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.charges.correct_charges_modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.charges.</span></span><span class="sig-name descname"><span class="pre">correct_charges_modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charge_report_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.charges.correct_charges_modelseed" title="Permalink to this definition"></a></dt>
<dd><p>wrapper function which completes the steps to charge correction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
new_file_path (Str): filepath + name for modified model</p>
</dd>
<dt>Returns:</dt><dd><p>dict: BiGG Id and possible charges of metabolites</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.comparison">
<span id="refinegems-comparison-module"></span><h2>refinegems.comparison module<a class="headerlink" href="#module-refinegems.comparison" title="Permalink to this heading"></a></h2>
<p>Provides functions to compare and visualize multiple models</p>
<p>Can mainly be used to compare growth behaviour of multiple models. All other stats are shown in the memote report.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.get_sbo_mapping_multiple">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">get_sbo_mapping_multiple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Model</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.comparison.get_sbo_mapping_multiple" title="Permalink to this definition"></a></dt>
<dd><p>Determines number of reactions per SBO Term and adss label of SBO Terms</p>
<dl class="simple">
<dt>Args:</dt><dd><p>models (list[libModel]): Models loaded with libSBML</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: SBO Terms, no of reactions per model and SBO Label</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.plot_heatmap_binary">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">plot_heatmap_binary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">growth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.comparison.plot_heatmap_binary" title="Permalink to this definition"></a></dt>
<dd><p>Creates a plot were if growth without additives is possible is marked blue otherwise red</p>
<dl class="simple">
<dt>Args:</dt><dd><p>growth (pd.DataFrame): Containing growth data from simulate_all</p>
</dd>
<dt>Returns:</dt><dd><p>plot: sns heatmap plot</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.plot_heatmap_dt">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">plot_heatmap_dt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">growth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.comparison.plot_heatmap_dt" title="Permalink to this definition"></a></dt>
<dd><p>Creates heatmap of simulated doubling times with additives</p>
<dl class="simple">
<dt>Args:</dt><dd><p>growth (pd.DataFrame): Containing growth data from simulate_all</p>
</dd>
<dt>Returns:</dt><dd><p>plot: sns heatmap plot</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.plot_initial_analysis">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">plot_initial_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Model</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.comparison.plot_initial_analysis" title="Permalink to this definition"></a></dt>
<dd><p>Creates bar plot of number of entities per Model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>models (list[libModel]): Models loaded with libSBML</p>
</dd>
<dt>Returns:</dt><dd><p>plot: pandas plot object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.plot_rea_sbo_multiple">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">plot_rea_sbo_multiple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Model</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.comparison.plot_rea_sbo_multiple" title="Permalink to this definition"></a></dt>
<dd><p>Plots reactions per SBO Term in horizontal bar chart with stacked bars for the models</p>
<dl class="simple">
<dt>Args:</dt><dd><p>models (list[libModel]): Models loaded with libSBML
rename (dict, optional): Rename model ids to custom names. Defaults to None.</p>
</dd>
<dt>Returns:</dt><dd><p>plot: pandas plot object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.plot_venn">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">plot_venn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">cobra.Model</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.comparison.plot_venn" title="Permalink to this definition"></a></dt>
<dd><p>Creates venn diagram to show the overlap of model entities</p>
<dl class="simple">
<dt>Args:</dt><dd><p>models (list[cobraModel]): Models loaded with cobrapy
entity (str): Compare on metabolite|reaction
perc (bool, optional): True if percentages should be used. Defaults to False.</p>
</dd>
<dt>Returns:</dt><dd><p>plot: venn diagram</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.comparison.simulate_all">
<span class="sig-prename descclassname"><span class="pre">refinegems.comparison.</span></span><span class="sig-name descname"><span class="pre">simulate_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">cobra.Model</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.comparison.simulate_all" title="Permalink to this definition"></a></dt>
<dd><p>Does a run of growth simulation for multiple models on different media</p>
<dl class="simple">
<dt>Args:</dt><dd><p>models (list[cobraModel]): Models loaded with cobrapy
mediumpath (string): Path to csv containing medium definitions
media (list): Media of interest (f.ex. LB, M9, …)
basis (string): Either default_uptake (adding metabs from default) or minimal_uptake (adding metabs from minimal medium)</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: table containing the results of the growth simulation</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.curate">
<span id="refinegems-curate-module"></span><h2>refinegems.curate module<a class="headerlink" href="#module-refinegems.curate" title="Permalink to this heading"></a></h2>
<p>Functions to enable annotation of entities using a manual curated table</p>
<p>While working on GEMs the user might come across ill-annotated or missing metabolites, reactions and genes. This module aims to enable faster manual curation by  allowing to edit an excel table directly which is used to update the given model. This module makes use of the cvterms module aswell.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.curate.add_reactions_from_table">
<span class="sig-prename descclassname"><span class="pre">refinegems.curate.</span></span><span class="sig-name descname"><span class="pre">add_reactions_from_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.curate.add_reactions_from_table" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>adds all reactions with their info given in the table to the given model</dt><dd><p>(wrapper function to use with table given in the data folder)</p>
</dd>
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libSBML
table (pd-DataFrame): manual_curation tabel loaded with pandas, sheet gapfill
email (string): User Email to access the NCBI Entrez database</p>
</dd>
<dt>Returns:</dt><dd><p>model: modified model with new reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.curate.update_annotations_from_others">
<span class="sig-prename descclassname"><span class="pre">refinegems.curate.</span></span><span class="sig-name descname"><span class="pre">update_annotations_from_others</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.curate.update_annotations_from_others" title="Permalink to this definition"></a></dt>
<dd><p>Synchronizes metabolite annotations for core, periplasm and extracelullar</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libSBML</p>
</dd>
<dt>Returns:</dt><dd><p>model: modified with synchronized annotations</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.curate.update_annotations_from_table">
<span class="sig-prename descclassname"><span class="pre">refinegems.curate.</span></span><span class="sig-name descname"><span class="pre">update_annotations_from_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.curate.update_annotations_from_table" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>updates annotation of metabolites given in the table</dt><dd><p>(wrapper function to use with table given in the data folder)</p>
</dd>
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libSBML
table (pd-DataFrame): manual_curation tabel loaded with pandas, sheet metabs</p>
</dd>
<dt>Returns:</dt><dd><p>model: modified model with new annotations</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.cvterms">
<span id="refinegems-cvterms-module"></span><h2>refinegems.cvterms module<a class="headerlink" href="#module-refinegems.cvterms" title="Permalink to this heading"></a></h2>
<p>Helper module to work with annotations (CVTerms)</p>
<p>Stores dictionaries which hold information the identifiers.org syntax, has functions to add CVTerms to different entities and parse CVTerms.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_genes">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lab_strain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_genes" title="Permalink to this definition"></a></dt>
<dd><p>Add CVTerm to a gene</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
gene (libsbml-gene): gene to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_metabolites">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metab</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_metabolites" title="Permalink to this definition"></a></dt>
<dd><p>Add CVTerm to a metabolite</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
metab (libsbml-species): metabolite to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_pathways" title="Permalink to this definition"></a></dt>
<dd><p>Add CVTerm to a groups pathway</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
entity (libsbml-group): pathway to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_pathways_to_entity">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_pathways_to_entity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_pathways_to_entity" title="Permalink to this definition"></a></dt>
<dd><p>Add CVTerm to a entity as OCCURS IN pathway</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
entity (libsbml-group): entity to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_reactions" title="Permalink to this definition"></a></dt>
<dd><p>Add CVTerm to a reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
reac (libsbml-reaction): reaction to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">unit_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Unit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_units" title="Permalink to this definition"></a></dt>
<dd><p>Adds CVTerm to a unit</p>
<dl class="simple">
<dt>Params:</dt><dd><p>unit_id (string):    ID to add as URI to annotation
unit (libSBML Unit): Unit to add CVTerm to
relation (str):      Provides model qualifier to be added</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.generate_cvterm">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">generate_cvterm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_m_qt</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.CVTerm</span></span></span><a class="headerlink" href="#refinegems.cvterms.generate_cvterm" title="Permalink to this definition"></a></dt>
<dd><p>Generates a CVTerm with the provided qualifier &amp;
biological or model qualifier types</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>qt: libSBML qualifier type</p></li>
<li><p>b_m_qt: libSBML biological or model qualifier</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>-&gt; A libSBML CVTerm with the provided qualifier &amp; biological or model qualifier types</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.get_id_from_cv_term">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">get_id_from_cv_term</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.get_id_from_cv_term" title="Permalink to this definition"></a></dt>
<dd><p>extract id for a specific database from CVTerm</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entity (libsbml Object): Species, Reaction, Gene
db_id (string): database of interest</p>
</dd>
<dt>Returns:</dt><dd><p>list: ids of entity belonging to db_id</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.print_cvterm">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">print_cvterm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cvterm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.CVTerm</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.print_cvterm" title="Permalink to this definition"></a></dt>
<dd><p>Debug function: 
Prints the URIs contained in the provided CVTerm 
along with the provided qualifier &amp; biological/model qualifier types</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>cvterm (CVTerm): A libSBML CVTerm</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.growth">
<span id="refinegems-growth-module"></span><h2>refinegems.growth module<a class="headerlink" href="#module-refinegems.growth" title="Permalink to this heading"></a></h2>
<p>Provides functions to simulate growth on any medium</p>
<p>Tailored to work with the Synthetic Nasal Medium (SNM), should
work with any medium as long as its defined in a csv with ; as
delimiter and BiGG Ids for the compounds.</p>
<p>Outputs a table where</p>
<p>essential = metabolites not present in the medium but necessary for growth</p>
<p>missing = all exchanges missing in the model but given in medium</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.find_additives">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">find_additives</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.find_additives" title="Permalink to this definition"></a></dt>
<dd><p>Iterates through all exchanges to find metabolites that lead to a higher growth rate compared to the growth rate yielded on the base_medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
base_medium (dict): exchanges as keys and their flux bound as value (f.ex {‘EX_glc__D_e’ : 10.0})</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: exchanges sorted from highest to lowest growth rate improvement</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.find_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">find_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">essential</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.find_minimum_essential" title="Permalink to this definition"></a></dt>
<dd><p>Report metabolites necessary for growth and not in custom medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>medium (df): dataframe of medium
essential (list): echanges of metabolites which lead to zero growth if blocked</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites not present in the medium but necessary for growth</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.find_missing_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">find_missing_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_uptake</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.find_missing_essential" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Report which exchange reactions are needed for growth</dt><dd><p>combines default uptake and valid new medium</p>
</dd>
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (dict): custom medium with exchanges that are present in the model
default_uptake (list): metabolites consumed in standard medium</p>
</dd>
<dt>Returns:</dt><dd><p>list: all metabolites which lead to zero growth if blocked</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_all_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_all_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_all_minimum_essential" title="Permalink to this definition"></a></dt>
<dd><p>Returns metabolites necessary for growth and not in media</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
media (list): a list containing the names of all media for which the growth essential metabolites not contained in the media should be returned</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on different media which metabs are missing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_default_secretion">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_default_secretion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_default_secretion" title="Permalink to this definition"></a></dt>
<dd><p>Checks fluxes after fba, if positive the metabolite is produced</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: BiGG Ids of produced metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_default_uptake">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_default_uptake</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_default_uptake" title="Permalink to this definition"></a></dt>
<dd><p>Determines which metabolites are used in the standard medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites consumed in standard medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_essential_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_essential_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_essential_reactions" title="Permalink to this definition"></a></dt>
<dd><p>Knocks out each reaction, if no growth is detected the reaction is seen as essential</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: BiGG Ids of essential reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_essential_reactions_via_bounds">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_essential_reactions_via_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_essential_reactions_via_bounds" title="Permalink to this definition"></a></dt>
<dd><p>Knocks out reactions by setting their bounds to 0, if no growth is detected the reaction is seen as essential</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: BiGG Ids of essential reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_growth_selected_media">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_growth_selected_media</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_growth_selected_media" title="Permalink to this definition"></a></dt>
<dd><p>Simulates growth on all selected media</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
media (list): media to simulate on (must be in csv)
basis (string): either default_uptake (adding metabs from default) or minimal_uptake (adding metabs from minimal medium)</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on growth behaviour on selected media</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_minimal_uptake">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_minimal_uptake</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_minimal_uptake" title="Permalink to this definition"></a></dt>
<dd><p>Determines which metabolites are used in a minimal medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites consumed in minimal medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_missing_exchanges">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_missing_exchanges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_missing_exchanges" title="Permalink to this definition"></a></dt>
<dd><p>Look for exchange reactions needed by the medium but not in the model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (df): dataframe of medium</p>
</dd>
<dt>Returns:</dt><dd><p>list: all exchanges missing in the model but given in medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.growth_one_medium_from_default">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">growth_one_medium_from_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.growth_one_medium_from_default" title="Permalink to this definition"></a></dt>
<dd><p>Simulates growth on given medium, adding missing metabolites from the default uptake</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (pandas-DataFrame): table containing metabolites present in the medium</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on growth behaviour on medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.growth_one_medium_from_minimal">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">growth_one_medium_from_minimal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.growth_one_medium_from_minimal" title="Permalink to this definition"></a></dt>
<dd><p>Simulates growth on given medium, adding missing metabolites from a minimal uptake</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (pandas-DataFrame): table containing metabolites present in the medium</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on growth behaviour on medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.modify_medium">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">modify_medium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_exchanges</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.modify_medium" title="Permalink to this definition"></a></dt>
<dd><p>Remove exchanges from medium that are not in the model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>medium (df): dataframe of medium
missing_exchanges (list): exchanges not in the model</p>
</dd>
<dt>Returns:</dt><dd><p>dict: growth medium that can be used with the model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.simulate_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">simulate_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">growth_medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.simulate_minimum_essential" title="Permalink to this definition"></a></dt>
<dd><p>Simulate growth with custom medium plus necessary uptakes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
growth_medium (dict): growth medium with exchanges present in the model
minimum (list): exchanges of metabolites not present in the medium but necessary for growth</p>
</dd>
<dt>Returns:</dt><dd><p>float: growth value</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.investigate">
<span id="refinegems-investigate-module"></span><h2>refinegems.investigate module<a class="headerlink" href="#module-refinegems.investigate" title="Permalink to this heading"></a></h2>
<p>Provides functions to investigate the model and test with memote</p>
<p>These functions enable simple testing of any model using memote and
access to its number of reactions, metabolites and genes.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_egc">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_egc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.investigate.get_egc" title="Permalink to this definition"></a></dt>
<dd><p>Energy-generating cycles represent thermodynamically infeasible states. Charging of energy metabolites without any energy source causes such cycles. Detection method is based on (Fritzemeier et al., 2017)</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>df: table with possible EGCs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_mass_charge_unbalanced">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_mass_charge_unbalanced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.get_mass_charge_unbalanced" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>creates lists of mass and charge unbalanced reactions,</dt><dd><p>without exchange reactions since they are unbalanced per definition</p>
</dd>
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (list of mass unbalanced, charge unbalanced reactions)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_metabs_with_one_cvterm">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_metabs_with_one_cvterm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="headerlink" href="#refinegems.investigate.get_metabs_with_one_cvterm" title="Permalink to this definition"></a></dt>
<dd><p>reports metabolites which have only one annotation, 
can be used as basis for further annotation research</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolite Ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_model_info">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_model_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.get_model_info" title="Permalink to this definition"></a></dt>
<dd><p>Reports core information of given model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (string): path to model file</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: overview on model parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_orphans_deadends_disconnected">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_orphans_deadends_disconnected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.get_orphans_deadends_disconnected" title="Permalink to this definition"></a></dt>
<dd><p>Uses memote functions to extract orphans, deadends and disconnected metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (list of orphans, deadends, disconnected metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.get_reactions_per_sbo">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">get_reactions_per_sbo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#refinegems.investigate.get_reactions_per_sbo" title="Permalink to this definition"></a></dt>
<dd><p>Counts number of reactions of all SBO Terms present</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libModel): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>dict: SBO Term as keys and number of reactions as values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.initial_analysis">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">initial_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.initial_analysis" title="Permalink to this definition"></a></dt>
<dd><p>extracts most important numbers of GEM</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (name of model, no of reactions, no of metabolites, no of genes)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.parse_reaction">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">parse_reaction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">eq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#refinegems.investigate.parse_reaction" title="Permalink to this definition"></a></dt>
<dd><p>Parses a reaction equation string to dictionary</p>
<dl class="simple">
<dt>Args:</dt><dd><p>eq (string): Equation of a reaction
model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>dict: metabolite ids as keys and their coefficients as values (negative = educts, positive = products)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.plot_rea_sbo_single">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">plot_rea_sbo_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.plot_rea_sbo_single" title="Permalink to this definition"></a></dt>
<dd><p>Plots reactions per SBO Term in horizontal bar chart</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libModel): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>plot: pandas plot object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.run_memote">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">run_memote</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#refinegems.investigate.run_memote" title="Permalink to this definition"></a></dt>
<dd><p>runs memote to obtain total score</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>float: memote score of model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.investigate.run_memote_sys">
<span class="sig-prename descclassname"><span class="pre">refinegems.investigate.</span></span><span class="sig-name descname"><span class="pre">run_memote_sys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cobra.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.investigate.run_memote_sys" title="Permalink to this definition"></a></dt>
<dd><p>run memote on linux machine</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.io">
<span id="refinegems-io-module"></span><h2>refinegems.io module<a class="headerlink" href="#module-refinegems.io" title="Permalink to this heading"></a></h2>
<p>Provides functions to load and write models, media definitions and the manual annotation table</p>
<p>Depending on the application the model needs to be loaded with cobra (memote)
or with libSBML (activation of groups). The media definitions are denoted in a csv within the data folder of this repository, thus the functions will only work if the user clones the repository. The manual_annotations table has to follow the specific layout given in the data folder in order to work with this module.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_a_table_from_database">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_a_table_from_database</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_name_or_query</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_a_table_from_database" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Loads the table for which the name is provided or a table containing all rows for which the query evaluates to </dt><dd><p>true from the refineGEMs database (‘data/database/data.db’)</p>
</dd>
<dt>Args:</dt><dd><p>table_name_or_query (str): Name of a table contained in the database ‘data.db’/ a SQL query</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: Containing the table for which the name was provided from the database ‘data.db’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_all_media_from_db">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_all_media_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_all_media_from_db" title="Permalink to this definition"></a></dt>
<dd><p>Helper function to extract media definitions from media_db.csv</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumpath (Str): path to csv file with medium database</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe of csv with metabs added as BiGG_EX exchange reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_document_libsbml">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_document_libsbml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.SBMLDocument</span></span></span><a class="headerlink" href="#refinegems.io.load_document_libsbml" title="Permalink to this definition"></a></dt>
<dd><p>loads model document using libsbml</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-document: loaded document by libsbml</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_manual_annotations">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_manual_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tablepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'data/manual_curation.xlsx'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'metab'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_manual_annotations" title="Permalink to this definition"></a></dt>
<dd><p>loads metabolite sheet from manual curation table</p>
<dl class="simple">
<dt>Args:</dt><dd><p>tablepath (str): Path to manual curation table. Defaults to ‘data/manual_curation.xlsx’.
sheet_name (str): Sheet name for metabolite annotations. Defaults to ‘metab’.</p>
</dd>
<dt>Returns:</dt><dd><p>df: table as pandas df</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_manual_gapfill">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_manual_gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tablepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'data/manual_curation.xlsx'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'gapfill'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_manual_gapfill" title="Permalink to this definition"></a></dt>
<dd><p>loads gapfill sheet from manual curation table</p>
<dl class="simple">
<dt>Args:</dt><dd><p>tablepath (str): Path to manual curation table. Defaults to ‘data/manual_curation.xlsx’.
sheet_name (str): Sheet name for reaction gapfilling. Defaults to ‘gapfill’.</p>
</dd>
<dt>Returns:</dt><dd><p>df: table as pandas df</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_medium_custom">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_medium_custom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_medium_custom" title="Permalink to this definition"></a></dt>
<dd><p>Helper function to read medium csv</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumpath (Str): path to csv file with medium</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe of csv</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_medium_from_db">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_medium_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.load_medium_from_db" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper function to extract subtable for the requested medium from the database ‘data.db’</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumname (Str): name of medium to test growth on</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe containing composition for one medium with metabs added as BiGG_EX exchange reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_model_cobra">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_model_cobra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">cobra.Model</span></span></span><a class="headerlink" href="#refinegems.io.load_model_cobra" title="Permalink to this definition"></a></dt>
<dd><p>loads model using cobrapy</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>cobra-model: loaded model by cobrapy</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_model_libsbml">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_model_libsbml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.Model</span></span></span><a class="headerlink" href="#refinegems.io.load_model_libsbml" title="Permalink to this definition"></a></dt>
<dd><p>loads model using libsbml</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: loaded model by libsbml</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.load_multiple_models">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">load_multiple_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">package</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="headerlink" href="#refinegems.io.load_multiple_models" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.parse_fasta_headers">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">parse_fasta_headers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_for_model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#refinegems.io.parse_fasta_headers" title="Permalink to this definition"></a></dt>
<dd><p>Parses FASTA file headers to obtain:
- the protein_id
- and the model_id (like it is obtained from CarveMe)
corresponding to the locus_tag</p>
<dl class="simple">
<dt>Args:</dt><dd><p>filepath (str): Path to FASTA file</p>
</dd>
<dt>Returns:</dt><dd><p>a pandas dataframe containing the columns locus_tag, Protein_id &amp; Model_id</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.parse_gff_for_gp_info">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">parse_gff_for_gp_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gff_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.parse_gff_for_gp_info" title="Permalink to this definition"></a></dt>
<dd><p>Parses gff file of organism to find gene protein reactions based on locus tags</p>
<dl class="simple">
<dt>Args:</dt><dd><p>gff_file (Str): path to gff file of organism of interest</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing mapping from locus tag to GPR</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.save_user_input">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">save_user_input</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">configpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.save_user_input" title="Permalink to this definition"></a></dt>
<dd><p>This aims to collect user input from the command line to create a config file, 
will also save the user input to a config if no config was given</p>
<dl class="simple">
<dt>Args:</dt><dd><p>configpath (str): path to config file if present</p>
</dd>
<dt>Returns:</dt><dd><p>dict: either loaded config file or created from user input</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.search_ncbi_for_gpr">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">search_ncbi_for_gpr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.search_ncbi_for_gpr" title="Permalink to this definition"></a></dt>
<dd><p>fetches protein name from NCBI</p>
<dl class="simple">
<dt>Args:</dt><dd><p>locus (string): NCBI compatible locus_tag</p>
</dd>
<dt>Returns:</dt><dd><p>str: protein name / description</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.search_sbo_label">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">search_sbo_label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sbo_number</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#refinegems.io.search_sbo_label" title="Permalink to this definition"></a></dt>
<dd><p>looks up the SBO label corresponding to a given SBO Term number</p>
<dl class="simple">
<dt>Args:</dt><dd><p>sbo_number (str): Last three digits of SBO-Term as str</p>
</dd>
<dt>Returns:</dt><dd><p>str: denoted label for given SBO Term</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.validate_libsbml_model">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">validate_libsbml_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.validate_libsbml_model" title="Permalink to this definition"></a></dt>
<dd><p>Debug method: Validates a libSBML model with the libSBML validator</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): A libSBML Model</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.write_report">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">write_report</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.write_report" title="Permalink to this definition"></a></dt>
<dd><p>Writes reports stored in dataframes to xlsx file</p>
<dl class="simple">
<dt>Args:</dt><dd><p>dataframe (pd.DataFrame): table containing output
filepath (str): path to file with filename</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.io.write_to_file">
<span class="sig-prename descclassname"><span class="pre">refinegems.io.</span></span><span class="sig-name descname"><span class="pre">write_to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.io.write_to_file" title="Permalink to this definition"></a></dt>
<dd><p>Writes modified model to new file</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
new_filename (Str): filename / path for modified model</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.modelseed">
<span id="refinegems-modelseed-module"></span><h2>refinegems.modelseed module<a class="headerlink" href="#module-refinegems.modelseed" title="Permalink to this heading"></a></h2>
<p>Reports mismatches in charges and formulae based on ModelSEED</p>
<p>Extracts modelseed data from a given tsv file, extracts all metabolites
from a given model. Both lists of metabolites are compared by charge and
formula.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.compare_model_modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">compare_model_modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_charges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modelseed_charges</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.compare_model_modelseed" title="Permalink to this definition"></a></dt>
<dd><p>compares tables with charges / formulae from model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_charges (df): charges and formulae of model metabolites
modelseed_charges (df): charges and formulae of modelseed metabolites</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing info whether charges / formulae match</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.compare_to_modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">compare_to_modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.compare_to_modelseed" title="Permalink to this definition"></a></dt>
<dd><p>Executes all steps to compare model metabolites to modelseed metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (table with charge mismatches, formula mismatches)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_charge_mismatch">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_charge_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_comp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_charge_mismatch" title="Permalink to this definition"></a></dt>
<dd><p>extracts metabolites with charge mismatch of model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>df_comp (df): charge and formula mismatches</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with charge mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_compared_formulae">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_compared_formulae</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">formula_mismatch</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_compared_formulae" title="Permalink to this definition"></a></dt>
<dd><p>compare formula by atom pattern</p>
<dl class="simple">
<dt>Args:</dt><dd><p>formula_mismatch (df): table with column containing atom comparison</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with formula mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_formula_mismatch">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_formula_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_comp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_formula_mismatch" title="Permalink to this definition"></a></dt>
<dd><p>extracts metabolites with formula mismatch of model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>df_comp (df): charge and formula mismatches</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with formula mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_model_charges">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_model_charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_model_charges" title="Permalink to this definition"></a></dt>
<dd><p>extracts all metabolites from model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing charges and formulae of model metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_modelseed_charges">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_modelseed_charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelseed_compounds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_modelseed_charges" title="Permalink to this definition"></a></dt>
<dd><p>extract table with BiGG, charges and formulae</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelseed_compounds (df): containing modelseed data</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing charges and formulae of modelseed metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_modelseed_compounds">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_modelseed_compounds</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_modelseed_compounds" title="Permalink to this definition"></a></dt>
<dd><p>extracts compounds from modelseed which have BiGG Ids</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>df: table containing modelseed data</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.pathways">
<span id="refinegems-pathways-module"></span><h2>refinegems.pathways module<a class="headerlink" href="#module-refinegems.pathways" title="Permalink to this heading"></a></h2>
<p>Provides functions for adding KEGG reactions as Group Pathways</p>
<p>If your organism occurs in the KEGG database, extract the KEGG reaction ID from the
annotations of your reactions and identify, in which KEGG pathways this reaction occurs. Add
all KEGG pathways for a reaction then as annotations with the biological qualifier ‘OCCURS_IN’
to the respective reaction.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.add_kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">add_kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_pathways</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.add_kegg_pathways" title="Permalink to this definition"></a></dt>
<dd><p>add KEGG reactions as BQB_OCCURS_IN</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
kegg_pathways (dict): reaction Id as key and Kegg Pathway Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: modified model with Kegg pathways</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.create_pathway_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">create_pathway_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway_groups</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.create_pathway_groups" title="Permalink to this definition"></a></dt>
<dd><p>use group module to add reactions to Kegg pathway</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml and containing Kegg pathways
pathway_groups (dict): Kegg Pathway Id as key and reactions Ids as values</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: modified model with groups for pathways</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.extract_kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">extract_kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_reactions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.extract_kegg_pathways" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>finds pathway for reactions in model with KEGG Ids</dt><dd><p>accesses KEGG API, uses tqdm to report progres to user</p>
</dd>
<dt>Args:</dt><dd><p>kegg_reactions (dict): reaction Id as key and Kegg Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>dict: reaction Id as key and Kegg Pathway Id as value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.extract_kegg_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">extract_kegg_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.extract_kegg_reactions" title="Permalink to this definition"></a></dt>
<dd><p>extract KEGG Ids from reactions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>(dict, list): reaction Id as key and Kegg Id as value, Ids of reactions without KEGG annotation</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.get_pathway_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">get_pathway_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_pathways</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.get_pathway_groups" title="Permalink to this definition"></a></dt>
<dd><p>group reaction into pathways</p>
<dl class="simple">
<dt>Args:</dt><dd><p>kegg_pathways (dict): reaction Id as key and Kegg Pathway Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>dict: Kegg Pathway Id as key and reactions Ids as values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.kegg_pathways" title="Permalink to this definition"></a></dt>
<dd><p>Executes all steps to add KEGG pathways as groups</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM
new_filename (Str): filename for modified model</p>
</dd>
<dt>Returns:</dt><dd><p>list: Ids of reactions without KEGG annotation</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.load_model_enable_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">load_model_enable_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.load_model_enable_groups" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>loads model as document using libsbml</dt><dd><p>enables groups extension</p>
</dd>
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-document: loaded document by libsbml</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.polish">
<span id="refinegems-polish-module"></span><h2>refinegems.polish module<a class="headerlink" href="#module-refinegems.polish" title="Permalink to this heading"></a></h2>
<p>Can be used to polish a model (created with CarveMe v.1.5.1)</p>
<p>The newer version of CarveMe leads to some irritations in the model, these scripts enable for example the addition of BiGG Ids to the annotations as well as a correct formatting of the annotations.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.add_compartment_structure_specs">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">add_compartment_structure_specs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.add_compartment_structure_specs" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Adds the required specifications for the compartment structure</dt><dd><p>if not set (size &amp; spatial dimension)</p>
</dd>
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.add_curie_set">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">add_curie_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity:</span> <span class="pre">libsbml.SBase</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_m_qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">curie_set:</span> <span class="pre">sortedcontainers.SortedSet.&lt;class</span> <span class="pre">'str'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.add_curie_set" title="Permalink to this definition"></a></dt>
<dd><p>Add a complete CURIE set to the provided CVTerm</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>entity (SBase):               A libSBML SBase object like model, GeneProduct, etc.</p></li>
<li><p>qt:                           A libSBML qualifier type: BIOLOGICAL_QUALIFIER|MODEL_QUALIFIER</p></li>
<li><p>b_m_qt:                       A libSBML biological or model qualifier type like BQB_IS|BQM_IS</p></li>
<li><p>curie_set (SortedSet[str]):   SortedSet containing CURIEs</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.add_fba_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">add_fba_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.add_fba_units" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Adds:</dt><dd><blockquote>
<div><ul class="simple">
<li><p>mmol per gDW per h</p></li>
<li><p>mmol per gDW</p></li>
<li><p>hour (h)</p></li>
<li><p>femto litre (fL)</p></li>
</ul>
</div></blockquote>
<p>to the list of unit definitions (needed for FBA)</p>
</dd>
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.add_metab">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">add_metab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Species</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.add_metab" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>adds the ID of metabolites as URI to the annotation field</dt><dd><dl class="simple">
<dt>For a VMH model, additionally, the corresponding BiGG IDs are added! </dt><dd><p>(Currently, only BiGG &amp; VMH IDs supported!)</p>
</dd>
</dl>
</dd>
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfSpecies</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.add_reac">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">add_reac</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Reaction</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.add_reac" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>adds the ID of reactions as URI to the annotation field</dt><dd><p>(Currently, only BiGG &amp; VMH IDs supported!)</p>
</dd>
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfReactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.change_all_qualifiers">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">change_all_qualifiers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lab_strain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.change_all_qualifiers" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper function to change qualifiers of all entities at once</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libsbml</p></li>
<li><p>lab_strain (bool): True if the strain was sequenced in a local lab</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>model: Model with all qualifiers updated to be MIRIAM compliant</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.change_qualifier_per_entity">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">change_qualifier_per_entity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.SBase</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_b_m_qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specific_db_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="headerlink" href="#refinegems.polish.change_qualifier_per_entity" title="Permalink to this definition"></a></dt>
<dd><p>Updates Qualifiers to be MIRIAM compliant for an entity</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>entity (SBase): A libSBML SBase object like model, GeneProduct, etc.</p></li>
<li><p>new_qt (Qualifier): A libSBML qualifier type: BIOLOGICAL_QUALIFIER|MODEL_QUALIFIER</p></li>
<li><p>new_b_m_qt (QualifierType): A libSBML biological or model qualifier type like BQB_IS|BQM_IS</p></li>
<li><p>specific_db_prefix (str): Has to be set if only for a specific database the qualifier type should be changed. Can be ‘kegg.genes’, ‘biocyc’, etc.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>list: CURIEs that are not MIRIAM compliant</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.change_qualifiers">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">change_qualifiers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_b_m_qt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specific_db_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.change_qualifiers" title="Permalink to this definition"></a></dt>
<dd><p>Updates Qualifiers to be MIRIAM compliant for an entity type of a given model</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (libsbml-model):   model loaded with libsbml</p></li>
<li><p>entity_type (str): model|compartment|metabolite|parameter|reaction|unit definition|unit|gene product|group</p></li>
<li><p>new_qt (Qualifier): A libSBML qualifier type: BIOLOGICAL_QUALIFIER|MODEL_QUALIFIER</p></li>
<li><p>new_b_m_qt (QualifierType): A libSBML biological or model qualifier type like BQB_IS|BQM_IS</p></li>
<li><p>specific_db_prefix (str): Has to be set if only for a specific database the qualifier type should be changed. Can be ‘kegg.genes’, ‘biocyc’, etc.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>model: Model with changed qualifier for given entity type</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.create_fba_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">create_fba_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.UnitDefinition</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#refinegems.polish.create_fba_units" title="Permalink to this definition"></a></dt>
<dd><p>Creates all fba units required for a constraint-based model</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
<dt>Return:</dt><dd><p>list of libSBML UnitDefinitions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.create_unit">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">create_unit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_specs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kind</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">e</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uri_is</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uri_idf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.Unit</span></span></span><a class="headerlink" href="#refinegems.polish.create_unit" title="Permalink to this definition"></a></dt>
<dd><p>Creates unit for SBML model according to arguments</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_specs (tuple):    Level &amp; Version of SBML model
meta_id (str):          Meta ID for unit (Neccessary for URI)
kind (str):             Unit kind constant (see libSBML for available constants)
e (int):                Exponent of unit
m (int):                Multiplier of unit
s (int):                Scale of unit 
uri_is (str):           URI supporting the specified unit
uri_idf (str):          URI supporting the derived from unit</p>
</dd>
<dt>Return:</dt><dd><p>libSBML unit object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.create_unit_definition">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">create_unit_definition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_specs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">units</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Unit</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.UnitDefinition</span></span></span><a class="headerlink" href="#refinegems.polish.create_unit_definition" title="Permalink to this definition"></a></dt>
<dd><p>Creates unit definition for SBML model according to arguments</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_specs (tuple): Level &amp; Version of SBML model
identifier (str):    Identifier for the defined unit
name (str):          Full name of the defined unit
units (list):        All units the defined unit consists of</p>
</dd>
<dt>Return:</dt><dd><p>libSBML unit definition object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.cv_ncbiprotein">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">cv_ncbiprotein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lab_strain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.cv_ncbiprotein" title="Permalink to this definition"></a></dt>
<dd><p>Adds NCBI Id to genes as annotation</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>gene_list (list): libSBML ListOfGenes</p></li>
<li><p>email (str): User Email to access the Entrez database</p></li>
<li><p>protein_fasta (str): The path to the CarveMe protein.fasta input file</p></li>
<li><dl class="simple">
<dt>lab_strain (bool): Needs to be set to True if strain was self-annotated</dt><dd><p>and/or the locus tags in the CarveMe input file should be kept</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.cv_notes_metab">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">cv_notes_metab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Species</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.cv_notes_metab" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>checks the notes field for information which should be in the annotation field</dt><dd><p>removes entry from notes and adds it as URL to the CVTerms of a metabolite</p>
</dd>
<dt>Args:</dt><dd><p>species_list (list): libSBML ListOfSpecies</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.cv_notes_reac">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">cv_notes_reac</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reaction_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.Reaction</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.cv_notes_reac" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>checks the notes field for information which should be in the annotation field</dt><dd><p>removes entry from notes and adds it as URL to the CVTerms of a reaction</p>
</dd>
<dt>Args:</dt><dd><p>reaction_list (list): libSBML ListOfReactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.generate_new_curie_set">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">generate_new_curie_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix2id:</span> <span class="pre">sortedcontainers.SortedDict.slice(&lt;class</span> <span class="pre">'str'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sortedcontainers.SortedSet.&lt;class</span> <span class="pre">'str'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">None)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_pattern:</span> <span class="pre">bool</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">sortedcontainers.SortedSet.&lt;class</span> <span class="pre">'str'&gt;</span></span></span><a class="headerlink" href="#refinegems.polish.generate_new_curie_set" title="Permalink to this definition"></a></dt>
<dd><p>Generate a set of complete CURIEs from the provided prefix to identifier mapping</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>prefix2id (SortedDict[str: SortedSet[str]]): Dictionary containing a mapping from database prefixes to their respective identifier sets</p></li>
<li><p>new_pattern (bool):                          True if new pattern is wanted, otherwise False</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>A sorted set containing complete CURIEs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.get_set_of_curies">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">get_set_of_curies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">curie_list:</span> <span class="pre">list[str])</span> <span class="pre">-&gt;</span> <span class="pre">sortedcontainers.SortedDict.slice(&lt;class</span> <span class="pre">'str'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sortedcontainers.SortedSet.&lt;class</span> <span class="pre">'str'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.get_set_of_curies" title="Permalink to this definition"></a></dt>
<dd><p>Gets a list of CURIEs
&amp; maps the database prefixes to their respective identifier sets</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>curie_list (list[str]): List containing CURIEs</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>Dictionary mapping database prefixes from the provided CURIEs to their respective identifier sets also provided by the CURIEs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.improve_curie_per_entity">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">improve_curie_per_entity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.SBase</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.improve_curie_per_entity" title="Permalink to this definition"></a></dt>
<dd><p>Helper function: Removes duplicates &amp; changes pattern according to new_pattern</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>entity (SBase):      A libSBML SBase object, either a model or an entity</p></li>
<li><p>new_pattern (bool):  True if new pattern is wanted, otherwise False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.improve_curies">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">improve_curies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entities</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.SBase</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.improve_curies" title="Permalink to this definition"></a></dt>
<dd><p>Removes duplicates &amp; changes pattern according to new_pattern</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>entity (SBase):      A libSBML SBase object, either a model or an entity</p></li>
<li><p>new_pattern (bool):  True if new pattern is wanted, otherwise False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.polish">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">polish</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lab_strain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.polish" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>completes all steps to polish a model</dt><dd><p>(Tested for models having either BiGG or VMH identifiers.)</p>
</dd>
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): model loaded with libsbml</p></li>
<li><p>new_filename (Str): filename for modified model</p></li>
<li><p>email (str): E-mail for Entrez</p></li>
<li><p>id_db (str): Main database identifiers in model come from</p></li>
<li><p>protein_fasta (str): File used as input for CarveMe</p></li>
<li><p>lab_strain (bool): True if the strain was sequenced in a local lab</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.polish_annotations">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">polish_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.polish_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Polishes all annotations in a model such that no duplicates are present 
&amp; the same pattern is used for all CURIEs</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model):        Model loaded with libsbml</p></li>
<li><p>new_pattern (bool):   True if new pattern is wanted, otherwise False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.polish_entities">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">polish_entities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabolite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.polish_entities" title="Permalink to this definition"></a></dt>
<dd><p>sets boundary condition and constant if not set for a metabolite</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfSpecies or ListOfReactions
metabolite (boolean): flag to determine whether entity = metabolite</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.print_UnitDefinitions">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">print_UnitDefinitions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contained_unit_defs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.UnitDefinition</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.print_UnitDefinitions" title="Permalink to this definition"></a></dt>
<dd><p>Prints a list of libSBML UnitDefinitions as XMLNodes</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>contained_unit_defs (list): List of libSBML UnitDefinition objects</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.print_remaining_UnitDefinitions">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">print_remaining_UnitDefinitions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_of_fba_units</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">libsbml.UnitDefinition</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.print_remaining_UnitDefinitions" title="Permalink to this definition"></a></dt>
<dd><p>Prints UnitDefinitions from the model that were removed as these were not contained in the list_of_fba_units</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
<li><p>list_of_fba_units (list):  List of libSBML UnitDefinitions</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.set_default_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">set_default_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.set_default_units" title="Permalink to this definition"></a></dt>
<dd><p>Sets default units of model</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.set_initial_amount">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">set_initial_amount</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.set_initial_amount" title="Permalink to this definition"></a></dt>
<dd><p>Sets initial amount to all metabolites if not already set or if initial concentration is not set</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish.set_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish.</span></span><span class="sig-name descname"><span class="pre">set_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish.set_units" title="Permalink to this definition"></a></dt>
<dd><p>Sets units of parameters in model</p>
<dl class="simple">
<dt>Params:</dt><dd><ul class="simple">
<li><p>model (Model): Model loaded with libSBML</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.sboann">
<span id="refinegems-sboann-module"></span><h2>refinegems.sboann module<a class="headerlink" href="#module-refinegems.sboann" title="Permalink to this heading"></a></h2>
<p>Provides functions to automate the addition of SBO terms to the model</p>
<p>Script written by Elisabeth Fritze in her bachelor thesis.
Modified by Gwendolyn O. Gusak during her master thesis.
Commented by Famke Bäuerle and extended by Nantia Leonidou.</p>
<p>It is splitted into a lot of small functions which are all annotated, however when using it for SBO-Term annotation it only makes sense to run the “main” function: sbo_annotation_write(model_libsbml, database_user, database_name, new_filename) if you want to write the modified model to a SBML file or sbo_annotation(model_libsbml, database_user, database_name) if you want to continue with the model. The smaller functions might be useful if special information is needed for a reaction without the context of a bigger model or when the automated annotation fails for some reason.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforCompartments">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforCompartments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforCompartments" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforGenes">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforGenes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforGenes" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforGroups">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforGroups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforGroups" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforMetabolites">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforMetabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforMetabolites" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforModel">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforModel" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOforParameters">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOforParameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOforParameters" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOfromDB">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOfromDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cur</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOfromDB" title="Permalink to this definition"></a></dt>
<dd><p>Adds SBO term based on bigg id of a reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
cur (sqlite3.connect.cursor): used to access the sqlite3 database</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if SBO Term was changed</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.addSBOviaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">addSBOviaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cur</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.addSBOviaEC" title="Permalink to this definition"></a></dt>
<dd><p>Adds SBO terms based on EC numbers given in the annotations of a reactions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
cur (sqlite3.connect.cursor): used to access the sqlite3 database</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkAcetylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkAcetylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkAcetylationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is acetylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkActiveTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkActiveTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkActiveTransport" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is active transport (uses atp/pep) and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkBiomass">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkBiomass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkBiomass" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is biomass / growth and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkCoTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkCoTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkCoTransport" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is co-transport and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDeamination">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDeamination</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDeamination" title="Permalink to this definition"></a></dt>
<dd><p>_tests if reac is deamination and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDeaminationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDeaminationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDeaminationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is deamination by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDecarbonylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDecarbonylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDecarbonylation" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is decarbonylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDecarboxylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDecarboxylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDecarboxylation" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is decarboxylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDecarboxylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDecarboxylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDecarboxylationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is decarboxylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkDemand">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkDemand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkDemand" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is demand and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkExchange">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkExchange</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkExchange" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is exchange and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkGlycosylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkGlycosylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkGlycosylation" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is glycosylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkGlycosylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkGlycosylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkGlycosylationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is glycosylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkHydrolysisViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkHydrolysisViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkHydrolysisViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is hydrolysis by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkIsomerisationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkIsomerisationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkIsomerisationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is isomerisation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkMethylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkMethylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkMethylationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is methylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkPassiveTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkPassiveTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkPassiveTransport" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is passive transport and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkPhosphorylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkPhosphorylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkPhosphorylation" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is phosphorylase / kinase and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkRedox">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkRedox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkRedox" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is redox and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkRedoxViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkRedoxViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkRedoxViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is redox by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkSink">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkSink</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkSink" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is sink and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.checkTransaminationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">checkTransaminationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.checkTransaminationViaEC" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is transamination by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentDict">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentDict" title="Permalink to this definition"></a></dt>
<dd><p>sorts metabolites by compartment</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>dict: compartment as key and metabolites as values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentFromSpeciesRef">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentFromSpeciesRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">speciesReference</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentFromSpeciesRef" title="Permalink to this definition"></a></dt>
<dd><p>extracts compartment from a species by its reference</p>
<dl class="simple">
<dt>Args:</dt><dd><p>speciesReference (libsbml-speciesreference): reference to species</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-compartment: compartment which the species lives in</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentList" title="Permalink to this definition"></a></dt>
<dd><p>extracts compartments of metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentlessMetaboliteIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessMetaboliteIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentlessMetaboliteIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts metabolites which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: all metabolites which have no compartment</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentlessProductIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessProductIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentlessProductIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts products which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: products (metabolites) without compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentlessReactantIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessReactantIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentlessReactantIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts reactants which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: reactants (metabolites) without compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getCompartmentlessSpeciesId">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessSpeciesId</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">speciesReference</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getCompartmentlessSpeciesId" title="Permalink to this definition"></a></dt>
<dd><p>determines wheter a species has compartment by its refernece</p>
<dl class="simple">
<dt>Args:</dt><dd><p>speciesReference (libsbml-speciesreference): reference to species</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-species-id: id of species without compartment</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getECNums">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getECNums</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getECNums" title="Permalink to this definition"></a></dt>
<dd><p>extracts EC-Code from the reaction annotations</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: all EC-Numbers of the reaction</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getListOfMetabolites">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getListOfMetabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getListOfMetabolites" title="Permalink to this definition"></a></dt>
<dd><p>extracts list of metabolites of the reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites that are part of the reaction</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getMetaboliteIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getMetaboliteIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getMetaboliteIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts list of metabolite ids of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolite ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getProductCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getProductCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getProductCompartmentList" title="Permalink to this definition"></a></dt>
<dd><p>extracts compartments of products</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all products (metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getProductIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getProductIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getProductIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts products (metabolites) of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: products (metabolites) ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getReactantCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getReactantCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getReactantCompartmentList" title="Permalink to this definition"></a></dt>
<dd><p>extracts compartments of reactants</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all reactants (metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.getReactantIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">getReactantIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.getReactantIds" title="Permalink to this definition"></a></dt>
<dd><p>extracts reactants (metabolites) of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: reactants (metabolites) ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.hasReactantPair">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">hasReactantPair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">met1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">met2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.hasReactantPair" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>checks if a pair of metabolites is present in reaction</dt><dd><p>needed for special reactions like redox or deamination</p>
</dd>
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
met1 (libsbml-metabolite): metabolite 1 of metabolite pair
met2 (libsbml-metabolite): metabolite 2 of metabolite pair</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if one of the metabolites is in reactants and the other in products</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.isProtonTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">isProtonTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.isProtonTransport" title="Permalink to this definition"></a></dt>
<dd><p>check if reaction is proton transport</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction is proton transport</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.moreThanTwoCompartmentTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">moreThanTwoCompartmentTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.moreThanTwoCompartmentTransport" title="Permalink to this definition"></a></dt>
<dd><p>check if reaction traverses more than 2 compartments</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction traverses more than 2 compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.returnCompartment">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">returnCompartment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.returnCompartment" title="Permalink to this definition"></a></dt>
<dd><p>helper to split compartment id</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.sbo_annotation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">sbo_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_libsbml</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.sbo_annotation" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>executes all steps to annotate SBO terms to a given model</dt><dd><p>(former main function of original script by Elisabeth Fritze)</p>
</dd>
<dt>Args:</dt><dd><p>model_libsbml (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: modified model with SBO terms</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.sbo_annotation_write">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">sbo_annotation_write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_libsbml</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.sbo_annotation_write" title="Permalink to this definition"></a></dt>
<dd><p>wrapper around sbo_annotation to include writing to a file</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_libsbml (libsbml-model): model loaded with libsbml
new_filename (Str): filename for modified model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.soleProtonTransported">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">soleProtonTransported</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.soleProtonTransported" title="Permalink to this definition"></a></dt>
<dd><p>check if reaction is transport powered by one H</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction is transport powered by one H</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.splitSymAntiPorter">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">splitSymAntiPorter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.splitSymAntiPorter" title="Permalink to this definition"></a></dt>
<dd><p>tests if reac is sym- or antiporter and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sboann.splitTransportBiochem">
<span class="sig-prename descclassname"><span class="pre">refinegems.sboann.</span></span><span class="sig-name descname"><span class="pre">splitTransportBiochem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sboann.splitTransportBiochem" title="Permalink to this definition"></a></dt>
<dd><p>test if reaction traverses more than 1 compartment and set SBO Term</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="explanation.html" class="btn btn-neutral float-left" title="Ideas behind some implementations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Famke Bäuerle.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>