
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>refinegems package &#8212; refinegems 1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="refinegems" href="modules.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="refinegems-package">
<h1>refinegems package<a class="headerlink" href="#refinegems-package" title="Permalink to this headline">¶</a></h1>
<section id="module-refinegems.charges">
<span id="refinegems-charges-module"></span><h2>refinegems.charges module<a class="headerlink" href="#module-refinegems.charges" title="Permalink to this headline">¶</a></h2>
<p>Provides functions for adding charges to metabolites</p>
<p>When iterating thorugh all metabolites present in a model, you will find several which have no defined charge (metab.getPlugin(‘fbc’).isSetCharge() = false). This can lead to charge imbalanced reactions. This script takes information on metabolite charges from the ModelSEED database. A charge is automatically added to a metabolite if it has no defined charge and if there is only one charge denoted in ModelSEED. When multiple charges are present, the metabolite and the possible charges are noted and later returned in a dictionary.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.charges.correct_charges">
<span class="sig-prename descclassname"><span class="pre">refinegems.charges.</span></span><span class="sig-name descname"><span class="pre">correct_charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modelseed_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.charges.correct_charges" title="Permalink to this definition">¶</a></dt>
<dd><p>wrapper function which completes the steps to charge correction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
new_file_path (Str): filepath + name for modified model
modelseed_path (str): path to modelseed compound definition</p>
</dd>
<dt>Returns:</dt><dd><p>dict: BiGG Id and possible charges of metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.charges.correct_charges_modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.charges.</span></span><span class="sig-name descname"><span class="pre">correct_charges_modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modelseed_compounds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.charges.correct_charges_modelseed" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds charges taken from the ModelSEED database to metabolites which have no defined charge</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
modelseed_compounds (df): containing modelseed data</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (model with added charges, metabolites with multiple charges)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.charges.get_modelseed_compounds">
<span class="sig-prename descclassname"><span class="pre">refinegems.charges.</span></span><span class="sig-name descname"><span class="pre">get_modelseed_compounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.charges.get_modelseed_compounds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compounds from modelseed which have BiGG Ids</p>
<dl class="simple">
<dt>Args:</dt><dd><p>path (str): path to modelseed compound definition</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing modelseed data</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.genecomp">
<span id="refinegems-genecomp-module"></span><h2>refinegems.genecomp module<a class="headerlink" href="#module-refinegems.genecomp" title="Permalink to this headline">¶</a></h2>
<p>Provides functions to compare genes found in KEGG and in the model</p>
<p>Extracts all KEGG ids from the annotations and compares them to a list for you organism from KEGG.
Reactions with KEGG ids not found in the model are expanded to a table containing the KEGG id, 
the locus_tag (old format), the EC number, the BiGG id and the Gene Protein Reaction (GPR) rule.
This section needs a gff file of your organism, the txt file from BiGG containing all reactions
and the KEGG identifier of your organism.</p>
<p>Due to the KEGG REST API this is relatively slow (model of size 1500 reactions - 20 min).</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.compare_bigg_model">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">compare_bigg_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_ec_kegg_bigg_gpr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_reactions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.compare_bigg_model" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Compares reactions by genes extracted via KEGG to reactions in the model</dt><dd><p>Needed to back check previous comparisons.</p>
</dd>
<dt>Args:</dt><dd><p>locus_ec_kegg_bigg_gpr (df): locus tag, EC number, KEGG Id, BiGG Id and GPR
model_reactions (df): BiGG Ids of reactions in the model</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing reactions present in KEGG but not in the model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.compare_kegg_model">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">compare_kegg_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_genes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.compare_kegg_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts list of genes present in KEGG but not in the model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_genes (df): all genes in the model
kegg_genes (df): all genes denoted in KEGG for the organism</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of all genes present in KEGG but not in the model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.genecomp">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">genecomp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">organismid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biggreactions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gff_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.genecomp" title="Permalink to this definition">¶</a></dt>
<dd><p>Exectues all steps to compare genes of the model to KEGG genes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
organismid (Str): KEGG Id of organism which the model is based on
biggreactions (Str): path to file containing BiGG database
gff_file (Str): path to gff file of organism of interest</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing missing reactions with locus tag, EC number, KEGG Id, BiGG Id and GPR</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_bigg_kegg">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_bigg_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">biggreactions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_bigg_kegg" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses list of BiGG reactions to get a mapping from BiGG to KEGG Id</p>
<dl class="simple">
<dt>Args:</dt><dd><p>biggreactions (Str): path to file containing BiGG database</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing BiGG Ids with corresponding KEGG Ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_kegg_genes">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_kegg_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organismid</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_kegg_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts list of genes from KEGG given an organism</p>
<dl class="simple">
<dt>Args:</dt><dd><p>organismid (Str): KEGG Id of organism which the model is based on</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of all genes denoted in KEGG for the organism</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_locus_ec">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_locus_ec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes_kegg_notmodel</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_locus_ec" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates columns with EC numbers for the locus tags of the genes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>genes_kegg_notmodel (df): genes present in KEGG but not in the model</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of genes with locus tag and EC number</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_locus_ec_kegg">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_locus_ec_kegg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_ec</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_locus_ec_kegg" title="Permalink to this definition">¶</a></dt>
<dd><p>Searches for KEGG reactions based on EC numbers</p>
<dl class="simple">
<dt>Args:</dt><dd><p>locus_ec (df): genes with locus tag and EC number</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of genes with locus tag, EC number and KEGG Id</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_locus_ec_kegg_bigg">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_locus_ec_kegg_bigg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_ec_kegg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bigg_kegg</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_locus_ec_kegg_bigg" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges table with genes from model with BiGG / KEGG mapping to add BiGG Ids</p>
<dl class="simple">
<dt>Args:</dt><dd><p>locus_ec_kegg (df): genes with locus tag, EC number and KEGG Id
bigg_kegg (df): BiGG Ids with corresponding KEGG Ids</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of genes with locus tag, EC number, KEGG Id and BiGG Id</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_locus_ec_kegg_bigg_gpr">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_locus_ec_kegg_bigg_gpr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_ec_kegg_bigg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_gpr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_locus_ec_kegg_bigg_gpr" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges table with genes from model if locus tag / GPR mapping to add GPRs</p>
<dl class="simple">
<dt>Args:</dt><dd><p>locus_ec_kegg_bigg (df): genes with locus tag, EC number, KEGG Id and BiGG Id
locus_gpr (df): mapping from locus tags to GPRs</p>
</dd>
<dt>Returns:</dt><dd><p>df: table of genes with locus tag, EC number, KEGG Id, BiGG Id and GPR</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_locus_gpr">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_locus_gpr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gff_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_locus_gpr" title="Permalink to this definition">¶</a></dt>
<dd><p>Searches gff file of organism for gene protein reactions based on locus tags</p>
<dl class="simple">
<dt>Args:</dt><dd><p>gff_file (Str): path to gff file of organism of interest</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing mapping from locus tag to GPR</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_model_genes">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_model_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_model_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts genes from given model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (model-libsbml): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>df: table with all genes in the model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.genecomp.get_model_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.genecomp.</span></span><span class="sig-name descname"><span class="pre">get_model_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.genecomp.get_model_reactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts table of reactions with BiGG Ids from model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>df: table with BiGG Ids of reactions in the model</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.growth">
<span id="refinegems-growth-module"></span><h2>refinegems.growth module<a class="headerlink" href="#module-refinegems.growth" title="Permalink to this headline">¶</a></h2>
<p>Provides functions to simulate growth on any medium</p>
<p>Tailored to work with the Synthetic Nasal Medium (SNM), should
work with any medium as long as its defined in a csv with ; as 
delimiter and BiGG Ids for the compounds.</p>
<p>Outputs a table where</p>
<p>essential = metabolites not present in the medium but necessary for growth</p>
<p>missing = all exchanges missing in the model but given in medium</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.find_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">find_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">essential</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.find_minimum_essential" title="Permalink to this definition">¶</a></dt>
<dd><p>Report metabolites necessary for growth and not in custom medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>medium (df): dataframe of medium csv
essential (list): metabolites which lead to zero growth if blocked</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites not present in the medium but necessary for growth</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.find_missing_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">find_missing_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_uptake</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.find_missing_essential" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Report which exchange reactions are needed for growth</dt><dd><p>combines default uptake and valid new medium</p>
</dd>
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (dict): medium with exchanges that are present in the model
default_uptake (list): metabolites consumed in standard medium</p>
</dd>
<dt>Returns:</dt><dd><p>list: all metabolites which lead to zero growth if blocked</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_all_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_all_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_all_minimum_essential" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns metabolites necessary for growth and not in media</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
mediumpath (string): path to csv with media definitions</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on different media which metabs are missing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_default_uptake">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_default_uptake</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_default_uptake" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines which metabolites are used in the standard medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites consumed in standard medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_growth_one_medium">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_growth_one_medium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_growth_one_medium" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates growth on given medium</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (pandas-DataFrame): table containing metabolites present in the medium</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on growth behaviour on medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_growth_selected_media">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_growth_selected_media</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">media</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_growth_selected_media" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates growth on all selected media</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
mediumpath (string): path to csv with media definitions
media (list): media to simulate on (must be in csv)</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: information on growth behaviour on selected media</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.get_missing_exchanges">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">get_missing_exchanges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.get_missing_exchanges" title="Permalink to this definition">¶</a></dt>
<dd><p>Look for exchange reactions needed by the medium but not in the model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (df): dataframe of medium csv</p>
</dd>
<dt>Returns:</dt><dd><p>list: all exchanges missing in the model but given in medium</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.modify_medium">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">modify_medium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_exchanges</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.modify_medium" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove exchanges from medium that are not in the model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>medium (df): dataframe of medium csv
missing_exchanges (list): exchanges not in the model</p>
</dd>
<dt>Returns:</dt><dd><p>dict: medium that can be used with the model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.growth.simulate_minimum_essential">
<span class="sig-prename descclassname"><span class="pre">refinegems.growth.</span></span><span class="sig-name descname"><span class="pre">simulate_minimum_essential</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medium</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.growth.simulate_minimum_essential" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate growth with custom medium plus necessary uptakes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy
medium (dict): medium with exchanges present in the model
minimum (list): metabolites not present in the medium but necessary for growth</p>
</dd>
<dt>Returns:</dt><dd><p>float: growth value</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.load">
<span id="refinegems-load-module"></span><h2>refinegems.load module<a class="headerlink" href="#module-refinegems.load" title="Permalink to this headline">¶</a></h2>
<p>Provides functions to load a modelpath with libSBML and cobrapy</p>
<p>Depending on the application the model needs to be loaded with cobra (memote)
or with libSBML (activation of groups).</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_all_media_from_db">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_all_media_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_all_media_from_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to extract media definitions from media_db.csv</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumpath (Str): path to csv file with medium database</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe of csv with metabs added as BiGG_EX exchange reactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_document_libsbml">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_document_libsbml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_document_libsbml" title="Permalink to this definition">¶</a></dt>
<dd><p>loads model document using libsbml</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-document: loaded document by libsbml</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_medium_custom">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_medium_custom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_medium_custom" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to read medium csv</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumpath (Str): path to csv file with medium</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe of csv</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_medium_from_db">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_medium_from_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mediumpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mediumname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_medium_from_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to read standard media_db.csv</p>
<dl class="simple">
<dt>Args:</dt><dd><p>mediumpath (Str): path to csv file with medium database
mediumname (Str): name of medium to test growth on</p>
</dd>
<dt>Returns:</dt><dd><p>df: pandas dataframe of csv</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_model_cobra">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_model_cobra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_model_cobra" title="Permalink to this definition">¶</a></dt>
<dd><p>loads model using cobrapy</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>cobra-model: loaded model by cobrapy</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.load_model_libsbml">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">load_model_libsbml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.load_model_libsbml" title="Permalink to this definition">¶</a></dt>
<dd><p>loads model using libsbml</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: loaded model by libsbml</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.write_report">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">write_report</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataframe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.write_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes reports stored in dataframes to xlsx file</p>
<dl class="simple">
<dt>Args:</dt><dd><p>dataframe (pd.DataFrame): table containing output
filepath (string): path to file with filename</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.load.write_to_file">
<span class="sig-prename descclassname"><span class="pre">refinegems.load.</span></span><span class="sig-name descname"><span class="pre">write_to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.load.write_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes modified model to new file</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
new_filename (Str): filename / path for modified model</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.modelseed">
<span id="refinegems-modelseed-module"></span><h2>refinegems.modelseed module<a class="headerlink" href="#module-refinegems.modelseed" title="Permalink to this headline">¶</a></h2>
<p>Reports mismatches in charges and formulae based on ModelSEED</p>
<p>Extracts modelseed data from a given tsv file, extracts all metabolites
from a given model. Both lists of metabolites are compared by charge and
formula.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.compare_model_modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">compare_model_modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_charges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modelseed_charges</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.compare_model_modelseed" title="Permalink to this definition">¶</a></dt>
<dd><p>compares tables with charges / formulae from model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model_charges (df): charges and formulae of model metabolites
modelseed_charges (df): charges and formulae of modelseed metabolites</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing info whether charges / formulae match</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_charge_mismatch">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_charge_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_comp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_charge_mismatch" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts metabolites with charge mismatch of model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>df_comp (df): charge and formula mismatches</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with charge mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_compared_formulae">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_compared_formulae</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">formula_mismatch</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_compared_formulae" title="Permalink to this definition">¶</a></dt>
<dd><p>compare formula by atom pattern</p>
<dl class="simple">
<dt>Args:</dt><dd><p>formula_mismatch (df): table with column containing atom comparison</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with formula mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_formula_mismatch">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_formula_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_comp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_formula_mismatch" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts metabolites with formula mismatch of model &amp; modelseed</p>
<dl class="simple">
<dt>Args:</dt><dd><p>df_comp (df): charge and formula mismatches</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing metabolites with formula mismatch</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_model_charges">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_model_charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_model_charges" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts all metabolites from model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing charges and formulae of model metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_modelseed_charges">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_modelseed_charges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelseed_compounds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_modelseed_charges" title="Permalink to this definition">¶</a></dt>
<dd><p>extract table with BiGG, charges and formulae</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelseed_compounds (df): containing modelseed data</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing charges and formulae of modelseed metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.get_modelseed_compounds">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">get_modelseed_compounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.get_modelseed_compounds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compounds from modelseed which have BiGG Ids</p>
<dl class="simple">
<dt>Args:</dt><dd><p>path (str): path to modelseed compound definition</p>
</dd>
<dt>Returns:</dt><dd><p>df: table containing modelseed data</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.modelseed.modelseed">
<span class="sig-prename descclassname"><span class="pre">refinegems.modelseed.</span></span><span class="sig-name descname"><span class="pre">modelseed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.modelseed.modelseed" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes all steps to compare model metabolites to modelseed metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>path (str): path to modelseed compound definition
model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (table with charge mismatches, formula mismatches)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.pathways">
<span id="refinegems-pathways-module"></span><h2>refinegems.pathways module<a class="headerlink" href="#module-refinegems.pathways" title="Permalink to this headline">¶</a></h2>
<p>Provides functions for adding KEGG reactions as Group Pathways</p>
<p>If your organism occurs in the KEGG database, extract the KEGG reaction ID from the
annotations of your reactions and identify, in which KEGG pathways this reaction occurs. Add
all KEGG pathways for a reaction then as annotations with the biological qualifier ‘OCCURS_IN’
to the respective reaction.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.add_kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">add_kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kegg_pathways</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.add_kegg_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>add KEGG reactions as BQB_OCCURS_IN</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
kegg_pathways (dict): reaction Id as key and Kegg Pathway Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: modified model with Kegg pathways</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.create_pathway_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">create_pathway_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pathway_groups</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.create_pathway_groups" title="Permalink to this definition">¶</a></dt>
<dd><p>use group module to add reactions to Kegg pathway</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml and containing Kegg pathways
pathway_groups (dict): Kegg Pathway Id as key and reactions Ids as values</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-model: modified model with groups for pathways</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.extract_kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">extract_kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_reactions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.extract_kegg_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>finds pathway for KEGG reactions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>kegg_reactions (dict): reaction Id as key and Kegg Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>dict: reaction Id as key and Kegg Pathway Id as value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.extract_kegg_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">extract_kegg_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.extract_kegg_reactions" title="Permalink to this definition">¶</a></dt>
<dd><p>extract KEGG Ids from reactions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>dict: reaction Id as key and Kegg Id as value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.get_pathway_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">get_pathway_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kegg_pathways</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.get_pathway_groups" title="Permalink to this definition">¶</a></dt>
<dd><p>group reaction into pathways</p>
<dl class="simple">
<dt>Args:</dt><dd><p>kegg_pathways (dict): reaction Id as key and Kegg Pathway Id as value</p>
</dd>
<dt>Returns:</dt><dd><p>dict: Kegg Pathway Id as key and reactions Ids as values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.kegg_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">kegg_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.kegg_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Executes all steps to add KEGG pathways as groups</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM
new_filename (Str): filename for modified model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.pathways.load_model_enable_groups">
<span class="sig-prename descclassname"><span class="pre">refinegems.pathways.</span></span><span class="sig-name descname"><span class="pre">load_model_enable_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.pathways.load_model_enable_groups" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>loads model as document using libsbml</dt><dd><p>enables groups extension</p>
</dd>
<dt>Args:</dt><dd><p>modelpath (Str): Path to GEM</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-document: loaded document by libsbml</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.polish_carveme">
<span id="refinegems-polish-carveme-module"></span><h2>refinegems.polish_carveme module<a class="headerlink" href="#module-refinegems.polish_carveme" title="Permalink to this headline">¶</a></h2>
<p>Can be used to polish a model created with CarveMe v.1.5.1</p>
<p>The newer version of CarveMe leads to some irritations in the model, these scripts enable for example the addition of BiGG Ids to the annotations as well as a correct formatting of the annotations.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.add_bigg_metab">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">add_bigg_metab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.add_bigg_metab" title="Permalink to this definition">¶</a></dt>
<dd><p>adds the BiGG Id of metabolites as URL to the annotation field</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfSpecies</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.add_bigg_reac">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">add_bigg_reac</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.add_bigg_reac" title="Permalink to this definition">¶</a></dt>
<dd><p>adds the BiGG Id of reactions as URL to the annotation field</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfReactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.cv_ncbiprotein">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">cv_ncbiprotein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.cv_ncbiprotein" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds NCBI Id to genes as annotation</p>
<dl class="simple">
<dt>Args:</dt><dd><p>gene_list (list): libSBML ListOfGenes
email (string): User Email to access the Entrez database</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.cv_notes_metab">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">cv_notes_metab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">species_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.cv_notes_metab" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>checks the notes field for information which should be in the annotation field</dt><dd><p>removes entry from notes and adds it as URL to the CVTerms of a metabolite</p>
</dd>
<dt>Args:</dt><dd><p>species_list (list): libSBML ListOfSpecies</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.cv_notes_reac">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">cv_notes_reac</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reaction_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.cv_notes_reac" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>checks the notes field for information which should be in the annotation field</dt><dd><p>removes entry from notes and adds it as URL to the CVTerms of a reaction</p>
</dd>
<dt>Args:</dt><dd><p>reaction_list (list): libSBML ListOfReactions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.polish_carveme">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">polish_carveme</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.polish_carveme" title="Permalink to this definition">¶</a></dt>
<dd><p>completes all steps to polish a model created with CarveMe</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml
new_filename (Str): filename for modified model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.polish_entities">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">polish_entities</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabolite</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.polish_entities" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>removes Notes field from entity</dt><dd><p>sets boundary condition and constant if not set for a metabolite</p>
</dd>
<dt>Args:</dt><dd><p>entity_list (list): libSBML ListOfSpecies or ListOfReactions
metabolite (boolean): flag to determine whether entity = metabolite</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.polish_carveme.set_units">
<span class="sig-prename descclassname"><span class="pre">refinegems.polish_carveme.</span></span><span class="sig-name descname"><span class="pre">set_units</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.polish_carveme.set_units" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds units to parameters in model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.sbo_annotation">
<span id="refinegems-sbo-annotation-module"></span><h2>refinegems.sbo_annotation module<a class="headerlink" href="#module-refinegems.sbo_annotation" title="Permalink to this headline">¶</a></h2>
<p>Provides functions to automate the addition of SBO terms to the model</p>
<p>Script written by Elisabeth Fritze in her bachelor thesis.</p>
<p>It is splitted into a lot of small functions which are all annotated, however when using it for SBO-Term annotation it only makes sense to run the “main” function: sbo_annotation(model_libsbml, database_user, database_name, new_filename). The smaller functions might be useful if special information is needed for a reaction without the context of a bigger model or when the automated annotation fails for some reason.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.addSBOfromDB">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">addSBOfromDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cur</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.addSBOfromDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds SBO term based on bigg id of a reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
cur (psycopg2.connect.cursor): used to access the psql database</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if SBO Term was changed</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.addSBOviaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">addSBOviaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cur</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.addSBOviaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds SBO terms based on EC numbers given in the annotations of a reactions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
cur (psycopg2.connect.cursor): used to access the psql database</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkAcetylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkAcetylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkAcetylationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is acetylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkActiveTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkActiveTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkActiveTransport" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is active transport (uses atp/pep) and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkBiomass">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkBiomass</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkBiomass" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is biomass / growth and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkCoTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkCoTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkCoTransport" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is co-transport and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDeamination">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDeamination</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDeamination" title="Permalink to this definition">¶</a></dt>
<dd><p>_tests if reac is deamination and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDeaminationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDeaminationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDeaminationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is deamination by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDecarbonylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDecarbonylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDecarbonylation" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is decarbonylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDecarboxylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDecarboxylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDecarboxylation" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is decarboxylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDecarboxylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDecarboxylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDecarboxylationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is decarboxylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkDemand">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkDemand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkDemand" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is demand and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkExchange">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkExchange</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkExchange" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is exchange and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkGlycosylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkGlycosylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkGlycosylation" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is glycosylation and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkGlycosylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkGlycosylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkGlycosylationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is glycosylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkHydrolysisViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkHydrolysisViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkHydrolysisViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is hydrolysis by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkIsomerisationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkIsomerisationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkIsomerisationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is isomerisation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkMethylationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkMethylationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkMethylationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is methylation by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkPassiveTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkPassiveTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkPassiveTransport" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is passive transport and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkPhosphorylation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkPhosphorylation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkPhosphorylation" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is phosphorylase / kinase and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkRedox">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkRedox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkRedox" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is redox and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkRedoxViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkRedoxViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkRedoxViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is redox by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkSink">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkSink</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkSink" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is sink and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.checkTransaminationViaEC">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">checkTransaminationViaEC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.checkTransaminationViaEC" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is transamination by its EC-Code and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentDict">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentDict" title="Permalink to this definition">¶</a></dt>
<dd><p>sorts metabolites by compartment</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>dict: compartment as key and metabolites as values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentFromSpeciesRef">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentFromSpeciesRef</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">speciesReference</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentFromSpeciesRef" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compartment from a species by its reference</p>
<dl class="simple">
<dt>Args:</dt><dd><p>speciesReference (libsbml-speciesreference): reference to species</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-compartment: compartment which the species lives in</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentList" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compartments of metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all metabolites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentlessMetaboliteIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessMetaboliteIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentlessMetaboliteIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts metabolites which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: all metabolites which have no compartment</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentlessProductIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessProductIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentlessProductIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts products which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: products (metabolites) without compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentlessReactantIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessReactantIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentlessReactantIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts reactants which have no compartment information</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: reactants (metabolites) without compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getCompartmentlessSpeciesId">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getCompartmentlessSpeciesId</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">speciesReference</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getCompartmentlessSpeciesId" title="Permalink to this definition">¶</a></dt>
<dd><p>determines wheter a species has compartment by its refernece</p>
<dl class="simple">
<dt>Args:</dt><dd><p>speciesReference (libsbml-speciesreference): reference to species</p>
</dd>
<dt>Returns:</dt><dd><p>libsbml-species-id: id of species without compartment</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getECNums">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getECNums</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getECNums" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts EC-Code from the reaction annotations</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: all EC-Numbers of the reaction</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getListOfMetabolites">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getListOfMetabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getListOfMetabolites" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts list of metabolites of the reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolites that are part of the reaction</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getMetaboliteIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getMetaboliteIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getMetaboliteIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts list of metabolite ids of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: metabolite ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getProductCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getProductCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getProductCompartmentList" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compartments of products</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all products (metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getProductIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getProductIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getProductIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts products (metabolites) of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: products (metabolites) ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getReactantCompartmentList">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getReactantCompartmentList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getReactantCompartmentList" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts compartments of reactants</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>set: compartment information of all reactants (metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.getReactantIds">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">getReactantIds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.getReactantIds" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts reactants (metabolites) of reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>list: reactants (metabolites) ids</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.hasReactantPair">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">hasReactantPair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">met1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">met2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.hasReactantPair" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>checks if a pair of metabolites is present in reaction</dt><dd><p>needed for special reactions like redox or deamination</p>
</dd>
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model
met1 (libsbml-metabolite): metabolite 1 of metabolite pair
met2 (libsbml-metabolite): metabolite 2 of metabolite pair</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if one of the metabolites is in reactants and the other in products</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.isProtonTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">isProtonTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.isProtonTransport" title="Permalink to this definition">¶</a></dt>
<dd><p>check if reaction is proton transport</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction is proton transport</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.moreThanTwoCompartmentTransport">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">moreThanTwoCompartmentTransport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.moreThanTwoCompartmentTransport" title="Permalink to this definition">¶</a></dt>
<dd><p>check if reaction traverses more than 2 compartments</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction traverses more than 2 compartments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.returnCompartment">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">returnCompartment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.returnCompartment" title="Permalink to this definition">¶</a></dt>
<dd><p>helper to split compartment id</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.sbo_annotation">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">sbo_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_libsbml</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database_user</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.sbo_annotation" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>executes all steps to annotate sbo terms to a given model</dt><dd><p>(former main function of original script by Elisabeth Fritze)</p>
</dd>
<dt>Args:</dt><dd><p>model_libsbml (libsbml-model): model loaded with libsbml
database_user (Str): username for postgresql database
database_name (Str): name of database in which create_dbs.sql was imported
new_filename (Str): filename for modified model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.soleProtonTransported">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">soleProtonTransported</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.soleProtonTransported" title="Permalink to this definition">¶</a></dt>
<dd><p>check if reaction is transport powered by one H</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
<dt>Returns:</dt><dd><p>bool: True if reaction is transport powered by one H</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.splitSymAntiPorter">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">splitSymAntiPorter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.splitSymAntiPorter" title="Permalink to this definition">¶</a></dt>
<dd><p>tests if reac is sym- or antiporter and sets SBO Term if true</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.sbo_annotation.splitTransportBiochem">
<span class="sig-prename descclassname"><span class="pre">refinegems.sbo_annotation.</span></span><span class="sig-name descname"><span class="pre">splitTransportBiochem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.sbo_annotation.splitTransportBiochem" title="Permalink to this definition">¶</a></dt>
<dd><p>test if reaction traverses more than 1 compartment and set SBO Term</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reac (libsbml-reaction): libsbml reaction from sbml model</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-refinegems.test">
<span id="refinegems-test-module"></span><h2>refinegems.test module<a class="headerlink" href="#module-refinegems.test" title="Permalink to this headline">¶</a></h2>
<p>Provides functions to investigate the model and test with memote</p>
<p>These functions enable simple testing of any model using memote and 
access to its number of reactions, metabolites and genes.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.get_mass_charge_unbalanced">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">get_mass_charge_unbalanced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.get_mass_charge_unbalanced" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>creates lists of mass and charge unbalanced reactions, </dt><dd><p>without exchange reactions since they are unbalanced per definition</p>
</dd>
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (list of mass unbalanced, charge unbalanced reactions)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.get_model_info">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">get_model_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.get_model_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Reports core information of given model</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelpath (string): path to model file</p>
</dd>
<dt>Returns:</dt><dd><p>DataFrame: overview on model parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.get_orphans_deadends_disconnected">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">get_orphans_deadends_disconnected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.get_orphans_deadends_disconnected" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses memote functions to extract orphans, deadends and disconnected metabolites</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (list of BiGG ids of orphans, deadends, disconnected metabolites)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.initial_analysis">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">initial_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.initial_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>extracts most important numbers of GEM</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (libsbml-model): model loaded with libsbml</p>
</dd>
<dt>Returns:</dt><dd><p>tuple: (name of model, no of reactions, no of metabolites, no of genes)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.run_memote">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">run_memote</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.run_memote" title="Permalink to this definition">¶</a></dt>
<dd><p>runs memote to obtain total score</p>
<dl class="simple">
<dt>Args:</dt><dd><p>model (cobra-model): model loaded with cobrapy</p>
</dd>
<dt>Returns:</dt><dd><p>float: memote score of model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.test.run_memote_sys">
<span class="sig-prename descclassname"><span class="pre">refinegems.test.</span></span><span class="sig-name descname"><span class="pre">run_memote_sys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modelfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.test.run_memote_sys" title="Permalink to this definition">¶</a></dt>
<dd><p>run memote on linux machine</p>
<dl class="simple">
<dt>Args:</dt><dd><p>modelfile (cobra-model): model loaded with cobrapy</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="refinegems-cvterms-module">
<h2>refinegems.cvterms module<a class="headerlink" href="#refinegems-cvterms-module" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-refinegems.cvterms"></span><p>Helper module to work with annotations (CVTerms)</p>
<p>Stores dictionaries which hold information the identifiers.org syntax, has functions to add CVTerms to different entities and parse CVTerms.</p>
<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_genes">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Add CVTerm to a gene</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
metab (libsbml-gene): gene to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_metabolites">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metab</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_metabolites" title="Permalink to this definition">¶</a></dt>
<dd><p>Add CVTerm to a metabolite</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
metab (libsbml-species): metabolite to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_pathways">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_pathways" title="Permalink to this definition">¶</a></dt>
<dd><p>Add CVTerm to a groups pathway</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
metab (libsbml-group): pathway to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.add_cv_term_reactions">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">add_cv_term_reactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reac</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.add_cv_term_reactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Add CVTerm to a reaction</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entry (string): id to add as annotation
db_id (string): database to which entry belongs
reac (libsbml-reaction): reaction to add CVTerm to</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="refinegems.cvterms.parse_id_from_cv_term">
<span class="sig-prename descclassname"><span class="pre">refinegems.cvterms.</span></span><span class="sig-name descname"><span class="pre">parse_id_from_cv_term</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#refinegems.cvterms.parse_id_from_cv_term" title="Permalink to this definition">¶</a></dt>
<dd><p>extract id for a specific database from CVTerm</p>
<dl class="simple">
<dt>Args:</dt><dd><p>entity (libsbml Object): Species, Reaction, Gene
db_id (string): database of interes</p>
</dd>
<dt>Returns:</dt><dd><p>list: ids of entity belonging to db_id</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="refinegems-strains-module">
<h2>refinegems.strains module<a class="headerlink" href="#refinegems-strains-module" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-refinegems.strains"></span><p>Implements functions to use refinegems with models of the same bacterium but different strains</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">refinegems</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="application.html">Usage as standalone application</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage as python module</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">refinegems</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">refinegems package</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="modules.html">refinegems</a><ul>
      <li>Previous: <a href="modules.html" title="previous chapter">refinegems</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Famke Bäuerle.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.5.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/refinegems.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>