<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Filling gaps with refineGEMs &mdash; refineGEMs 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom_theme.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/refineGEMs_logo.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Growth simulation with RefineGEMs" href="growth.html" />
    <link rel="prev" title="Main modules" href="../main-modules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/refineGEMs_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../main-modules.html">Main modules</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Gap filling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#manual-gap-filling">Manual gap filling</a></li>
<li class="toctree-l3"><a class="reference internal" href="#automated-gap-filling">Automated gap filling</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-module-gapfill-can-be-used-to">The module <code class="docutils literal notranslate"><span class="pre">gapfill</span></code> can be used to:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#relevant-parameters">Relevant parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-acquisition-from-biocyc">Data acquisition from BioCyc</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="growth.html">Growth simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathways.html">Pathway addition</a></li>
<li class="toctree-l2"><a class="reference internal" href="sboann.html">SBO-Term annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="polish.html">Automated polishing</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples.html">Application examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../in_silico_media_generation.html">From laboratory to <em>in silico</em> medium</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API access</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline.html">Pipeline: From genome sequence to draft model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../development.html">Notes for developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">refineGEMs</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../main-modules.html">Main modules</a></li>
      <li class="breadcrumb-item active">Filling gaps with refineGEMs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modules/gapfill.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="filling-gaps-with-refinegems">
<h1>Filling gaps with refineGEMs<a class="headerlink" href="#filling-gaps-with-refinegems" title="Permalink to this heading"></a></h1>
<p>There are two possibilities to use refineGEMs to fill gaps.</p>
<section id="manual-gap-filling">
<h2>Manual gap filling<a class="headerlink" href="#manual-gap-filling" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">refinegems.curate.</span></span><span class="sig-name descname"><span class="pre">add_reactions_from_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">email</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.Model</span></span></span></dt>
<dd><p>Wrapper function to use with table format given in data/manual_curation.xlsx, sheet gapfill: Adds all reactions with their info given in the table to the given model</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model (libModel): Model loaded with libSBML</p></li>
<li><p>table (pd-DataFrame): Table in format of sheet gapfill from manual_curation.xlsx located in the data folder</p></li>
<li><p>email (str): User Email to access the NCBI Entrez database</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>libModel: Modified model with new reactions</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="automated-gap-filling">
<h2>Automated gap filling<a class="headerlink" href="#automated-gap-filling" title="Permalink to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">gapfill</span></code> module was created to enable an automatic way of filling gaps in a model via genes.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<dl class="simple">
<dt>Current restrictions:</dt><dd><ul class="simple">
<li><p>Only bacteria supported as missing reactions are filtered by compartments (‘c’, ‘e’, ‘p’).</p></li>
<li><p>Only models where the organisms have entries in KEGG and/or BioCyc can be gap filled.</p></li>
</ul>
</dd>
</dl>
</div>
<dl class="simple">
<dt>Run times:</dt><dd><ul class="simple">
<li><p>‘KEGG’: ~ 2h</p></li>
<li><p>‘BioCyc’: ~ 45mins - 1h</p></li>
<li><p>‘KEGG+BioCyc’: ~ 3 - 4h</p></li>
</ul>
</dd>
</dl>
<section id="the-module-gapfill-can-be-used-to">
<h3>The module <code class="docutils literal notranslate"><span class="pre">gapfill</span></code> can be used to:<a class="headerlink" href="#the-module-gapfill-can-be-used-to" title="Permalink to this heading"></a></h3>
<section id="a-perform-gap-analysis">
<h4>a. perform gap analysis<a class="headerlink" href="#a-perform-gap-analysis" title="Permalink to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">refinegems.gapfill.</span></span><span class="sig-name descname"><span class="pre">gap_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">model_libsbml:</span> <span class="pre">libsbml.Model,</span> <span class="pre">gapfill_params:</span> <span class="pre">dict[slice(&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">None)],</span> <span class="pre">filename:</span> <span class="pre">str</span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></span></dt>
<dd><div class="line-block">
<div class="line">Main function to infer gaps in a model by comparing the locus tags of the GeneProducts 
| to KEGG/BioCyc/both</div>
</div>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model_libsbml (libModel): Model loaded with libSBML</p></li>
<li><p>gapfill_params (dict): Dictionary obtained from YAML file containing the parameter mappings</p></li>
<li><p>filename (str): Path to output file for gapfill analysis result</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>Case ‘KEGG’</dt><dd><p>pd.DataFrame: Table containing the columns ‘bigg_id’ ‘locus_tag’ ‘EC’ ‘KEGG’ ‘name’ ‘GPR’</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Case ‘BioCyc’</dt><dd><dl class="simple">
<dt>tuple: Five tables (1) - (4)</dt><dd><ol class="arabic simple">
<li><dl class="simple">
<dt>pd.DataFrame: Gap fill statistics with the columns </dt><dd><p>‘Missing entity’ ‘Total’ ‘Have BiGG ID’ ‘Can be added’ ‘Notes’</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pd.DataFrame: Genes with the columns </dt><dd><p>‘locus_tag’ ‘protein_id’ ‘model_id’ ‘name’</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pd.DataFrame: Metabolites with the columns </dt><dd><p>‘bigg_id’ ‘name’ ‘BioCyc’ ‘compartment’ ‘Chemical Formula’ ‘InChI-Key’ ‘ChEBI’ ‘charge’</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pd.DataFrame: Reactions with the columns </dt><dd><p>‘bigg_id’ ‘name’ ‘BioCyc’ ‘locus_tag’ ‘Reactants’ ‘Products’ ‘EC’ ‘Fluxes’ ‘Spontaneous?’ 
‘bigg_reaction’</p>
</dd>
</dl>
</li>
</ol>
</dd>
</dl>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Case ‘KEGG+BioCyc’: </dt><dd><dl class="simple">
<dt>tuple: Five tables (1)-(4) from output of ‘BioCyc’ &amp; (5) from output of ‘KEGG’</dt><dd><p>-&gt; Table reactions contains additionally column ‘KEGG’</p>
</dd>
</dl>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="b-add-genes-metabolites-and-reactions-from-an-excel-table-to-a-model">
<h4>b. add genes, metabolites and reactions from an Excel table to a model<a class="headerlink" href="#b-add-genes-metabolites-and-reactions-from-an-excel-table-to-a-model" title="Permalink to this heading"></a></h4>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">refinegems.gapfill.</span></span><span class="sig-name descname"><span class="pre">gapfill_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_libsbml</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">libsbml.Model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_analysis_result</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">libsbml.Model</span></span></span></dt>
<dd><p>Main function to fill gaps in a model from a table</p>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model_libsbml (libModel): Model loaded with libSBML</p></li>
<li><p>gap_analysis_result (str|tuple): Path to Excel file from gap_analysis|Tuple of pd.DataFrames obtained from gap_analysis</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>libModel: Gap filled model</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="c-or-perform-gap-analysis-and-add-the-result-directly-to-a-model">
<h4>c. or perform gap analysis and add the result directly to a model.<a class="headerlink" href="#c-or-perform-gap-analysis-and-add-the-result-directly-to-a-model" title="Permalink to this heading"></a></h4>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>To use the gap analysis and directly add the result to a model, currently, one of the options ‘BioCyc’ or ‘KEGG+BioCyc’ has to be selected.
For all other options the usage of gap_analysis combined with gapfill_model will result in an error.</p>
</div>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">refinegems.gapfill.</span></span><span class="sig-name descname"><span class="pre">gapfill</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">model_libsbml:</span> <span class="pre">libsbml.Model,</span> <span class="pre">gapfill_params:</span> <span class="pre">dict[slice(&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">None)],</span> <span class="pre">filename:</span> <span class="pre">str</span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">pandas.DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">libsbml.Model</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">libsbml.Model</span><span class="p"><span class="pre">]</span></span></span></span></dt>
<dd><div class="line-block">
<div class="line">Main function to fill gaps in a model by comparing the locus tags of the GeneProducts to 
| KEGG/BioCyc/(Genbank) GFF file</div>
</div>
<dl class="simple">
<dt>Args:</dt><dd><ul class="simple">
<li><p>model_libsbml (libModel): Model loaded with libSBML</p></li>
<li><p>gapfill_params (dict): Dictionary obtained from YAML file containing the parameter mappings</p></li>
<li><p>filename (str): Path to output file for gapfill analysis result</p></li>
<li><p>gapfill_model_out (str): Path where gapfilled model should be written to</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>tuple: <code class="docutils literal notranslate"><span class="pre">gap_analysis()</span></code> table(s) (1) &amp; libSBML model (2)</dt><dd><ol class="arabic simple">
<li><p>pd.DataFrame|tuple(pd.DataFrame): Result from function <code class="docutils literal notranslate"><span class="pre">gap_analysis()</span></code></p></li>
<li><p>libModel: Gap filled model</p></li>
</ol>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="relevant-parameters">
<h3>Relevant parameters<a class="headerlink" href="#relevant-parameters" title="Permalink to this heading"></a></h3>
<p>To perform the gap analysis the following parameters are relevant for the config.yaml file:
(See <a class="reference internal" href="#data-acquisition-from-biocyc"><span class="std std-ref">Data acquisition from BioCyc</span></a> on how to obtain the files 1 to 3)</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">gap_analysis</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRUE</span>
<span class="w">    </span><span class="l l-Scalar l-Scalar-Plain">gap_analysis_params</span><span class="p p-Indicator">:</span>
<span class="w">      </span><span class="nt">db_to_compare</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;One</span><span class="nv"> </span><span class="s">of</span><span class="nv"> </span><span class="s">the</span><span class="nv"> </span><span class="s">choices</span><span class="nv"> </span><span class="s">KEGG|BioCyc|KEGG+BioCyc&#39;</span>
<span class="w">      </span><span class="nt">organismid</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;KEGG</span><span class="nv"> </span><span class="s">Organism</span><span class="nv"> </span><span class="s">ID&#39;</span><span class="w"> </span><span class="c1"># Needs to be specified for KEGG</span>
<span class="w">      </span><span class="nt">gff_file</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">RefSeq</span><span class="nv"> </span><span class="s">GFF</span><span class="nv"> </span><span class="s">file&#39;</span><span class="w"> </span><span class="c1"># Needs to be specified for KEGG</span>
<span class="w">      </span><span class="nt">biocyc_files</span><span class="p">:</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;File</span><span class="nv"> </span><span class="s">1:</span><span class="nv"> </span><span class="s">Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">gene</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">reaction</span><span class="nv"> </span><span class="s">mapping</span><span class="nv"> </span><span class="s">table&#39;</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;File</span><span class="nv"> </span><span class="s">2:</span><span class="nv"> </span><span class="s">Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">reaction</span><span class="nv"> </span><span class="s">table&#39;</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;File</span><span class="nv"> </span><span class="s">3:</span><span class="nv"> </span><span class="s">Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">compounds</span><span class="nv"> </span><span class="s">table&#39;</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&#39;File</span><span class="nv"> </span><span class="s">4:</span><span class="nv"> </span><span class="s">Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">protein</span><span class="nv"> </span><span class="s">FASTA</span><span class="nv"> </span><span class="s">file</span><span class="nv"> </span><span class="s">used</span><span class="nv"> </span><span class="s">as</span><span class="nv"> </span><span class="s">input</span><span class="nv"> </span><span class="s">for</span><span class="nv"> </span><span class="s">CarveMe&#39;</span>
</pre></div>
</div>
<p>To add genes, metabolites and reactions from an Excel table to a model the following parameters need to be set:
(The Excel file is either obtained by running gapfill_analysis or created by hand with the same structure as the result file from gapfill_analysis.
An example Excel file to fill in by hand can be found in the cloned repository under ‘data/modelName_gapfill_analysis_date_example.xlsx’)</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">gapfill_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRUE</span>
<span class="w">    </span><span class="l l-Scalar l-Scalar-Plain">gap_analysis_file</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="s">&#39;Path</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">Excel</span><span class="nv"> </span><span class="s">file</span><span class="nv"> </span><span class="s">with</span><span class="nv"> </span><span class="s">which</span><span class="nv"> </span><span class="s">gaps</span><span class="nv"> </span><span class="s">in</span><span class="nv"> </span><span class="s">model</span><span class="nv"> </span><span class="s">should</span><span class="nv"> </span><span class="s">be</span><span class="nv"> </span><span class="s">filled&#39;</span>
</pre></div>
</div>
</section>
<section id="data-acquisition-from-biocyc">
<h3>Data acquisition from BioCyc<a class="headerlink" href="#data-acquisition-from-biocyc" title="Permalink to this heading"></a></h3>
<ol class="arabic">
<li><p>If you have no BioCyc account you will need to create one. See <cite>BioCyc /&gt; Create Free Account &lt;https://biocyc.org/new-account.shtml&gt;</cite> to create an account.</p></li>
<li><p>Then you need to search for the strain of your organism.</p></li>
<li><p>Within the database of your organism you need to click on <cite>Tools</cite> in the menu bar and select <cite>Special SmartTables</cite> under <cite>SmartTables</cite>.
There you need to make an adjustable copy of each of the tables “All genes of &lt;organism&gt;” and “All reactions of &lt;organism&gt;”.</p></li>
<li><p><strong>For the gene to reaction mapping table:</strong></p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Remove all columns except ‘Gene Name’ from the “All genes of &lt;organism&gt;” table,</p></li>
<li><p>then click <cite>choose a transform</cite> and select ‘Reactions of gene’,</p></li>
<li><p>then add the <cite>property</cite> ‘Accession-2’</p></li>
<li><p>and delete the ‘Gene Name’ column.</p></li>
<li><p>After that select the column ‘Accession-2’ and use the filter function in the box on the right side of the page to delete all empty rows.</p></li>
<li><p>Finally, click <cite>Export to Spreadsheet File</cite> from the box on the right side and choose <cite>Frame IDs</cite>.</p></li>
</ol>
</div></blockquote>
</li>
<li><p><strong>For the reactions table:</strong></p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Remove all columns except ‘Reaction’ from the “All reactions of &lt;organism&gt;” table,</p></li>
<li><p>then click <cite>choose a transform</cite>:</p>
<ol class="loweralpha simple">
<li><p>select ‘Reactants of reaction’,</p></li>
<li><p>then select ‘Products of reaction’</p></li>
</ol>
</li>
<li><p>and then choose the <cite>property</cite>:</p></li>
</ol>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>‘EC-Number’,</p></li>
<li><p>then ‘Reaction-Direction’,</p></li>
<li><p>and then ‘Spontaneous?’.</p></li>
</ol>
</div></blockquote>
<ol class="lowerroman simple" start="4">
<li><p>Finally, click <cite>Export to Spreadsheet File</cite> in the box on the right side and choose <cite>Frame IDs</cite>.</p></li>
</ol>
</div></blockquote>
</li>
<li><p><strong>For the metabolites table:</strong></p>
<blockquote>
<div><ol class="lowerroman simple">
<li><p>Use the MetaCyc database to get the table “All compounds of MetaCyc”.</p></li>
<li><p>Remove all columns except ‘Compound’,</p></li>
<li><p>then choose the <cite>property</cite>:</p></li>
</ol>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>‘Object ID’,</p></li>
<li><p>then ‘Chemical Formula’,</p></li>
<li><p>then ‘InChI-Key’,</p></li>
<li><p>and then ‘database links’ &gt; ‘ChEBI’.</p></li>
</ol>
</div></blockquote>
<ol class="lowerroman simple" start="4">
<li><p>Finally, click <cite>Export to Spreadsheet File</cite> in the box on the right side and choose <cite>common names</cite>.</p></li>
</ol>
</div></blockquote>
</li>
</ol>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../main-modules.html" class="btn btn-neutral float-left" title="Main modules" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="growth.html" class="btn btn-neutral float-right" title="Growth simulation with RefineGEMs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Famke Bäuerle and Gwendolyn O. Gusak.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>